--- a/src/main/java/org/bukkit/craftbukkit/block/CraftBlock.java
+++ b/src/main/java/org/bukkit/craftbukkit/block/CraftBlock.java
@@ -1,5 +_,6 @@
 package org.bukkit.craftbukkit.block;
 
+import ca.spottedleaf.moonrise.common.util.TickThread;
 import com.google.common.base.Preconditions;
 import java.util.ArrayList;
 import java.util.Collection;
@@ -80,6 +_,7 @@
     }
 
     public net.minecraft.world.level.block.state.BlockState getBlockState() {
+        if (this.level instanceof ServerLevel serverLevel) TickThread.ensureTickThread(serverLevel, this.position, "Cannot get block asynchronously!");
         return this.level.getBlockState(this.position);
     }
 
@@ -156,6 +_,7 @@
 
     @Override
     public byte getData() {
+        if (this.level instanceof ServerLevel serverLevel) TickThread.ensureTickThread(serverLevel, this.position, "Cannot get block asynchronously!");
         return CraftMagicNumbers.toLegacyData(this.getBlockState());
     }
 
@@ -187,6 +_,7 @@
     }
 
     boolean setBlockState(final net.minecraft.world.level.block.state.BlockState state, final boolean applyPhysics) {
+        if (this.level instanceof ServerLevel serverLevel) TickThread.ensureTickThread(serverLevel, this.position, "Cannot set block asynchronously!");
         return setBlockState(this.level, this.position, state, applyPhysics);
     }
 
@@ -203,6 +_,7 @@
 
     @Override
     public Material getType() {
+        if (this.level instanceof ServerLevel serverLevel) TickThread.ensureTickThread(serverLevel, this.position, "Cannot get block asynchronously!");
         return this.getBlockState().getBukkitMaterial();
     }
 
@@ -520,19 +_,19 @@
 
         // SPIGOT-6895: Call StructureGrowEvent and BlockFertilizeEvent
         List<org.bukkit.craftbukkit.block.CraftBlockState> capturedBlockStates;
-        world.captureTreeGeneration = true;
+        world.captureTreeGenerationThreadLocal.set(true); // ShreddedPaper - use thread local
         TreeType treeType;
         InteractionResult result;
         try {
             result = BoneMealItem.applyBonemeal(context);
         } finally {
-            world.captureTreeGeneration = false;
-
-            capturedBlockStates = new ArrayList<>(world.capturedBlockStates.values());
-            world.capturedBlockStates.clear();
-
-            treeType = SaplingBlock.treeType;
-            SaplingBlock.treeType = null;
+            world.captureTreeGenerationThreadLocal.remove(); // ShreddedPaper - use thread local
+
+            capturedBlockStates = new ArrayList<>(world.capturedBlockStatesThreadLocal.get().values()); // ShreddedPaper - use thread local
+            world.capturedBlockStatesThreadLocal.get().clear(); // ShreddedPaper - use thread local
+
+            treeType = SaplingBlock.treeTypeThreadLocal.get(); // ShreddedPaper - use thread local
+            SaplingBlock.treeTypeThreadLocal.remove(); // ShreddedPaper - use thread local
         }
 
         if (!capturedBlockStates.isEmpty()) {
