--- a/src/main/java/org/bukkit/craftbukkit/block/CraftBlock.java
+++ b/src/main/java/org/bukkit/craftbukkit/block/CraftBlock.java
@@ -1,5 +_,6 @@
 package org.bukkit.craftbukkit.block;
 
+import ca.spottedleaf.moonrise.common.util.TickThread;
 import com.google.common.base.Preconditions;
 import java.util.ArrayList;
 import java.util.Collection;
@@ -75,6 +_,7 @@
     }
 
     public net.minecraft.world.level.block.state.BlockState getNMS() {
+        if (this.world instanceof ServerLevel serverLevel) TickThread.ensureTickThread(serverLevel, this.position, "Cannot get block asynchronously!");
         return this.world.getBlockState(this.position);
     }
 
@@ -160,6 +_,7 @@
 
     @Override
     public byte getData() {
+        if (this.world instanceof ServerLevel serverLevel) TickThread.ensureTickThread(serverLevel, this.position, "Cannot get block asynchronously!");
         net.minecraft.world.level.block.state.BlockState state = this.world.getBlockState(this.position);
         return CraftMagicNumbers.toLegacyData(state);
     }
@@ -192,6 +_,7 @@
     }
 
     boolean setBlockState(final net.minecraft.world.level.block.state.BlockState state, final boolean applyPhysics) {
+        if (this.world instanceof ServerLevel serverLevel) TickThread.ensureTickThread(serverLevel, this.position, "Cannot set block asynchronously!");
         return CraftBlock.setBlockState(this.world, this.position, this.getNMS(), state, applyPhysics);
     }
 
@@ -227,6 +_,7 @@
 
     @Override
     public Material getType() {
+        if (this.world instanceof ServerLevel serverLevel) TickThread.ensureTickThread(serverLevel, this.position, "Cannot get block asynchronously!");
         return this.world.getBlockState(this.position).getBukkitMaterial(); // Paper - optimise getType calls
     }
 
@@ -554,15 +_,15 @@
         UseOnContext context = new UseOnContext(world, null, InteractionHand.MAIN_HAND, Items.BONE_MEAL.getDefaultInstance(), new BlockHitResult(Vec3.ZERO, direction, this.getPosition(), false));
 
         // SPIGOT-6895: Call StructureGrowEvent and BlockFertilizeEvent
-        world.captureTreeGeneration = true;
+        world.captureTreeGenerationThreadLocal.set(true); // ShreddedPaper - use thread local
         InteractionResult result = BoneMealItem.applyBonemeal(context);
-        world.captureTreeGeneration = false;
+        world.captureTreeGenerationThreadLocal.set(false); // ShreddedPaper - use thread local
 
-        if (!world.capturedBlockStates.isEmpty()) {
-            TreeType treeType = SaplingBlock.treeType;
-            SaplingBlock.treeType = null;
-            List<BlockState> states = new ArrayList<>(world.capturedBlockStates.values());
-            world.capturedBlockStates.clear();
+        if (!world.capturedBlockStatesThreadLocal.get().isEmpty()) { // ShreddedPaper - use thread local
+            TreeType treeType = SaplingBlock.treeTypeThreadLocal.get(); // ShreddedPaper - use thread local
+            SaplingBlock.treeTypeThreadLocal.remove(); // ShreddedPaper - use thread local
+            List<BlockState> states = new ArrayList<>(world.capturedBlockStatesThreadLocal.get().values()); // ShreddedPaper - use thread local
+            world.capturedBlockStatesThreadLocal.get().clear(); // ShreddedPaper - use thread local
             StructureGrowEvent structureEvent = null;
 
             if (treeType != null) {
