--- a/ca/spottedleaf/moonrise/patches/chunk_system/level/entity/server/ServerEntityLookup.java
+++ b/ca/spottedleaf/moonrise/patches/chunk_system/level/entity/server/ServerEntityLookup.java
@@ -22,7 +_,7 @@
     private static final Entity[] EMPTY_ENTITY_ARRAY = new Entity[0];
 
     private final ServerLevel serverWorld;
-    public final ReferenceList<Entity> trackerEntities = new ReferenceList<>(EMPTY_ENTITY_ARRAY); // Moonrise - entity tracker
+    // public final ReferenceList<Entity> trackerEntities = new ReferenceList<>(EMPTY_ENTITY_ARRAY); // Moonrise - entity tracker // ShreddedPaper - remove trackerEntities, handled ourselves
 
     // Vanilla does not increment ticket timeouts if the chunk is progressing in generation. They made this change in 1.21.6 so that the ender pearl
     // ticket does not expire if the chunk fails to generate before the timeout expires. Rather than blindly adjusting the entire system behavior
@@ -112,14 +_,14 @@
     @Override
     protected void entityStartLoaded(final Entity entity) {
         // Moonrise start - entity tracker
-        this.trackerEntities.add(entity);
+        // this.trackerEntities.add(entity); // ShreddedPaper - remove trackerEntities, handled ourselves
         // Moonrise end - entity tracker
     }
 
     @Override
     protected void entityEndLoaded(final Entity entity) {
         // Moonrise start - entity tracker
-        this.trackerEntities.remove(entity);
+        // this.trackerEntities.remove(entity); // ShreddedPaper - remove trackerEntities, handled ourselves
         // Moonrise end - entity tracker
     }
 
