--- a/src/main/java/io/papermc/paper/command/brigadier/bukkit/BukkitCommandNode.java
+++ b/src/main/java/io/papermc/paper/command/brigadier/bukkit/BukkitCommandNode.java
@@ -8,6 +_,8 @@
 import com.mojang.brigadier.suggestion.Suggestions;
 import com.mojang.brigadier.suggestion.SuggestionsBuilder;
 import com.mojang.brigadier.tree.LiteralCommandNode;
+import io.multipaper.shreddedpaper.threading.SynchronousPluginExecution;
+import io.multipaper.shreddedpaper.util.ObjectHolder;
 import io.papermc.paper.command.brigadier.CommandSourceStack;
 import java.util.ArrayList;
 import net.kyori.adventure.text.Component;
@@ -23,6 +_,8 @@
 import java.util.List;
 import java.util.concurrent.CompletableFuture;
 import java.util.logging.Level;
+
+import org.bukkit.command.PluginCommand;
 import org.bukkit.entity.Player;
 import org.bukkit.event.server.TabCompleteEvent;
 
@@ -80,7 +_,14 @@
             String[] args = org.apache.commons.lang3.StringUtils.split(content, ' '); // fix adjacent spaces (from console/plugins) causing empty array elements
 
             // Note: we don't return the result of target.execute as thats success / failure, we return handled (true) or not handled (false)
-            this.command.execute(sender, this.literal, Arrays.copyOfRange(args, 1, args.length));
+            // ShreddedPaper start - run unsupported plugins in sync
+            Runnable runnable = () -> this.command.execute(sender, this.literal, Arrays.copyOfRange(args, 1, args.length));
+            if (this.command instanceof PluginCommand pluginCommand) {
+                SynchronousPluginExecution.executeNoException(pluginCommand.getPlugin(), runnable::run);
+            } else {
+                runnable.run();
+            }
+            // ShreddedPaper end
 
             // return true as command was handled
             return 1;
@@ -103,32 +_,45 @@
             org.bukkit.command.CommandSender sender = context.getSource().getSender();
             String[] args = builder.getRemaining().split(" ", -1); // We need the command included -- Set limit to -1, allow for trailing spaces
 
-            List<String> results = null;
+            ObjectHolder<List<String>> results = new ObjectHolder<>(); // ShreddedPaper
             Location pos = context.getSource().getLocation();
             try {
-                results = this.command.tabComplete(sender, this.literal, args, pos.clone());
+                // ShreddedPaper start - run unsupported plugins in sync
+                try {
+                    SynchronousPluginExecution.RunnableWithException runnable = () -> results.set(this.command.tabComplete(sender, this.literal, args, pos.clone()));
+                    if (this.command instanceof PluginCommand pluginCommand) {
+                        SynchronousPluginExecution.execute(pluginCommand.getPlugin(), runnable);
+                    } else {
+                        runnable.run();
+                    }
+                } catch (Error | RuntimeException e) {
+                    throw e;
+                } catch (Exception e) {
+                    throw new RuntimeException(e);
+                }
+                // ShreddedPaper end
             } catch (CommandException ex) {
                 sender.sendMessage(Component.text("An internal error occurred while attempting to tab-complete this command", NamedTextColor.RED));
                 Bukkit.getServer().getLogger().log(Level.SEVERE, "Exception when " + sender.getName() + " attempted to tab complete " + builder.getRemaining(), ex);
             }
 
             if (sender instanceof final Player player) {
-                TabCompleteEvent tabEvent = new org.bukkit.event.server.TabCompleteEvent(player, builder.getInput(), results != null ? results : new ArrayList<>(), true, pos); // Paper - AsyncTabCompleteEvent
+                TabCompleteEvent tabEvent = new org.bukkit.event.server.TabCompleteEvent(player, builder.getInput(), results.get() != null ? results.get() : new ArrayList<>(), true, pos); // Paper - AsyncTabCompleteEvent // ShreddedPaper
                 if (!tabEvent.callEvent()) {
-                    results = null;
+                    results.set(null); // ShreddedPaper
                 } else {
-                    results = tabEvent.getCompletions();
+                    results.set(tabEvent.getCompletions()); // ShreddedPaper
                 }
             }
             // Paper end
-            if (results == null) {
+            if (results.get() == null) { // ShreddedPaper
                 return builder.buildFuture();
             }
 
             // Defaults to sub nodes, but we have just one giant args node, so offset accordingly
             builder = builder.createOffset(builder.getInput().lastIndexOf(' ') + 1);
 
-            for (String s : results) {
+            for (String s : results.get()) { // ShreddedPaper
                 builder.suggest(s);
             }
 
