--- a/src/main/java/org/bukkit/plugin/SimplePluginManager.java
+++ b/src/main/java/org/bukkit/plugin/SimplePluginManager.java
@@ -22,6 +_,7 @@
 import java.util.Map;
 import java.util.Set;
 import java.util.WeakHashMap;
+import java.util.concurrent.ConcurrentHashMap;
 import java.util.logging.Level;
 import java.util.regex.Matcher;
 import java.util.regex.Pattern;
@@ -57,8 +_,8 @@
     // Paper start
     public final Map<String, Permission> permissions = new HashMap<String, Permission>();
     public final Map<Boolean, Set<Permission>> defaultPerms = new LinkedHashMap<Boolean, Set<Permission>>();
-    public final Map<String, Map<Permissible, Boolean>> permSubs = new HashMap<String, Map<Permissible, Boolean>>();
-    public final Map<Boolean, Map<Permissible, Boolean>> defSubs = new HashMap<Boolean, Map<Permissible, Boolean>>();
+    public final Map<String, Map<Permissible, Boolean>> permSubs = new ConcurrentHashMap<String, Map<Permissible, Boolean>>(); // ShreddedPaper - thread-safe permission manager
+    public final Map<Boolean, Map<Permissible, Boolean>> defSubs = new ConcurrentHashMap<Boolean, Map<Permissible, Boolean>>(); // ShreddedPaper - thread-safe permission manager
     public PluginManager paperPluginManager;
     // Paper end
     private boolean useTimings = false;
