--- a/net/minecraft/server/level/DistanceManager.java
+++ b/net/minecraft/server/level/DistanceManager.java
@@ -34,7 +_,7 @@
 public abstract class DistanceManager implements ca.spottedleaf.moonrise.patches.chunk_system.level.chunk.ChunkSystemDistanceManager, ca.spottedleaf.moonrise.patches.chunk_tick_iteration.ChunkTickDistanceManager { // Paper - rewrite chunk system // Paper - chunk tick iteration optimisation
     private static final Logger LOGGER = LogUtils.getLogger();
     static final int PLAYER_TICKET_LEVEL = ChunkLevel.byStatus(FullChunkStatus.ENTITY_TICKING);
-    final Long2ObjectMap<ObjectSet<ServerPlayer>> playersPerChunk = new Long2ObjectOpenHashMap<>();
+    // final Long2ObjectMap<ObjectSet<ServerPlayer>> playersPerChunk = new Long2ObjectOpenHashMap<>(); // ShreddedPaper - Remove unused DistanceManager methods
     // Paper - rewrite chunk system
     public final TicketStorage ticketStorage;
     // Paper - chunk tick iteration optimisation
@@ -99,25 +_,29 @@
     }
 
     public void addPlayer(SectionPos sectionPos, ServerPlayer player) {
-        ChunkPos chunkPos = sectionPos.chunk();
-        long packedChunkPos = chunkPos.toLong();
-        this.playersPerChunk.computeIfAbsent(packedChunkPos, l -> new ObjectOpenHashSet<>()).add(player);
-        // Paper - chunk tick iteration optimisation
-        // Paper - rewrite chunk system
+        // ShreddedPaper start - Remove unused DistanceManager methods
+//        ChunkPos chunkPos = sectionPos.chunk();
+//        long packedChunkPos = chunkPos.toLong();
+//        this.playersPerChunk.computeIfAbsent(packedChunkPos, l -> new ObjectOpenHashSet<>()).add(player);
+//        // Paper - chunk tick iteration optimisation
+//        // Paper - rewrite chunk system
+        // ShreddedPaper end - Remove unused DistanceManager methods
     }
 
     public void removePlayer(SectionPos sectionPos, ServerPlayer player) {
-        ChunkPos chunkPos = sectionPos.chunk();
-        long packedChunkPos = chunkPos.toLong();
-        ObjectSet<ServerPlayer> set = this.playersPerChunk.get(packedChunkPos);
-        // Paper start - some state corruption happens here, don't crash, clean up gracefully
-        if (set != null) set.remove(player);
-        if (set == null || set.isEmpty()) {
-        // Paper end - some state corruption happens here, don't crash, clean up gracefully
-            this.playersPerChunk.remove(packedChunkPos);
-            // Paper - chunk tick iteration optimisation
-            // Paper - rewrite chunk system
-        }
+        // ShreddedPaper start - Remove unused DistanceManager methods
+//        ChunkPos chunkPos = sectionPos.chunk();
+//        long packedChunkPos = chunkPos.toLong();
+//        ObjectSet<ServerPlayer> set = this.playersPerChunk.get(packedChunkPos);
+//        // Paper start - some state corruption happens here, don't crash, clean up gracefully
+//        if (set != null) set.remove(player);
+//        if (set == null || set.isEmpty()) {
+//        // Paper end - some state corruption happens here, don't crash, clean up gracefully
+//            this.playersPerChunk.remove(packedChunkPos);
+//            // Paper - chunk tick iteration optimisation
+//            // Paper - rewrite chunk system
+//        }
+        // ShreddedPaper end - Remove unused DistanceManager methods
     }
 
     private int getPlayerTicketLevel() {
@@ -230,8 +_,11 @@
         }
 
         private boolean havePlayer(long chunkPos) {
-            ObjectSet<ServerPlayer> set = DistanceManager.this.playersPerChunk.get(chunkPos);
-            return set != null && !set.isEmpty();
+            // ShreddedPaper start - Remove unused DistanceManager methods
+            throw new UnsupportedOperationException("REPORT THIS STACK TRACE TO SHREDDEDPAPER"); // ShreddedPaper - Remove unused DistanceManager methods
+//            ObjectSet<ServerPlayer> set = DistanceManager.this.playersPerChunk.get(chunkPos);
+//            return set != null && !set.isEmpty();
+            // ShreddedPaper end - Remove unused DistanceManager methods
         }
 
         public void runAllUpdates() {
