--- a/net/minecraft/server/level/ServerPlayer.java
+++ b/net/minecraft/server/level/ServerPlayer.java
@@ -23,6 +_,9 @@
 import java.util.UUID;
 import java.util.concurrent.CompletableFuture;
 import java.util.stream.Collectors;
+
+import io.multipaper.shreddedpaper.ShreddedPaper;
+import io.multipaper.shreddedpaper.util.ObjectHolder;
 import net.minecraft.ChatFormatting;
 import net.minecraft.CrashReport;
 import net.minecraft.CrashReportCategory;
@@ -252,6 +_,7 @@
     private @Nullable Entity camera;
     private boolean isChangingDimension;
     public boolean seenCredits = false;
+    public boolean hasMaximumTrackerBypassPermission; // ShreddedPaper
     private final ServerRecipeBook recipeBook;
     private @Nullable Vec3 levitationStartPos;
     private int levitationStartTime;
@@ -266,6 +_,7 @@
     private @Nullable Entity currentExplosionCause;
     private SectionPos lastSectionPos = SectionPos.of(0, 0, 0);
     private ChunkTrackingView chunkTrackingView = ChunkTrackingView.EMPTY;
+    public ChunkPos previousChunkPosRegion; // ShreddedPaper
     private ServerPlayer.@Nullable RespawnConfig respawnConfig;
     private final TextFilter textFilter;
     private boolean textFilteringEnabled;
@@ -648,6 +_,8 @@
         if (!pearlLevelKey.isEmpty()) {
             ServerLevel pearlLevel = this.level().getServer().getLevel(pearlLevelKey.get());
             if (pearlLevel != null) {
+                BlockPos enderpearlPos = BlockPos.containing(pearlInput.read("Pos", Vec3.CODEC).orElse(Vec3.ZERO)); // ShreddedPaper - spawn ender pearl on ender pearl thread
+                ShreddedPaper.ensureSync(pearlLevel, enderpearlPos, () -> { // ShreddedPaper - spawn ender pearl on ender pearl thread
                 Entity pearl = EntityType.loadEntityRecursive(
                     pearlInput, pearlLevel, EntitySpawnReason.LOAD, entity -> !pearlLevel.addWithUUID(entity) ? null : entity
                 );
@@ -656,6 +_,7 @@
                 } else {
                     LOGGER.warn("Failed to spawn player ender pearl in level ({}), skipping", pearlLevelKey.get());
                 }
+                }); // ShreddedPaper - spawn ender pearl on ender pearl thread
             } else {
                 LOGGER.warn("Trying to load ender pearl without level ({}) being loaded, skipping", pearlLevelKey.get());
             }
@@ -755,9 +_,11 @@
 
         Entity camera = this.getCamera();
         if (camera != this) {
-            if (camera.isAlive()) {
+            if (camera.isAlive() && camera.level() == this.level()) { // ShreddedPaper - ensure telelport to camera sync
+                ShreddedPaper.ensureSync(this, camera, () -> { // ShreddedPaper - ensure telelport to camera sync
                 this.absSnapTo(camera.getX(), camera.getY(), camera.getZ(), camera.getYRot(), camera.getXRot());
                 this.level().getChunkSource().move(this);
+                }); // ShreddedPaper - ensure telelport to camera sync
                 if (this.wantsToStopRiding()) {
                     this.setCamera(this);
                 }
@@ -1355,45 +_,49 @@
     public record RespawnResult(TeleportTransition transition, boolean isBedSpawn, boolean isAnchorSpawn) {
     }
 
-    public @Nullable TeleportTransition findRespawnPositionAndUseSpawnBlock(
+    public CompletableFuture<TeleportTransition> findRespawnPositionAndUseSpawnBlockAsync( // ShreddedPaper - respawn async
         final boolean consumeSpawnBlock, final TeleportTransition.PostTeleportTransition postTeleportTransition,
         final org.bukkit.event.player.PlayerRespawnEvent.RespawnReason respawnReason
     ) {
-        RespawnResult result = this.findRespawnPositionAndUseSpawnBlock0(consumeSpawnBlock, postTeleportTransition, respawnReason);
+        return this.findRespawnPositionAndUseSpawnBlock0Async(consumeSpawnBlock, postTeleportTransition, respawnReason).thenApply(result -> { // ShreddedPaper - respawn async
         return result == null ? null : result.transition();
+        }); // ShreddedPaper - respawn async
     }
 
-    public @Nullable RespawnResult findRespawnPositionAndUseSpawnBlock0(
+    public CompletableFuture<@Nullable RespawnResult> findRespawnPositionAndUseSpawnBlock0Async( // ShreddedPaper - respawn async
         final boolean consumeSpawnBlock, final TeleportTransition.PostTeleportTransition postTeleportTransition,
         final org.bukkit.event.player.PlayerRespawnEvent.RespawnReason respawnReason
     ) {
-        TeleportTransition teleportTransition;
-        boolean isBedSpawn = false;
-        boolean isAnchorSpawn = false;
-        Runnable consumeAnchorCharge = null;
+        // TeleportTransition teleportTransition; // ShreddedPaper - respawn async
+        ObjectHolder<Boolean> isBedSpawn = new ObjectHolder<>(false); // ShreddedPaper - respawn async
+        ObjectHolder<Boolean> isAnchorSpawn = new ObjectHolder<>(false); // ShreddedPaper - respawn async
+        ObjectHolder<Runnable> consumeAnchorCharge = new ObjectHolder<>(null); // ShreddedPaper - respawn async
         // Paper end
         ServerPlayer.RespawnConfig respawnConfig = this.getRespawnConfig();
         ServerLevel respawnLevel = this.server.getLevel(ServerPlayer.RespawnConfig.getDimensionOrDefault(respawnConfig));
+        CompletableFuture<TeleportTransition> future; // ShreddedPaper - respawn async
         if (respawnLevel != null && respawnConfig != null) {
-            Optional<ServerPlayer.RespawnPosAngle> respawn = findRespawnAndUseSpawnBlock(respawnLevel, respawnConfig, consumeSpawnBlock);
+            future = findRespawnAndUseSpawnBlockAsync(respawnLevel, respawnConfig, consumeSpawnBlock).thenCompose(respawn -> { // ShreddedPaper - respawn async
             if (respawn.isPresent()) {
                 ServerPlayer.RespawnPosAngle respawnPosAngle = respawn.get();
                 // CraftBukkit start
-                isBedSpawn = respawnPosAngle.isBedSpawn();
-                isAnchorSpawn = respawnPosAngle.isAnchorSpawn();
-                consumeAnchorCharge = respawnPosAngle.consumeAnchorCharge();
-                teleportTransition = new TeleportTransition(
+                isBedSpawn.set(respawnPosAngle.isBedSpawn()); // ShreddedPaper - respawn async
+                isAnchorSpawn.set(respawnPosAngle.isAnchorSpawn()); // ShreddedPaper - respawn async
+                consumeAnchorCharge.set(respawnPosAngle.consumeAnchorCharge()); // ShreddedPaper - respawn async
+                return CompletableFuture.completedFuture(new TeleportTransition( // ShreddedPaper - respawn async
                     respawnLevel, respawnPosAngle.position(), Vec3.ZERO, respawnPosAngle.yaw(), respawnPosAngle.pitch(), postTeleportTransition
-                );
+                )); // ShreddedPaper - respawn async
                 // CraftBukkit end
             } else {
-                teleportTransition = TeleportTransition.missingRespawnBlock(this, postTeleportTransition); // CraftBukkit
+                return TeleportTransition.missingRespawnBlockAsync(this, postTeleportTransition); // CraftBukkit // ShreddedPaper - respawn async
             }
+            }); // ShreddedPaper - respawn async
         } else {
             // CraftBukkit start
-            teleportTransition = TeleportTransition.createDefault(this, postTeleportTransition);
+            future = TeleportTransition.createDefaultAsync(this, postTeleportTransition); // ShreddedPaper - respawn async
         }
 
+        return future.thenApply(teleportTransition -> { // ShreddedPaper - respawn async
         org.bukkit.entity.Player respawnPlayer = this.getBukkitEntity();
         org.bukkit.Location location = org.bukkit.craftbukkit.util.CraftLocation.toBukkit(
             teleportTransition.position(),
@@ -1406,8 +_,8 @@
         org.bukkit.event.player.PlayerRespawnEvent respawnEvent = new org.bukkit.event.player.PlayerRespawnEvent(
             respawnPlayer,
             location,
-            isBedSpawn,
-            isAnchorSpawn,
+            isBedSpawn.get(), // ShreddedPaper - respawn async
+            isAnchorSpawn.get(), // ShreddedPaper - respawn async
             teleportTransition.missingRespawnBlock(),
             respawnReason
         );
@@ -1420,17 +_,18 @@
         // Spigot end
 
         // Paper start - consume anchor charge if location hasn't changed
-        if (location.equals(respawnEvent.getRespawnLocation()) && consumeAnchorCharge != null) {
-            consumeAnchorCharge.run();
+        if (location.equals(respawnEvent.getRespawnLocation()) && consumeAnchorCharge.get() != null) {
+            consumeAnchorCharge.get().run(); // ShreddedPaper - respawn async
         }
         // Paper end - consume anchor charge if location hasn't changed
         location = respawnEvent.getRespawnLocation();
 
         return new RespawnResult(
             teleportTransition.withChangedLocation(location),
-            isBedSpawn,
-            isAnchorSpawn
+            isBedSpawn.get(), // ShreddedPaper - respawn async
+            isAnchorSpawn.get() // ShreddedPaper - respawn async
         );
+        }); // ShreddedPaper - respawn async
         // CraftBukkit end
     }
 
@@ -1452,7 +_,7 @@
         super.onAttributeUpdated(attribute);
     }
 
-    public static Optional<ServerPlayer.RespawnPosAngle> findRespawnAndUseSpawnBlock(
+    public static CompletableFuture<Optional<ServerPlayer.RespawnPosAngle>> findRespawnAndUseSpawnBlockAsync( // ShreddedPaper - respawn async
         final ServerLevel level, final ServerPlayer.RespawnConfig respawnConfig, final boolean consumeSpawnBlock
     ) {
         LevelData.RespawnData respawnData = respawnConfig.respawnData;
@@ -1460,6 +_,7 @@
         float yaw = respawnData.yaw();
         float pitch = respawnData.pitch();
         boolean forced = respawnConfig.forced;
+        return CompletableFuture.supplyAsync(() -> { // ShreddedPaper - respawn async
         BlockState blockState = level.getBlockState(pos);
         Block block = blockState.getBlock();
         if (block instanceof RespawnAnchorBlock && (forced || blockState.getValue(RespawnAnchorBlock.CHARGE) > 0) && RespawnAnchorBlock.canSetSpawn(level, pos)
@@ -1468,7 +_,16 @@
             Runnable consumeAnchorCharge = null; // Paper - Fix SPIGOT-5989 (don't use charge until after respawn event)
             Optional<Vec3> standUpPosition = RespawnAnchorBlock.findStandUpPosition(EntityTypes.PLAYER, level, pos);
             if (!forced && consumeSpawnBlock && standUpPosition.isPresent()) {
-                consumeAnchorCharge = () -> level.setBlock(pos, blockState.setValue(RespawnAnchorBlock.CHARGE, blockState.getValue(RespawnAnchorBlock.CHARGE) - 1), Block.UPDATE_ALL); // Paper - Fix SPIGOT-5989 (don't use charge until after respawn event)
+                // ShreddedPaper start - respawn anchor block update async
+                consumeAnchorCharge = () -> {
+                    ShreddedPaper.ensureSync(level, pos, () -> {
+                        BlockState blockState2 = level.getBlockState(pos); // Use the latest block state in case it has been modified since
+                        if (blockState2.getBlock() instanceof RespawnAnchorBlock && blockState2.getValue(RespawnAnchorBlock.CHARGE) > 0) { // Ensure this block is still a respawn anchor block
+                            level.setBlock(pos, blockState2.setValue(RespawnAnchorBlock.CHARGE, blockState2.getValue(RespawnAnchorBlock.CHARGE) - 1), Block.UPDATE_ALL); // Paper - Fix SPIGOT-5989 (don't use charge until after respawn event)
+                        }
+                    });
+                };
+                // ShreddedPaper end - respawn anchor block update async
             }
             final Runnable finalConsumeAnchorCharge = consumeAnchorCharge; // Paper - Fix SPIGOT-5989
 
@@ -1490,6 +_,7 @@
                 ? Optional.of(new ServerPlayer.RespawnPosAngle(new Vec3(pos.getX() + 0.5, pos.getY() + 0.1, pos.getZ() + 0.5), yaw, pitch, false, false, null)) // Paper - Fix SPIGOT-5989
                 : Optional.empty();
         }
+        }, r -> ShreddedPaper.ensureSync(level, pos, r));
     }
 
     public void showEndCredits() {
@@ -2675,6 +_,7 @@
     }
 
     public void setCamera(final @Nullable Entity newCamera) {
+        ShreddedPaper.ensureSync(this, newCamera == null ? this : newCamera, () -> { // ShreddedPaper - run on entity's thread
         Entity oldCamera = this.getCamera();
         this.camera = newCamera == null ? this : newCamera;
         if (oldCamera != this.camera) {
@@ -2704,6 +_,7 @@
             this.connection.send(new ClientboundSetCameraPacket(this.camera));
             this.connection.resetPosition();
         }
+        });  // ShreddedPaper - run on entity's thread
     }
 
     @Override
@@ -2973,6 +_,8 @@
             for (MobEffectInstance effect : livingEntity.getActiveEffects()) {
                 this.connection.send(new ClientboundRemoveMobEffectPacket(oldVehicle.getId(), effect.getEffect()));
             }
+
+            if (oldVehicle.moonrise$getTrackedEntity() != null) oldVehicle.moonrise$getTrackedEntity().serverEntity.sendChanges(); // ShreddedPaper - send changes now
         }
 
         if (oldVehicle != null) {
