--- a/net/minecraft/server/level/ServerLevel.java
+++ b/net/minecraft/server/level/ServerLevel.java
@@ -1,10 +_,20 @@
 package net.minecraft.server.level;
 
+import ca.spottedleaf.moonrise.common.util.TickThread;
+import ca.spottedleaf.moonrise.common.util.WorldUtil;
 import com.google.common.annotations.VisibleForTesting;
 import com.google.common.collect.Lists;
 import com.mojang.datafixers.DataFixer;
 import com.mojang.datafixers.util.Pair;
 import com.mojang.logging.LogUtils;
+import io.multipaper.shreddedpaper.ShreddedPaper;
+import io.multipaper.shreddedpaper.config.ShreddedPaperConfiguration;
+import io.multipaper.shreddedpaper.region.LevelChunkRegion;
+import io.multipaper.shreddedpaper.region.LevelTicksRegionProxy;
+import io.multipaper.shreddedpaper.region.RegionPos;
+import io.multipaper.shreddedpaper.threading.ShreddedPaperRegionScheduler;
+import io.multipaper.shreddedpaper.threading.ShreddedPaperTickThread;
+import io.multipaper.shreddedpaper.util.NoOpServerWaypointManager;
 import it.unimi.dsi.fastutil.ints.Int2ObjectMap;
 import it.unimi.dsi.fastutil.ints.Int2ObjectOpenHashMap;
 import it.unimi.dsi.fastutil.longs.LongSet;
@@ -26,7 +_,10 @@
 import java.util.Optional;
 import java.util.Set;
 import java.util.UUID;
+import java.util.concurrent.CompletableFuture;
+import java.util.concurrent.CopyOnWriteArrayList;
 import java.util.concurrent.Executor;
+import java.util.concurrent.Executors;
 import java.util.function.BooleanSupplier;
 import java.util.function.Function;
 import java.util.function.Predicate;
@@ -191,11 +_,11 @@
     private static final Logger LOGGER = LogUtils.getLogger();
     private static final int EMPTY_TIME_NO_TICK = 300;
     private static final int MAX_SCHEDULED_TICKS_PER_TICK = 65536;
-    private final List<ServerPlayer> players = Lists.newArrayList();
+    private final List<ServerPlayer> players = new CopyOnWriteArrayList<>(); // ShreddedPaper - thread-safe (players should not be changing worlds often, thus copy-on-write is sufficient)
     public final ServerChunkCache chunkSource;
     private final MinecraftServer server;
     public final io.papermc.paper.world.saveddata.PaperLevelOverrides serverLevelData; // Paper - type
-    private final EntityTickList entityTickList = new EntityTickList();
+    // private final EntityTickList entityTickList = new EntityTickList(); // ShreddedPaper - moved into each region
     private final ServerWaypointManager waypointManager;
     private EnvironmentAttributeSystem environmentAttributes;
     // Paper - rewrite chunk system
@@ -204,15 +_,15 @@
     private final SleepStatus sleepStatus;
     private int emptyTime;
     private final PortalForcer portalForcer;
-    private final LevelTicks<Block> blockTicks = new LevelTicks<>(this::isPositionTickingWithEntitiesLoaded);
-    private final LevelTicks<Fluid> fluidTicks = new LevelTicks<>(this::isPositionTickingWithEntitiesLoaded);
+    public final LevelTicksRegionProxy<Block> blockTicks = new LevelTicksRegionProxy<>(this::isPositionTickingWithEntitiesLoaded); // ShreddedPaper
+    public final LevelTicksRegionProxy<Fluid> fluidTicks = new LevelTicksRegionProxy<>(this::isPositionTickingWithEntitiesLoaded); // ShreddedPaper
     private final PathTypeCache pathTypesByPosCache = new PathTypeCache();
-    private final Set<Mob> navigatingMobs = new ObjectOpenHashSet<>();
-    private volatile boolean isUpdatingNavigations;
+    // private final Set<Mob> navigatingMobs = new ObjectOpenHashSet<>(); // ShreddedPaper - handled ourselves
+    private final ThreadLocal<Boolean> isUpdatingNavigations = ThreadLocal.withInitial(() -> false); // ShreddedPaper - make thread local
     protected final Raids raids;
-    private final ObjectLinkedOpenHashSet<BlockEventData> blockEvents = new ObjectLinkedOpenHashSet<>();
-    private final List<BlockEventData> blockEventsToReschedule = new ArrayList<>(64);
-    private boolean handlingTick;
+    // private final ObjectLinkedOpenHashSet<BlockEventData> blockEvents = new ObjectLinkedOpenHashSet<>(); // ShreddedPaper - moved into each region
+    private final ThreadLocal<List<BlockEventData>> blockEventsToRescheduleThreadLocal = ThreadLocal.withInitial(() -> new ArrayList<>(64)); // ShreddedPaper
+    public ThreadLocal<Boolean> handlingTickThreadLocal = ThreadLocal.withInitial(() -> false); // ShreddedPaper
     private final List<CustomSpawner> customSpawners;
     private @Nullable EnderDragonFight dragonFight;
     private final Int2ObjectMap<EnderDragonPart> dragonParts = new Int2ObjectOpenHashMap<>();
@@ -221,6 +_,8 @@
     private final boolean tickTime;
     private double preciseTime; // Purpur - Configurable daylight cycle
     private boolean forceTime; // Purpur - Configurable daylight cycle
+    public final ShreddedPaperRegionScheduler chunkScheduler = new ShreddedPaperRegionScheduler(); // ShreddedPaper
+    public final Executor levelThread; // ShreddedPaper
     private final LevelDebugSynchronizers debugSynchronizers = new LevelDebugSynchronizers(this);
 
     // CraftBukkit start
@@ -234,7 +_,7 @@
     public final WorldGenSettings worldGenSettings;
     public boolean hasPhysicsEvent = true; // Paper - BlockPhysicsEvent
     public boolean hasEntityMoveEvent; // Paper - Add EntityMoveEvent
-    private final alternate.current.wire.WireHandler wireHandler = new alternate.current.wire.WireHandler(this); // Paper - optimize redstone (Alternate Current)
+    private final ThreadLocal<alternate.current.wire.WireHandler> wireHandlerThreadLocal = ThreadLocal.withInitial(() -> new alternate.current.wire.WireHandler(this)); // Paper - optimize redstone (Alternate Current) // ShreddedPaper - ThreadLocal
     public boolean hasRidableMoveEvent = false; // Purpur - Ridables
 
     @Override
@@ -321,9 +_,9 @@
     private long tickedBlocksOrFluids;
     private final ca.spottedleaf.moonrise.common.misc.NearbyPlayers nearbyPlayers = new ca.spottedleaf.moonrise.common.misc.NearbyPlayers((ServerLevel)(Object)this);
     private static final LevelChunk[] EMPTY_LEVEL_CHUNKS = new LevelChunk[0];
-    private final ca.spottedleaf.moonrise.common.list.ReferenceList<LevelChunk> loadedChunks = new ca.spottedleaf.moonrise.common.list.ReferenceList<>(EMPTY_LEVEL_CHUNKS);
-    private final ca.spottedleaf.moonrise.common.list.ReferenceList<LevelChunk> tickingChunks = new ca.spottedleaf.moonrise.common.list.ReferenceList<>(EMPTY_LEVEL_CHUNKS);
-    private final ca.spottedleaf.moonrise.common.list.ReferenceList<LevelChunk> entityTickingChunks = new ca.spottedleaf.moonrise.common.list.ReferenceList<>(EMPTY_LEVEL_CHUNKS);
+    // private final ca.spottedleaf.moonrise.common.list.ReferenceList<LevelChunk> loadedChunks = new ca.spottedleaf.moonrise.common.list.ReferenceList<>(EMPTY_LEVEL_CHUNKS); // ShreddedPaper - loadedChunks is unused, remove
+    // private final ca.spottedleaf.moonrise.common.list.ReferenceList<LevelChunk> tickingChunks = new ca.spottedleaf.moonrise.common.list.ReferenceList<>(EMPTY_LEVEL_CHUNKS); // ShreddedPaper - unused, ensure it remains that way
+    // private final ca.spottedleaf.moonrise.common.list.ReferenceList<LevelChunk> entityTickingChunks = new ca.spottedleaf.moonrise.common.list.ReferenceList<>(EMPTY_LEVEL_CHUNKS); // ShreddedPaper - unused, ensure it remains that way
 
     @Override
     public final LevelChunk moonrise$getFullChunkIfLoaded(final int chunkX, final int chunkZ) {
@@ -497,19 +_,21 @@
         return this.nearbyPlayers;
     }
 
-    @Override
-    public final ca.spottedleaf.moonrise.common.list.ReferenceList<LevelChunk> moonrise$getLoadedChunks() {
-        return this.loadedChunks;
-    }
+    // ShreddedPaper start - loadedChunks is unused, remove
+//    @Override
+//    public final ca.spottedleaf.moonrise.common.list.ReferenceList<LevelChunk> moonrise$getLoadedChunks() {
+//        return this.loadedChunks;
+//    }
+    // ShreddedPaper start - loadedChunks is unused, remove
 
     @Override
     public final ca.spottedleaf.moonrise.common.list.ReferenceList<LevelChunk> moonrise$getTickingChunks() {
-        return this.tickingChunks;
+        throw new UnsupportedOperationException("Removed by ShreddedPaper"); // return this.tickingChunks; // ShreddedPaper - unused, ensure it remains that way
     }
 
     @Override
     public final ca.spottedleaf.moonrise.common.list.ReferenceList<LevelChunk> moonrise$getEntityTickingChunks() {
-        return this.entityTickingChunks;
+        throw new UnsupportedOperationException("Removed by ShreddedPaper"); // return this.entityTickingChunks; // ShreddedPaper - unused, ensure it remains that way
     }
 
     @Override
@@ -535,75 +_,77 @@
    }
     // Paper end - rewrite chunk system
     // Paper start - chunk tick iteration
-    private final ca.spottedleaf.moonrise.common.list.ReferenceList<LevelChunk> playerTickingChunks = new ca.spottedleaf.moonrise.common.list.ReferenceList<>(EMPTY_LEVEL_CHUNKS);
-    private final it.unimi.dsi.fastutil.longs.Long2IntOpenHashMap playerTickingRequests = new it.unimi.dsi.fastutil.longs.Long2IntOpenHashMap();
-
-    @Override
-    public final ca.spottedleaf.moonrise.common.list.ReferenceList<LevelChunk> moonrise$getPlayerTickingChunks() {
-        return this.playerTickingChunks;
-    }
-
-    @Override
-    public final void moonrise$markChunkForPlayerTicking(final LevelChunk chunk) {
-        final ChunkPos pos = chunk.getPos();
-        if (!this.playerTickingRequests.containsKey(ca.spottedleaf.moonrise.common.util.CoordinateUtils.getChunkKey(pos))) {
-            return;
-        }
-
-        this.playerTickingChunks.add(chunk);
-    }
-
-    @Override
-    public final void moonrise$removeChunkForPlayerTicking(final LevelChunk chunk) {
-        this.playerTickingChunks.remove(chunk);
-    }
-
-    @Override
-    public final void moonrise$addPlayerTickingRequest(final int chunkX, final int chunkZ) {
-        ca.spottedleaf.moonrise.common.util.TickThread.ensureTickThread((ServerLevel)(Object)this, chunkX, chunkZ, "Cannot add ticking request async");
-
-        final long chunkKey = ca.spottedleaf.moonrise.common.util.CoordinateUtils.getChunkKey(chunkX, chunkZ);
-
-        if (this.playerTickingRequests.addTo(chunkKey, 1) != 0) {
-            // already added
-            return;
-        }
-
-        final ca.spottedleaf.moonrise.patches.chunk_system.scheduling.NewChunkHolder chunkHolder = ((ca.spottedleaf.moonrise.patches.chunk_system.level.ChunkSystemServerLevel)(ServerLevel)(Object)this).moonrise$getChunkTaskScheduler()
-            .chunkHolderManager.getChunkHolder(chunkKey);
-
-        if (chunkHolder == null || !chunkHolder.isTickingReady()) {
-            return;
-        }
-
-        this.playerTickingChunks.add((LevelChunk)chunkHolder.getCurrentChunk());
-    }
-
-    @Override
-    public final void moonrise$removePlayerTickingRequest(final int chunkX, final int chunkZ) {
-        ca.spottedleaf.moonrise.common.util.TickThread.ensureTickThread((ServerLevel)(Object)this, chunkX, chunkZ, "Cannot remove ticking request async");
-
-        final long chunkKey = ca.spottedleaf.moonrise.common.util.CoordinateUtils.getChunkKey(chunkX, chunkZ);
-        final int val = this.playerTickingRequests.addTo(chunkKey, -1);
-
-        if (val <= 0) {
-            throw new IllegalStateException("Negative counter");
-        }
-
-        if (val != 1) {
-            // still has at least one request
-            return;
-        }
-
-        final ca.spottedleaf.moonrise.patches.chunk_system.scheduling.NewChunkHolder chunkHolder = ((ca.spottedleaf.moonrise.patches.chunk_system.level.ChunkSystemServerLevel)(ServerLevel)(Object)this).moonrise$getChunkTaskScheduler()
-            .chunkHolderManager.getChunkHolder(chunkKey);
-
-        if (chunkHolder == null || !chunkHolder.isTickingReady()) {
-            return;
-        }
-
-        this.playerTickingChunks.remove((LevelChunk)chunkHolder.getCurrentChunk());
-    }
+    // ShreddedPaper start - handled by the regions
+//    private final ca.spottedleaf.moonrise.common.list.ReferenceList<LevelChunk> playerTickingChunks = new ca.spottedleaf.moonrise.common.list.ReferenceList<>(EMPTY_LEVEL_CHUNKS);
+//    private final it.unimi.dsi.fastutil.longs.Long2IntOpenHashMap playerTickingRequests = new it.unimi.dsi.fastutil.longs.Long2IntOpenHashMap();
+//
+//    @Override
+//    public final ca.spottedleaf.moonrise.common.list.ReferenceList<LevelChunk> moonrise$getPlayerTickingChunks() {
+//        return this.playerTickingChunks;
+//    }
+//
+//    @Override
+//    public final void moonrise$markChunkForPlayerTicking(final LevelChunk chunk) {
+//        final ChunkPos pos = chunk.getPos();
+//        if (!this.playerTickingRequests.containsKey(ca.spottedleaf.moonrise.common.util.CoordinateUtils.getChunkKey(pos))) {
+//            return;
+//        }
+//
+//        this.playerTickingChunks.add(chunk);
+//    }
+//
+//    @Override
+//    public final void moonrise$removeChunkForPlayerTicking(final LevelChunk chunk) {
+//        this.playerTickingChunks.remove(chunk);
+//    }
+//
+//    @Override
+//    public final void moonrise$addPlayerTickingRequest(final int chunkX, final int chunkZ) {
+//        ca.spottedleaf.moonrise.common.util.TickThread.ensureTickThread((ServerLevel)(Object)this, chunkX, chunkZ, "Cannot add ticking request async");
+//
+//        final long chunkKey = ca.spottedleaf.moonrise.common.util.CoordinateUtils.getChunkKey(chunkX, chunkZ);
+//
+//        if (this.playerTickingRequests.addTo(chunkKey, 1) != 0) {
+//            // already added
+//            return;
+//        }
+//
+//        final ca.spottedleaf.moonrise.patches.chunk_system.scheduling.NewChunkHolder chunkHolder = ((ca.spottedleaf.moonrise.patches.chunk_system.level.ChunkSystemServerLevel)(ServerLevel)(Object)this).moonrise$getChunkTaskScheduler()
+//            .chunkHolderManager.getChunkHolder(chunkKey);
+//
+//        if (chunkHolder == null || !chunkHolder.isTickingReady()) {
+//            return;
+//        }
+//
+//        this.playerTickingChunks.add((LevelChunk)chunkHolder.getCurrentChunk());
+//    }
+//
+//    @Override
+//    public final void moonrise$removePlayerTickingRequest(final int chunkX, final int chunkZ) {
+//        ca.spottedleaf.moonrise.common.util.TickThread.ensureTickThread((ServerLevel)(Object)this, chunkX, chunkZ, "Cannot remove ticking request async");
+//
+//        final long chunkKey = ca.spottedleaf.moonrise.common.util.CoordinateUtils.getChunkKey(chunkX, chunkZ);
+//        final int val = this.playerTickingRequests.addTo(chunkKey, -1);
+//
+//        if (val <= 0) {
+//            throw new IllegalStateException("Negative counter");
+//        }
+//
+//        if (val != 1) {
+//            // still has at least one request
+//            return;
+//        }
+//
+//        final ca.spottedleaf.moonrise.patches.chunk_system.scheduling.NewChunkHolder chunkHolder = ((ca.spottedleaf.moonrise.patches.chunk_system.level.ChunkSystemServerLevel)(ServerLevel)(Object)this).moonrise$getChunkTaskScheduler()
+//            .chunkHolderManager.getChunkHolder(chunkKey);
+//
+//        if (chunkHolder == null || !chunkHolder.isTickingReady()) {
+//            return;
+//        }
+//
+//        this.playerTickingChunks.remove((LevelChunk)chunkHolder.getCurrentChunk());
+//    }
+    // ShreddedPaper end - handled by the regions
     // Paper end - chunk tick iteration
 
     public ServerLevel(
@@ -733,7 +_,7 @@
 
         this.sleepStatus = new SleepStatus();
         this.gameEventDispatcher = new GameEventDispatcher(this);
-        this.waypointManager = new ServerWaypointManager(this); // Paper - optimize ServerWaypointManager with locator bar disabled
+        this.waypointManager = ShreddedPaperConfiguration.get().optimizations.disableLocatorBar ? new NoOpServerWaypointManager(this) : new ServerWaypointManager(this); // Paper - optimize ServerWaypointManager with locator bar disabled // ShreddedPaper - Optimise ServerWaypointManager
         this.environmentAttributes = EnvironmentAttributeSystem.builder().addDefaultLayers(this).build();
         this.updateSkyBrightness();
         // Paper start - rewrite chunk system
@@ -752,6 +_,7 @@
         // Paper end - rewrite chunk system
         this.getCraftServer().addWorld(this.getWorld()); // CraftBukkit
         this.preciseTime = this.serverLevelData.getGameTime(); // Purpur - Configurable daylight cycle
+        this.levelThread = Executors.newSingleThreadExecutor(r -> new ShreddedPaperTickThread(r, "ServerLevel-%s".formatted(WorldUtil.getWorldName(this)))); // ShreddedPaper
     }
 
     // Paper start
@@ -794,9 +_,9 @@
         return previous;
     }
 
-    public void tick(final BooleanSupplier haveTime) {
+    public CompletableFuture<Void> tickAsync(final BooleanSupplier haveTime) {
         ProfilerFiller profiler = Profiler.get();
-        this.handlingTick = true;
+        this.handlingTickThreadLocal.set(true); // ShreddedPaper
         TickRateManager tickRateManager = this.tickRateManager();
         boolean runs = tickRateManager.runsNormally();
         if (runs) {
@@ -854,13 +_,13 @@
         }
 
         profiler.popPush("chunkSource");
-        this.getChunkSource().tick(haveTime, true);
+        return this.getChunkSource().tick(haveTime, true).thenRun(() -> { // ShreddedPaper - run async
         profiler.popPush("blockEvents");
         if (runs) {
-            this.runBlockEvents();
+            // this.runBlockEvents(); // ShreddedPaper - moved into each region
         }
 
-        this.handlingTick = false;
+        this.handlingTickThreadLocal.set(false); // ShreddedPaper
         profiler.pop();
         boolean isActive = !paperConfig().unsupportedSettings.disableWorldTickingWhenEmpty || this.chunkSource.hasActiveTickets(); // CraftBukkit - this prevents entity cleanup, other issues on servers with no players // Paper - restore this
         if (isActive) {
@@ -880,34 +_,36 @@
             }
 
             io.papermc.paper.entity.activation.ActivationRange.activateEntities(this); // Paper - EAR
-            this.entityTickList
-                .forEach(
-                    entity -> {
-                        if (!entity.isRemoved()) {
-                            if (!tickRateManager.isEntityFrozen(entity)) {
-                                profiler.push("checkDespawn");
-                                entity.checkDespawn();
-                                profiler.pop();
-                                if (true) { // Paper - rewrite chunk system
-                                    Entity vehicle = entity.getVehicle();
-                                    if (vehicle != null) {
-                                        if (!vehicle.isRemoved() && vehicle.hasPassenger(entity)) {
-                                            return;
-                                        }
-
-                                        entity.stopRiding();
-                                    }
-
-                                    profiler.push("tick");
-                                    this.guardEntityTick(this::tickNonPassenger, entity);
-                                    profiler.pop();
-                                }
-                            }
-                        }
-                    }
-                );
+            // ShreddedPaper start - moved to ShreddedPaperEntityTicker
+//            this.entityTickList
+//                .forEach(
+//                    entity -> {
+//                        if (!entity.isRemoved()) {
+//                            if (!tickRateManager.isEntityFrozen(entity)) {
+//                                profiler.push("checkDespawn");
+//                                entity.checkDespawn();
+//                                profiler.pop();
+//                                if (true) { // Paper - rewrite chunk system
+//                                    Entity vehicle = entity.getVehicle();
+//                                    if (vehicle != null) {
+//                                        if (!vehicle.isRemoved() && vehicle.hasPassenger(entity)) {
+//                                            return;
+//                                        }
+//
+//                                        entity.stopRiding();
+//                                    }
+//
+//                                    profiler.push("tick");
+//                                    this.guardEntityTick(this::tickNonPassenger, entity);
+//                                    profiler.pop();
+//                                }
+//                            }
+//                        }
+//                    }
+//                );
+            // ShreddedPaper end - moved to ShreddedPaperEntityTicker
             profiler.popPush("blockEntities");
-            this.tickBlockEntities();
+            // this.tickBlockEntities(); // ShreddedPaper - handled locally in the region
             profiler.pop();
         }
 
@@ -916,15 +_,16 @@
         profiler.pop();
         profiler.push("debugSynchronizers");
         if (this.debugSynchronizers.hasAnySubscriberFor(DebugSubscriptions.NEIGHBOR_UPDATES)) {
-            this.neighborUpdater
-                .setDebugListener(blockPos -> this.debugSynchronizers.broadcastEventToTracking(blockPos, DebugSubscriptions.NEIGHBOR_UPDATES, blockPos));
+            // this.neighborUpdater // ShreddedPaper - unneeded
+            //     .setDebugListener(blockPos -> this.debugSynchronizers.broadcastEventToTracking(blockPos, DebugSubscriptions.NEIGHBOR_UPDATES, blockPos)); // ShreddedPaper - unneeded
         } else {
-            this.neighborUpdater.setDebugListener(null);
+            // this.neighborUpdater.setDebugListener(null); // ShreddedPaper - unneeded
         }
 
         this.debugSynchronizers.tick(this.server.debugSubscribers());
         profiler.pop();
         this.environmentAttributes().invalidateTickCache();
+        }); // ShreddedPaper - run async
     }
 
     @Override
@@ -953,7 +_,7 @@
 
     private void wakeUpAllPlayers() {
         this.sleepStatus.removeAllSleepers();
-        this.players.stream().filter(LivingEntity::isSleeping).collect(Collectors.toList()).forEach(player -> player.stopSleepInBed(false, false));
+        this.players.stream().filter(LivingEntity::isSleeping).collect(Collectors.toList()).forEach(player -> ShreddedPaper.runSync(player, () -> player.stopSleepInBed(false, false))); // ShreddedPaper - run on player's thread
     }
 
     // Paper start - optimise random ticking
@@ -1177,7 +_,7 @@
     }
 
     public boolean isHandlingTick() {
-        return this.handlingTick;
+        return this.handlingTickThreadLocal.get();
     }
 
     public boolean canSleepThroughNights() {
@@ -1398,27 +_,29 @@
         this.emptyTime = 0;
     }
 
-    private void tickFluid(final BlockPos pos, final Fluid type) {
+    public void tickFluid(final BlockPos pos, final Fluid type) { // ShreddedPaper - make public
+        TickThread.ensureTickThread(this, pos, "Cannot tick fluid outside of tick thread"); // ShreddedPaper
         BlockState blockState = this.getBlockState(pos);
         FluidState fluidState = blockState.getFluidState();
         if (fluidState.is(type)) {
             fluidState.tick(this, pos, blockState);
         }
         // Paper start - rewrite chunk system
-        if ((++this.tickedBlocksOrFluids & 7L) != 0L) {
+        if (false && (++this.tickedBlocksOrFluids & 7L) != 0L) { // ShreddedPaper - no mid tick tasks
             ((ca.spottedleaf.moonrise.patches.chunk_system.server.ChunkSystemMinecraftServer)this.server).moonrise$executeMidTickTasks();
         }
         // Paper end - rewrite chunk system
 
     }
 
-    private void tickBlock(final BlockPos pos, final Block type) {
+    public void tickBlock(final BlockPos pos, final Block type) { // ShreddedPaper - make public
+        TickThread.ensureTickThread(this, pos, "Cannot tick block outside of tick thread"); // ShreddedPaper
         BlockState state = this.getBlockState(pos);
         if (state.is(type)) {
             state.tick(this, pos, this.random);
         }
         // Paper start - rewrite chunk system
-        if ((++this.tickedBlocksOrFluids & 7L) != 0L) {
+        if (false && (++this.tickedBlocksOrFluids & 7L) != 0L) { // ShreddedPaper - no mid tick tasks
             ((ca.spottedleaf.moonrise.patches.chunk_system.server.ChunkSystemMinecraftServer)this.server).moonrise$executeMidTickTasks();
         }
         // Paper end - rewrite chunk system
@@ -1473,7 +_,7 @@
     private void tickPassenger(final Entity vehicle, final Entity entity, final boolean isActive) { // Paper - EAR 2
         if (entity.isRemoved() || entity.getVehicle() != vehicle) {
             entity.stopRiding();
-        } else if (entity instanceof Player || this.entityTickList.contains(entity)) {
+        } else if (entity instanceof Player || entity.isTicking()) { // ShreddedPaper - use a different method to check if passenger is ticking
             entity.setOldPosAndRot();
             entity.tickCount++;
             entity.totalEntityAge++; // Paper - age-like counter for all entities
@@ -1697,8 +_,8 @@
         } else {
             if (entity instanceof net.minecraft.world.entity.item.ItemEntity item && item.getItem().isEmpty()) return false; // Paper - Prevent empty items from being added
             // Paper start - capture all item additions to the world
-            if (captureDrops != null && entity instanceof net.minecraft.world.entity.item.ItemEntity item) {
-                captureDrops.add(item);
+            if (captureDropsThreadLocal.get() != null && entity instanceof net.minecraft.world.entity.item.ItemEntity item) { // ShreddedPaper - use thread local
+                captureDropsThreadLocal.get().add(item); // ShreddedPaper - use thread local
                 return true;
             }
             // Paper end - capture all item additions to the world
@@ -1784,7 +_,9 @@
                 .callEvent();
         }
         // Paper end - Add BlockBreakProgressUpdateEvent
-        for (ServerPlayer player : this.server.getPlayerList().getPlayers()) {
+        ClientboundBlockDestructionPacket packet = new ClientboundBlockDestructionPacket(id, blockPos, progress); // ShreddedPaper - BroadcastPacketEvent - move packet creation out of loop
+        new io.multipaper.shreddedpaper.event.BroadcastPacketEvent(this.getWorld(), packet, entity == null ? null : entity.getBukkitEntity()).callEvent(); // ShreddedPaper - BroadcastPacketEvent
+        for (ServerPlayer player : this.players) { // ShreddedPaper - BroadcastPacketEvent - optimization: use the level's players instead of all players
             if (player.level() == this && player.getId() != id) {
                 double xd = blockPos.getX() - player.getX();
                 double yd = blockPos.getY() - player.getY();
@@ -1795,7 +_,7 @@
                         continue;
                     }
                     // CraftBukkit end
-                    player.connection.send(new ClientboundBlockDestructionPacket(id, blockPos, progress));
+                    player.connection.send(packet); // ShreddedPaper - BroadcastPacketEvent - move packet creation out of loop
                 }
             }
         }
@@ -1904,7 +_,7 @@
 
     @Override
     public void sendBlockUpdated(final BlockPos pos, final BlockState old, final BlockState current, final int updateFlags) {
-        if (this.isUpdatingNavigations) {
+        if (this.isUpdatingNavigations.get()) { // ShreddedPaper - use thread local
             String message = "recursive call to sendBlockUpdated";
             Util.logAndPauseIfInIde("recursive call to sendBlockUpdated", new IllegalStateException("recursive call to sendBlockUpdated"));
         }
@@ -1918,7 +_,7 @@
             List<PathNavigation> navigationsToUpdate = new ObjectArrayList<>();
 
             try { // Paper - catch CME see below why
-            for (Mob navigatingMob : this.navigatingMobs) {
+            for (Mob navigatingMob : this.chunkSource.tickingRegions.collectRelevantNavigatingMobs(RegionPos.forChunk(ChunkPos.containing(pos)))) { // ShreddedPaper
                 PathNavigation pathNavigation = navigatingMob.getNavigation();
                 if (pathNavigation.shouldRecomputePath(pos)) {
                     navigationsToUpdate.add(pathNavigation);
@@ -1935,13 +_,13 @@
             // Paper end - catch CME see below why
 
             try {
-                this.isUpdatingNavigations = true;
+                this.isUpdatingNavigations.set(true); // ShreddedPaper - use thread local
 
                 for (PathNavigation navigation : navigationsToUpdate) {
                     navigation.recomputePath();
                 }
             } finally {
-                this.isUpdatingNavigations = false;
+                this.isUpdatingNavigations.set(false); // ShreddedPaper - use thread local
             }
         }
         } // Paper - option to disable pathfinding updates
@@ -1950,17 +_,17 @@
     @Override
     public void updateNeighborsAt(final BlockPos pos, final Block sourceBlock) {
         // CraftBukkit start
-        if (this.populating) {
+        if (this.populatingThreadLocal.get()) { // ShreddedPaper - thread local for populating
             return;
         }
         // CraftBukkit end
-        if (captureBlockStates) { return; } // Paper - Cancel all physics during placement
+        if (captureBlockStatesThreadLocal.get()) { return; } // Paper - Cancel all physics during placement // ShreddedPaper - use thread local
         this.updateNeighborsAt(pos, sourceBlock, ExperimentalRedstoneUtils.initialOrientation(this, null, null));
     }
 
     @Override
     public void updateNeighborsAt(final BlockPos pos, final Block sourceBlock, final @Nullable Orientation orientation) {
-        if (captureBlockStates) { return; } // Paper - Cancel all physics during placement
+        if (captureBlockStatesThreadLocal.get()) { return; } // Paper - Cancel all physics during placement // ShreddedPaper - use thread local
         this.neighborUpdater.updateNeighborsAtExceptFromFacing(pos, sourceBlock, null, orientation);
     }
 
@@ -2089,14 +_,15 @@
 
     @Override
     public void blockEvent(final BlockPos pos, final Block block, final int b0, final int b1) {
-        this.blockEvents.add(new BlockEventData(pos, block, b0, b1));
+        this.chunkSource.tickingRegions.addBlockEvent(new BlockEventData(pos, block, b0, b1)); // ShreddedPaper
     }
 
-    private void runBlockEvents() {
-        this.blockEventsToReschedule.clear();
+    public void runBlockEvents(final LevelChunkRegion region) { // ShreddedPaper
+        List<BlockEventData> blockEventsToReschedule = blockEventsToRescheduleThreadLocal.get(); // ShreddedPaper
+        blockEventsToReschedule.clear(); // ShreddedPaper
 
-        while (!this.blockEvents.isEmpty()) {
-            BlockEventData eventData = this.blockEvents.removeFirst();
+        while (region.hasBlockEvents()) { // ShreddedPaper
+            BlockEventData eventData = region.removeFirstBlockEvent(); // ShreddedPaper
             if (this.shouldTickBlocksAt(eventData.pos())) {
                 if (this.doBlockEvent(eventData)) {
                     this.server
@@ -2112,11 +_,11 @@
                         );
                 }
             } else {
-                this.blockEventsToReschedule.add(eventData);
+                blockEventsToReschedule.add(eventData); // ShreddedPaper
             }
         }
 
-        this.blockEvents.addAll(this.blockEventsToReschedule);
+        region.addAllBlockEvents(blockEventsToReschedule); // ShreddedPaper
     }
 
     private boolean doBlockEvent(final BlockEventData eventData) {
@@ -2370,7 +_,7 @@
             final MapItemSavedData mapData = storage.get(MapItemSavedData.type(id)); // get populates the cache
             if (mapData != null) { // map was read, init it and return
                 mapData.id = id;
-                new org.bukkit.event.server.MapInitializeEvent(mapData.mapView).callEvent();
+                if (mapData.hasCalledMapInitializeEvent.compareAndSet(false, true)) new org.bukkit.event.server.MapInitializeEvent(mapData.mapView).callEvent(); // ShreddedPaper - ensure only called once
                 return mapData;
             }
 
@@ -2389,8 +_,7 @@
     public void setMapData(final MapId id, final MapItemSavedData data) {
         // CraftBukkit start
         data.id = id;
-        org.bukkit.event.server.MapInitializeEvent event = new org.bukkit.event.server.MapInitializeEvent(data.mapView);
-        event.callEvent();
+        if (data.hasCalledMapInitializeEvent.compareAndSet(false, true)) new org.bukkit.event.server.MapInitializeEvent(data.mapView).callEvent(); // ShreddedPaper - ensure only called once
         // CraftBukkit end
         this.getServer().getDataStorage().set(MapItemSavedData.type(id), data);
     }
@@ -2446,11 +_,11 @@
         Optional<Holder<PoiType>> newType = PoiTypes.forState(newState);
         if (!Objects.equals(oldType, newType)) {
             BlockPos immutable = pos.immutable();
-            oldType.ifPresent(poiType -> this.getServer().execute(() -> {
+            oldType.ifPresent(poiType -> chunkSource.tickingRegions.execute(RegionPos.forChunk(ChunkPos.containing(pos)), () -> { // ShreddedPaper
                 this.getPoiManager().remove(immutable);
                 this.debugSynchronizers.dropPoi(immutable);
             }));
-            newType.ifPresent(poiType -> this.getServer().execute(() -> {
+            newType.ifPresent(poiType -> chunkSource.tickingRegions.execute(RegionPos.forChunk(ChunkPos.containing(pos)), () -> { // ShreddedPaper
                 // Paper start - Remove stale POIs
                 if (oldType.isEmpty() && this.getPoiManager().exists(pos, _ -> true)) {
                     this.getPoiManager().remove(pos);
@@ -2513,7 +_,7 @@
             }
 
             output.write(String.format(Locale.ROOT, "entities: %s\n", this.moonrise$getEntityLookup().getDebugInfo())); // Paper - rewrite chunk system
-            output.write(String.format(Locale.ROOT, "block_entity_tickers: %d\n", this.blockEntityTickers.size()));
+            // output.write(String.format(Locale.ROOT, "block_entity_tickers: %d\n", this.blockEntityTickers.size())); // ShreddedPaper - removed blockEntityTickers
             output.write(String.format(Locale.ROOT, "block_ticks: %d\n", this.getBlockTicks().count()));
             output.write(String.format(Locale.ROOT, "fluid_ticks: %d\n", this.getFluidTicks().count()));
             output.write("distance_manager: " + chunkMap.getDistanceManager().getDebugStatus() + "\n");
@@ -2581,17 +_,25 @@
     }
 
     private void dumpBlockEntityTickers(final Writer output) throws IOException {
-        CsvOutput csvOutput = CsvOutput.builder().addColumn("x").addColumn("y").addColumn("z").addColumn("type").build(output);
-
-        for (TickingBlockEntity ticker : this.blockEntityTickers) {
-            BlockPos blockPos = ticker.getPos();
-            csvOutput.writeRow(blockPos.getX(), blockPos.getY(), blockPos.getZ(), ticker.getType());
-        }
+        // ShreddedPaper start - removed blockEntityTickers
+//        CsvOutput csvOutput = CsvOutput.builder().addColumn("x").addColumn("y").addColumn("z").addColumn("type").build(output);
+//
+//        for (TickingBlockEntity ticker : this.blockEntityTickers) {
+//            BlockPos blockPos = ticker.getPos();
+//            csvOutput.writeRow(blockPos.getX(), blockPos.getY(), blockPos.getZ(), ticker.getType());
+//        }
+        // ShreddedPaper end - removed blockEntityTickers
     }
 
     @VisibleForTesting
     public void clearBlockEvents(final BoundingBox bb) {
-        this.blockEvents.removeIf(e -> bb.isInside(e.pos()));
+        // ShreddedPaper start - moved blockEvents into regions
+        this.getLevel().getChunkSource().tickingRegions.forEachRegionInBoundingBox(bb, region -> {
+            region.removeBlockEventsIf((blockactiondata) -> {
+                return bb.isInside(blockactiondata.pos());
+            });
+        });
+        // ShreddedPaper end - moved blockEvents into regions
     }
 
     public Iterable<Entity> getAllEntities() {
@@ -2633,8 +_,8 @@
             this.players.size(),
             this.moonrise$getEntityLookup().getDebugInfo(), // Paper - rewrite chunk system
             getTypeCount(this.moonrise$getEntityLookup().getAll(), e -> e.typeHolder().getRegisteredName()), // Paper - rewrite chunk system
-            this.blockEntityTickers.size(),
-            getTypeCount(this.blockEntityTickers, TickingBlockEntity::getType),
+            0, // this.blockEntityTickers.size(),, // ShreddedPaper - removed blockEntityTickers
+            0, // getTypeCount(this.blockEntityTickers, TickingBlockEntity::getType),, // ShreddedPaper - removed blockEntityTickers
             this.getBlockTicks().count(),
             this.getFluidTicks().count(),
             this.gatherChunkSourceStats()
@@ -2664,7 +_,7 @@
 
     @Override
     public LevelEntityGetter<Entity> getEntities() {
-        org.spigotmc.AsyncCatcher.catchOp("Chunk getEntities call"); // Spigot
+        // org.spigotmc.AsyncCatcher.catchOp("Chunk getEntities call"); // Spigot // ShreddedPaper - thread safe now
         return this.moonrise$getEntityLookup(); // Paper - rewrite chunk system
     }
 
@@ -2854,7 +_,7 @@
     // Paper start - optimize redstone (Alternate Current)
     @Override
     public alternate.current.wire.WireHandler getWireHandler() {
-        return wireHandler;
+        return wireHandlerThreadLocal.get(); // ShreddedPaper - ThreadLocal
     }
     // Paper end - optimize redstone (Alternate Current)
 
@@ -2900,12 +_,12 @@
         @Override
         public void onTickingStart(final Entity entity) {
             if (entity instanceof net.minecraft.world.entity.Marker && !paperConfig().entities.markers.tick) return; // Paper - Configurable marker ticking
-            ServerLevel.this.entityTickList.add(entity);
+            ServerLevel.this.getChunkSource().tickingRegions.addTickingEntity(entity); // ShreddedPaper - move entityTickList to the regions
         }
 
         @Override
         public void onTickingEnd(final Entity entity) {
-            ServerLevel.this.entityTickList.remove(entity);
+            ServerLevel.this.getChunkSource().tickingRegions.removeTickingEntity(entity); // ShreddedPaper - move entityTickList to the regions
             // Paper start - Reset pearls when they stop being ticked
             if (ServerLevel.this.paperConfig().fixes.disableUnloadedChunkEnderpearlExploit && ServerLevel.this.paperConfig().misc.legacyEnderPearlBehavior && entity instanceof net.minecraft.world.entity.projectile.throwableitemprojectile.ThrownEnderpearl pearl) {
                 pearl.setOwner(null);
@@ -2915,10 +_,11 @@
 
         @Override
         public void onTrackingStart(final Entity entity) {
-            org.spigotmc.AsyncCatcher.catchOp("entity register"); // Spigot
+            TickThread.ensureTickThread(entity, "entity register"); // org.spigotmc.AsyncCatcher.catchOp("entity register"); // Spigot // ShreddedPaper
             // ServerLevel.this.getChunkSource().addEntity(entity); // Paper - ignore and warn about illegal addEntity calls instead of crashing server; moved down below valid=true
             if (entity instanceof ServerPlayer player) {
                 ServerLevel.this.players.add(player);
+                ServerLevel.this.getChunkSource().tickingRegions.addPlayer(player); // ShreddedPaper - add player to region
                 if (player.isReceivingWaypoints()) {
                     ServerLevel.this.getWaypointManager().addPlayer(player);
                 }
@@ -2931,14 +_,14 @@
             }
 
             if (entity instanceof Mob mob) {
-                if (false && ServerLevel.this.isUpdatingNavigations) { // Paper - Remove unnecessary onTrackingStart during navigation warning
+                if (false && ServerLevel.this.isUpdatingNavigations.get()) { // Paper - Remove unnecessary onTrackingStart during navigation warning // ShreddedPaper - use thread local
                     String message = "onTrackingStart called during navigation iteration";
                     Util.logAndPauseIfInIde(
                         "onTrackingStart called during navigation iteration", new IllegalStateException("onTrackingStart called during navigation iteration")
                     );
                 }
 
-                ServerLevel.this.navigatingMobs.add(mob);
+                // ServerLevel.this.navigatingMobs.add(mob); // ShreddedPaper - handled ourselves
             }
 
             if (entity instanceof EnderDragon dragon) {
@@ -2965,7 +_,7 @@
 
         @Override
         public void onTrackingEnd(final Entity entity) {
-            org.spigotmc.AsyncCatcher.catchOp("entity unregister"); // Spigot
+            TickThread.ensureTickThread(entity, "entity unregister"); // org.spigotmc.AsyncCatcher.catchOp("entity unregister"); // Spigot // ShreddedPaper
             // Spigot start // TODO I don't think this is needed anymore
             if (entity instanceof Player player) {
                 for (final ServerLevel level : ServerLevel.this.getServer().getAllLevels()) {
@@ -2974,10 +_,12 @@
                             continue;
                         }
 
+                        synchronized (map) { // ShreddedPaper - thread-safe
                         map.carriedByPlayers.remove(player);
                         if (map.carriedBy.removeIf(holdingPlayer -> holdingPlayer.player == player)) {
                             map.decorations.remove(player.getName().getString());
                         }
+                        } // ShreddedPaper - thread-safe
                     }
                 }
             }
@@ -2996,20 +_,21 @@
             // Spigot end
             ServerLevel.this.getChunkSource().removeEntity(entity);
             if (entity instanceof ServerPlayer player) {
+                ServerLevel.this.getChunkSource().tickingRegions.removePlayer(player); // ShreddedPaper - remove player from region
                 ServerLevel.this.players.remove(player);
                 ServerLevel.this.getWaypointManager().removePlayer(player);
                 ServerLevel.this.updateSleepingPlayerList();
             }
 
             if (entity instanceof Mob mob) {
-                if (false && ServerLevel.this.isUpdatingNavigations) { // Paper - Remove unnecessary onTrackingStart during navigation warning
+                if (false && ServerLevel.this.isUpdatingNavigations.get()) { // Paper - Remove unnecessary onTrackingStart during navigation warning // ShreddedPaper - use thread local
                     String message = "onTrackingStart called during navigation iteration";
                     Util.logAndPauseIfInIde(
                         "onTrackingStart called during navigation iteration", new IllegalStateException("onTrackingStart called during navigation iteration")
                     );
                 }
 
-                ServerLevel.this.navigatingMobs.remove(mob);
+                // ServerLevel.this.navigatingMobs.remove(mob); // ShreddedPaper - handled ourselves
             }
 
             if (entity instanceof EnderDragon dragon) {
@@ -3034,6 +_,9 @@
         @Override
         public void onSectionChange(final Entity entity) {
             entity.updateDynamicGameEventListener(DynamicGameEventListener::move);
+            ServerLevel.this.chunkSource.tickingRegions.moveTickingEntity(entity); // ShreddedPaper - move ticking entity into the new region
+            ServerLevel.this.chunkSource.tickingRegions.moveTrackedEntity(entity); // ShreddedPaper - move tracked entity into the new region
+            if (entity instanceof ServerPlayer serverPlayer) ServerLevel.this.chunkSource.tickingRegions.movePlayer(serverPlayer); // ShreddedPaper - move player into the new region
         }
     }
 
