--- a/net/minecraft/world/level/storage/PlayerDataStorage.java
+++ b/net/minecraft/world/level/storage/PlayerDataStorage.java
@@ -1,5 +_,6 @@
 package net.minecraft.world.level.storage;
 
+import ca.spottedleaf.moonrise.common.util.TickThread;
 import com.mojang.datafixers.DataFixer;
 import com.mojang.logging.LogUtils;
 import java.io.File;
@@ -8,6 +_,9 @@
 import java.nio.file.StandardCopyOption;
 import java.time.ZonedDateTime;
 import java.util.Optional;
+import java.util.concurrent.Executor;
+
+import io.multipaper.shreddedpaper.config.ShreddedPaperConfiguration;
 import net.minecraft.nbt.CompoundTag;
 import net.minecraft.nbt.NbtAccounter;
 import net.minecraft.nbt.NbtIo;
@@ -24,6 +_,8 @@
     private final File playerDir;
     protected final DataFixer fixerUpper;
 
+    private static final Executor IO_WORKER = Util.ioPool(); // ShreddedPaper
+
     public PlayerDataStorage(LevelStorageSource.LevelStorageAccess levelStorageAccess, DataFixer fixerUpper) {
         this.fixerUpper = fixerUpper;
         this.playerDir = levelStorageAccess.getLevelPath(LevelResource.PLAYER_DATA_DIR).toFile();
@@ -35,6 +_,7 @@
         try (ProblemReporter.ScopedCollector scopedCollector = new ProblemReporter.ScopedCollector(player.problemPath(), LOGGER)) {
             TagValueOutput tagValueOutput = TagValueOutput.createWithContext(scopedCollector, player.registryAccess());
             player.saveWithoutId(tagValueOutput);
+            Runnable r = () -> { try { // ShreddedPaper - write async
             Path path = this.playerDir.toPath();
             Path path1 = Files.createTempFile(path, player.getStringUUID() + "-", ".dat");
             CompoundTag compoundTag = tagValueOutput.buildResult();
@@ -42,6 +_,8 @@
             Path path2 = path.resolve(player.getStringUUID() + ".dat");
             Path path3 = path.resolve(player.getStringUUID() + ".dat_old");
             Util.safeReplaceFile(path2, path1, path3);
+            } catch (Exception e) { LOGGER.warn("Failed to save player data for {}", player.getPlainTextName(), e); throw new RuntimeException(e); }}; // ShreddedPaper - write async
+            if (ShreddedPaperConfiguration.get().optimizations.writePlayerSavesAsync && !TickThread.isShutdownThread()) IO_WORKER.execute(r); else r.run(); // ShreddedPaper - write async
         } catch (Exception var11) {
             LOGGER.warn("Failed to save player data for {}", player.getPlainTextName(), var11); // Paper - Print exception
         }
