--- a/purpur-api/build.gradle.kts
+++ b/purpur-api/build.gradle.kts
@@ -107,7 +_,7 @@
     testRuntimeOnly("org.junit.platform:junit-platform-launcher:6.0.3")
 
     // checkstyle
-    checkstyle(project(":purpur-checkstyle"))
+    // checkstyle(project(":purpur-checkstyle")) // ShreddedPaper - unused
 }
 
 val generatedDir: java.nio.file.Path = rootProject.layout.projectDirectory.dir("paper-api/src/generated/java").asFile.toPath() // Purpur
@@ -121,17 +_,21 @@
         java {
             srcDir(generatedDir)
             srcDir(file("../paper-api/src/main/java"))
+            srcDir(file("../purpur-api/src/main/java"))
         }
         resources {
             srcDir(file("../paper-api/src/main/resources"))
+            srcDir(file("../purpur-api/src/main/resources"))
         }
     }
     test {
         java {
             srcDir(file("../paper-api/src/test/java"))
+            srcDir(file("../purpur-api/src/test/java"))
         }
         resources {
             srcDir(file("../paper-api/src/test/resources"))
+            srcDir(file("../purpur-api/src/test/resources"))
         }
     }
 }
