--- a/net/minecraft/world/entity/npc/villager/Villager.java
+++ b/net/minecraft/world/entity/npc/villager/Villager.java
@@ -5,6 +_,7 @@
 import com.google.common.collect.ImmutableSet;
 import com.mojang.datafixers.util.Pair;
 import com.mojang.logging.LogUtils;
+import io.multipaper.shreddedpaper.ShreddedPaper;
 import java.util.ArrayList;
 import java.util.List;
 import java.util.Map;
@@ -811,13 +_,16 @@
             this.brain.getMemory(memoryType).ifPresent(memory -> {
                 ServerLevel poiLevel = server.getLevel(memory.dimension());
                 if (poiLevel != null) {
+                    BiPredicate<Villager, Holder<PoiType>> poiTypePredicate = POI_MEMORIES.get(memoryType); // ShreddedPaper - don't run on poi's region
+                    ShreddedPaper.ensureSync(poiLevel, memory.pos(), () -> { // ShreddedPaper - run on poi's region
                     PoiManager poiManager = poiLevel.getPoiManager();
                     Optional<Holder<PoiType>> type = poiManager.getType(memory.pos());
-                    BiPredicate<Villager, Holder<PoiType>> poiTypePredicate = POI_MEMORIES.get(memoryType);
+                    // BiPredicate<Villager, Holder<PoiType>> poiTypePredicate = POI_MEMORIES.get(memoryType); // ShreddedPaper - move up
                     if (type.isPresent() && poiTypePredicate.test(this, type.get())) {
                         poiManager.release(memory.pos());
                         poiLevel.debugSynchronizers().updatePoi(memory.pos());
                     }
+                    }); // ShreddedPaper - run on poi's region
                 }
             });
         }
