--- a/net/minecraft/server/commands/ClearInventoryCommands.java
+++ b/net/minecraft/server/commands/ClearInventoryCommands.java
@@ -6,7 +_,10 @@
 import com.mojang.brigadier.exceptions.DynamicCommandExceptionType;
 import java.util.Collection;
 import java.util.Collections;
+import java.util.concurrent.CompletableFuture;
 import java.util.function.Predicate;
+
+import io.multipaper.shreddedpaper.ShreddedPaper;
 import net.minecraft.commands.CommandBuildContext;
 import net.minecraft.commands.CommandSourceStack;
 import net.minecraft.commands.Commands;
@@ -73,9 +_,16 @@
         int i = 0;
 
         for (ServerPlayer serverPlayer : targetPlayers) {
-            i += serverPlayer.getInventory().clearOrCountMatchingItems(itemPredicate, maxCount, serverPlayer.inventoryMenu.getCraftSlots());
-            serverPlayer.containerMenu.broadcastChanges();
-            serverPlayer.inventoryMenu.slotsChanged(serverPlayer.getInventory());
+            // ShreddedPaper start - run on correct thread
+            CompletableFuture<Integer> future = new CompletableFuture<>();
+            ShreddedPaper.ensureSync(serverPlayer, () -> {
+                int j = serverPlayer.getInventory().clearOrCountMatchingItems(itemPredicate, maxCount, serverPlayer.inventoryMenu.getCraftSlots());
+                serverPlayer.containerMenu.broadcastChanges();
+                serverPlayer.inventoryMenu.slotsChanged(serverPlayer.getInventory());
+                future.complete(j);
+            });
+            i += future.getNow(1); // ShreddedPaper - default to 1 if player in on a different thread
+            // ShreddedPaper end - run on correct thread
         }
 
         if (i == 0) {
