--- a/net/minecraft/server/commands/CloneCommands.java
+++ b/net/minecraft/server/commands/CloneCommands.java
@@ -8,9 +_,12 @@
 import com.mojang.brigadier.exceptions.Dynamic2CommandExceptionType;
 import com.mojang.brigadier.exceptions.SimpleCommandExceptionType;
 import com.mojang.logging.LogUtils;
+import io.multipaper.shreddedpaper.ShreddedPaper;
 import java.util.Deque;
 import java.util.List;
 import java.util.function.Predicate;
+
+import io.multipaper.shreddedpaper.util.ObjectHolder;
 import net.minecraft.commands.CommandBuildContext;
 import net.minecraft.commands.CommandSourceStack;
 import net.minecraft.commands.Commands;
@@ -233,6 +_,7 @@
             } else if (serverLevel1.isDebug()) {
                 throw ERROR_FAILED.create();
             } else {
+                ShreddedPaper.ensureSync(serverLevel1, blockPos2, () -> { // ShreddedPaper - run on correct region thread
                 List<CloneCommands.CloneBlockInfo> list = Lists.newArrayList();
                 List<CloneCommands.CloneBlockInfo> list1 = Lists.newArrayList();
                 List<CloneCommands.CloneBlockInfo> list2 = Lists.newArrayList();
@@ -336,17 +_,19 @@
                         var35.addSuppressed(var34);
                     }
 
-                    throw var35;
+                    throw new RuntimeException(var35); // ShreddedPaper - run on correct region thread - wrap in RuntimeException for lambda
                 }
 
                 scopedCollector.close();
                 if (i2 == 0) {
-                    throw ERROR_FAILED.create();
+                    source.sendFailure(Component.translatable("commands.clone.failed")); // ShreddedPaper - run on correct region thread - send failure message instead of throwing
                 } else {
                     int i3 = i2;
                     source.sendSuccess(() -> Component.translatable("commands.clone.success", i3), true);
-                    return i2;
+                    // return i2; // ShreddedPaper - run on correct region thread
                 }
+                }); // ShreddedPaper - run on correct region thread
+                return 1; // ShreddedPaper - run on correct region thread
             }
         }
     }
