--- a/net/minecraft/server/commands/SpreadPlayersCommand.java
+++ b/net/minecraft/server/commands/SpreadPlayersCommand.java
@@ -246,17 +_,12 @@
                 position = positions[i++];
             }
 
-            entity.teleportTo(
-                level,
-                Mth.floor(position.x) + 0.5,
-                position.getSpawnY(level, maxHeight),
-                Mth.floor(position.z) + 0.5,
-                Set.of(),
-                entity.getYRot(),
-                entity.getXRot(),
-                true
-                , org.bukkit.event.player.PlayerTeleportEvent.TeleportCause.COMMAND // CraftBukkit - handle teleport reason
+            // ShreddedPaper start - use teleportAsync to handle cross-region teleportation
+            entity.getBukkitEntity().teleportAsync(
+                new org.bukkit.Location(level.getWorld(), Mth.floor(position.x) + 0.5, position.getSpawnY(level, maxHeight), Mth.floor(position.z) + 0.5, entity.getYRot(), entity.getXRot()),
+                org.bukkit.event.player.PlayerTeleportEvent.TeleportCause.COMMAND
             );
+            // ShreddedPaper end - use teleportAsync to handle cross-region teleportation
             double d1 = Double.MAX_VALUE;
 
             for (SpreadPlayersCommand.Position position1 : positions) {
