--- a/net/minecraft/server/commands/TeleportCommand.java
+++ b/net/minecraft/server/commands/TeleportCommand.java
@@ -9,6 +_,8 @@
 import java.util.EnumSet;
 import java.util.Locale;
 import java.util.Set;
+
+import io.multipaper.shreddedpaper.ShreddedPaper;
 import net.minecraft.commands.CommandSourceStack;
 import net.minecraft.commands.Commands;
 import net.minecraft.commands.arguments.EntityAnchorArgument;
@@ -24,6 +_,7 @@
 import net.minecraft.world.entity.LivingEntity;
 import net.minecraft.world.entity.PathfinderMob;
 import net.minecraft.world.entity.Relative;
+import net.minecraft.world.level.ChunkPos;
 import net.minecraft.world.level.Level;
 import net.minecraft.world.phys.Vec2;
 import net.minecraft.world.phys.Vec3;
@@ -255,6 +_,7 @@
         float relativeOrAbsoluteYRot = relatives.contains(Relative.Y_ROT) ? yRot - victim.getYRot() : yRot;
         float relativeOrAbsoluteXRot = relatives.contains(Relative.X_ROT) ? xRot - victim.getXRot() : xRot;
         float newYRot = Mth.wrapDegrees(relativeOrAbsoluteYRot);
+        ShreddedPaper.ensureSync(victim, level, ChunkPos.containing(new BlockPos((int) Math.floor(x), (int) Math.floor(y), (int) Math.floor(z))), () -> { // ShreddedPaper - run on correct thread
         float newXRot = Mth.wrapDegrees(relativeOrAbsoluteXRot);
         if (victim.teleportTo(level, relativeOrAbsoluteX, relativeOrAbsoluteY, relativeOrAbsoluteZ, relatives, newYRot, newXRot, true, org.bukkit.event.player.PlayerTeleportEvent.TeleportCause.COMMAND)) { // Paper - teleport cause
             if (lookAt != null) {
@@ -270,5 +_,6 @@
                 mob.getNavigation().stop();
             }
         }
+        }); // ShreddedPaper - run on correct thread
     }
 }
