--- a/net/minecraft/world/level/dimension/end/EndDragonFight.java
+++ b/net/minecraft/world/level/dimension/end/EndDragonFight.java
@@ -9,6 +_,7 @@
 import com.mojang.logging.LogUtils;
 import com.mojang.serialization.Codec;
 import com.mojang.serialization.codecs.RecordCodecBuilder;
+import io.multipaper.shreddedpaper.region.RegionPos;
 import it.unimi.dsi.fastutil.objects.ObjectArrayList;
 import java.util.List;
 import java.util.Optional;
@@ -157,6 +_,7 @@
         }
 
         if (!this.dragonEvent.getPlayers().isEmpty()) {
+            final Runnable r = () -> { // ShreddedPaper - run on end island thread
             this.level.getChunkSource().addTicketWithRadius(TicketType.DRAGON, new ChunkPos(0, 0), 9);
             boolean isArenaLoaded = this.isArenaLoaded();
             if (this.needsStateScanning && isArenaLoaded) {
@@ -184,6 +_,13 @@
                     this.ticksSinceCrystalsScanned = 0;
                 }
             }
+            // ShreddedPaper start - run on end island thread
+            };
+            while (!this.level.chunkScheduler.getRegionLocker().tryLockNow(RegionPos.forChunk(0, 0), r)) {
+                long startBlockingTime = System.nanoTime();
+                this.level.chunkSource.mainThreadProcessor.managedBlock(() -> System.nanoTime() - startBlockingTime < 1_000_000); // Wait for 1ms
+            }
+            // ShreddedPaper end - run on end island thread
         } else {
             this.level.getChunkSource().removeTicketWithRadius(TicketType.DRAGON, new ChunkPos(0, 0), 9);
         }
