--- a/src/main/java/org/bukkit/craftbukkit/util/CraftLocation.java
+++ b/src/main/java/org/bukkit/craftbukkit/util/CraftLocation.java
@@ -1,7 +_,9 @@
 package org.bukkit.craftbukkit.util;
 
 import net.minecraft.core.BlockPos;
+import net.minecraft.core.SectionPos;
 import net.minecraft.core.Vec3i;
+import net.minecraft.world.level.ChunkPos;
 import net.minecraft.world.level.Level;
 import net.minecraft.world.level.pathfinder.Node;
 import net.minecraft.world.phys.Vec3;
@@ -68,6 +_,12 @@
     public static Location fromGlobalPos(net.minecraft.core.GlobalPos globalPos) {
         return CraftLocation.toBukkit(globalPos.pos(), net.minecraft.server.MinecraftServer.getServer().getLevel(globalPos.dimension()));
     }
+
+    // ShreddedPaper start
+    public static ChunkPos toChunkPosition(Location location) {
+        return new ChunkPos(SectionPos.blockToSectionCoord(location.getBlockX()), SectionPos.blockToSectionCoord(location.getBlockZ()));
+    }
+    // ShreddedPaper end
 
     public static Vec3 toVec3(Location loc) {
         return new Vec3(loc.getX(), loc.getY(), loc.getZ());
