--- a/net/minecraft/server/network/ServerGamePacketListenerImpl.java
+++ b/net/minecraft/server/network/ServerGamePacketListenerImpl.java
@@ -1,5 +_,6 @@
 package net.minecraft.server.network;
 
+import ca.spottedleaf.moonrise.common.util.TickThread;
 import com.google.common.collect.Lists;
 import com.google.common.primitives.Floats;
 import com.mojang.authlib.GameProfile;
@@ -8,6 +_,7 @@
 import com.mojang.brigadier.StringReader;
 import com.mojang.brigadier.suggestion.Suggestions;
 import com.mojang.logging.LogUtils;
+import io.multipaper.shreddedpaper.ShreddedPaper;
 import it.unimi.dsi.fastutil.ints.Int2ObjectMaps;
 import it.unimi.dsi.fastutil.ints.Int2ObjectMap.Entry;
 import it.unimi.dsi.fastutil.objects.Object2ObjectOpenHashMap;
@@ -378,7 +_,7 @@
         }
 
         if (this.server.isPaused() || !this.tickPlayer()) {
-            this.keepConnectionAlive();
+            // this.keepConnectionAlive(); // ShreddedPaper - called elsewhere
             this.chatSpamThrottler.tick();
             this.commandSpamThrottler.tick();
             this.dropSpamThrottler.tick();
@@ -637,6 +_,12 @@
                     return;
                 }
                 // Paper end - Prevent moving into unloaded chunks
+                // ShreddedPaper start - Check we own the destination location
+                if (!TickThread.isTickThreadFor(this.player.level(), new Vec3(targetX, targetY, targetZ))) {
+                    this.connection.send(ClientboundMoveVehiclePacket.fromEntity(vehicle));
+                    return;
+                }
+                // ShreddedPaper end - Check we own the destination location
                 if (movedDist - expectedDist > Math.max(100.0, Mth.square(org.spigotmc.SpigotConfig.movedTooQuicklyMultiplier * (float) i * speed)) && !this.isSingleplayerOwner()) {
                     // CraftBukkit end
                     LOGGER.warn(
@@ -1423,7 +_,7 @@
             Consumer<List<FilteredText>> handler = title.isPresent()
                 ? filteredContents -> this.signBook(filteredContents.get(0), filteredContents.subList(1, filteredContents.size()), slot, hasSignPerm) // Purpur - Allow color codes in books
                 : filteredContents -> this.updateBookContents(filteredContents, slot, hasEditPerm); // Purpur - Allow color codes in books
-            this.filterTextPacket(contents).thenAcceptAsync(handler, this.server);
+            this.filterTextPacket(contents).thenAcceptAsync(handler, r -> ShreddedPaper.ensureSync(this.player, r)); // ShreddedPaper - run on region thread
         }
     }
 
@@ -1616,6 +_,12 @@
                                         // Paper end - Add fail move event
                                     }
                                     // Paper end - Prevent moving into unloaded chunks
+                                    // ShreddedPaper start - Check we own the destination location
+                                    if (!TickThread.isTickThreadFor(this.player.level(), new Vec3(targetX, targetY, targetZ))) {
+                                        this.internalTeleport(PositionMoveRotation.of(this.player), Collections.emptySet());
+                                        return;
+                                    }
+                                    // ShreddedPaper end - Check we own the destination location
 
                                     if (this.shouldCheckPlayerMovement(isFallFlying)) {
                                         float metersPerTick = isFallFlying ? 300.0F : 100.0F;
@@ -2331,7 +_,9 @@
             for (ServerLevel level : this.server.getAllLevels()) {
                 Entity entity = packet.getEntity(level);
                 if (entity != null) {
+                    ShreddedPaper.ensureSync(this.player, entity, () -> { // ShreddedPaper - ensure on entity's thread
                     this.player.teleportTo(level, entity.getX(), entity.getY(), entity.getZ(), Set.of(), entity.getYRot(), entity.getXRot(), true, org.bukkit.event.player.PlayerTeleportEvent.TeleportCause.SPECTATE); // CraftBukkit
+                    }); // ShreddedPaper - ensure on entity's thread
                     return;
                 }
             }
@@ -2348,12 +_,15 @@
 
     @Override
     public void onDisconnect(final DisconnectionDetails details) {
+        final Consumer<Entity> runOnPlayersThread = e -> { // ShreddedPaper - run on player's thread
         LOGGER.info("{} lost connection: {}", this.player.getPlainTextName(), details.reason().getString());
         // Paper start - Fix kick event leave message not being sent
         final net.kyori.adventure.text.Component quitMessage = details.quitMessage().map(io.papermc.paper.adventure.PaperAdventure::asAdventure).orElse(null);
         this.removePlayerFromWorld(quitMessage);
         // Paper end - Fix kick event leave message not being sent
         super.onDisconnect(details);
+        }; // ShreddedPaper - run on player's thread
+        if (TickThread.isShutdownThread()) runOnPlayersThread.accept(this.player); else player.getBukkitEntity().taskScheduler.schedule(runOnPlayersThread, e -> LOGGER.error("Failed to run disconnect task for player {}", player.getName().getString(), new Exception("Disconnect task retired")), 1); // ShreddedPaper - run on player's thread
     }
 
     private void removePlayerFromWorld() {
@@ -2613,7 +_,7 @@
             this.player.resetLastActionTime();
             // CraftBukkit start
             if (sync) {
-                this.server.execute(chatHandler);
+                ShreddedPaper.runSync(player, chatHandler); // ShreddedPaper - run on player's thread
             } else {
                 chatHandler.run();
             }
@@ -3075,22 +_,26 @@
         switch (action) {
             case PERFORM_RESPAWN:
                 if (this.player.wonGame) {
+                    this.server.getPlayerList().respawnAsync(this.player, true, Entity.RemovalReason.CHANGED_DIMENSION, RespawnReason.END_PORTAL).thenAccept(player -> { // CraftBukkit // ShreddedPaper - async respawn
+                    this.player = player; // ShreddedPaper - async respawn
                     this.player.wonGame = false;
-                    this.player = this.server.getPlayerList().respawn(this.player, true, Entity.RemovalReason.CHANGED_DIMENSION, RespawnReason.END_PORTAL); // CraftBukkit
                     this.resetPosition();
                     this.restartClientLoadTimerAfterRespawn();
                     CriteriaTriggers.CHANGED_DIMENSION.trigger(this.player, Level.END, Level.OVERWORLD);
+                    }); // ShreddedPaper - async respawn
                 } else {
                     if (this.player.getHealth() > 0.0F) {
                         return;
                     }
 
-                    this.player = this.server.getPlayerList().respawn(this.player, false, Entity.RemovalReason.KILLED, RespawnReason.DEATH); // CraftBukkit
+                    this.server.getPlayerList().respawnAsync(this.player, false, Entity.RemovalReason.KILLED, RespawnReason.DEATH).thenAccept(player -> { // CraftBukkit // ShreddedPaper - async respawn
+                    this.player = player; // ShreddedPaper - async respawn
                     this.resetPosition();
                     this.restartClientLoadTimerAfterRespawn();
                     if (this.server.isHardcore()) {
                         this.player.setGameMode(GameType.SPECTATOR, org.bukkit.event.player.PlayerGameModeChangeEvent.Cause.HARDCORE_DEATH, null); // Paper - Expand PlayerGameModeChangeEvent
                     }
+                    }); // ShreddedPaper - async respawn
                 }
                 break;
             case REQUEST_STATS:
@@ -3637,7 +_,7 @@
         }
         List<String> lines = Stream.of(linesArray).map(ChatFormatting::stripFormatting).collect(Collectors.toList());
         // Paper end - Limit client sign length
-        this.filterTextPacket(lines).thenAcceptAsync(filteredLines -> this.updateSignText(packet, (List<FilteredText>)filteredLines), this.server);
+        this.filterTextPacket(lines).thenAcceptAsync(filteredLines -> this.updateSignText(packet, (List<FilteredText>)filteredLines), r -> ShreddedPaper.ensureSync(this.player, r)); // ShreddedPaper - run on region thread
     }
 
     private void updateSignText(final ServerboundSignUpdatePacket packet, final List<FilteredText> lines) {
@@ -3895,7 +_,7 @@
 
     @Override
     public void disconnectAsync(final net.minecraft.network.DisconnectionDetails disconnectionInfo) {
-        if (this.cserver.isPrimaryThread()) {
+        if (TickThread.isTickThreadFor(this.player)) { // ShreddedPaper
             this.disconnect(disconnectionInfo);
             return;
         }
