--- a/net/minecraft/network/Connection.java
+++ b/net/minecraft/network/Connection.java
@@ -1,7 +_,9 @@
 package net.minecraft.network;
 
+import ca.spottedleaf.moonrise.common.util.TickThread;
 import com.google.common.collect.Queues;
 import com.mojang.logging.LogUtils;
+import io.multipaper.shreddedpaper.config.ShreddedPaperConfiguration;
 import io.netty.bootstrap.Bootstrap;
 import io.netty.channel.Channel;
 import io.netty.channel.ChannelException;
@@ -45,8 +_,10 @@
 import net.minecraft.network.protocol.login.LoginProtocols;
 import net.minecraft.network.protocol.status.ClientStatusPacketListener;
 import net.minecraft.network.protocol.status.StatusProtocols;
+import net.minecraft.server.MinecraftServer;
 import net.minecraft.server.RunningOnDifferentThreadException;
 import net.minecraft.server.network.EventLoopGroupHolder;
+import net.minecraft.server.network.ServerGamePacketListenerImpl;
 import net.minecraft.util.Mth;
 import net.minecraft.util.Util;
 import net.minecraft.util.debugchart.LocalSampleLogger;
@@ -303,6 +_,7 @@
 
     private static void syncAfterConfigurationChange(ChannelFuture future) {
         try {
+            if (TickThread.isTickThread()) MinecraftServer.getServer().managedBlock(future::isDone); // ShreddedPaper - don't block main thread
             future.syncUninterruptibly();
         } catch (Exception var2) {
             if (var2 instanceof ClosedChannelException) {
@@ -400,7 +_,7 @@
 
     public void send(Packet<?> packet, @Nullable ChannelFutureListener sendListener, boolean flush) {
         // Paper start - Optimize network: Handle oversized packets better
-        final boolean connected = this.isConnected();
+        final boolean connected = true || this.isConnected(); // ShreddedPaper - Checking if the player is connected is expensive
         if (!connected && !this.preparing) {
             return;
         }
@@ -500,7 +_,7 @@
     }
 
     // Paper start - Optimize network: Rewrite this to be safer if ran off main thread
-    private boolean flushQueue() {
+    public boolean flushQueue() { // ShreddedPaper - make public
         if (!this.isConnected()) {
             return true;
         }
@@ -556,7 +_,8 @@
     private static int currTick; // Paper - Buffer joins to world
     private static int tickSecond; // Purpur - Max joins per second
     public void tick() {
-        this.flushQueue();
+        if (this.packetListener instanceof ServerGamePacketListenerImpl serverGamePacketListener && !TickThread.isTickThreadFor(serverGamePacketListener.player)) return; // ShreddedPaper
+        if (!ShreddedPaperConfiguration.get().optimizations.flushQueueInParallel) this.flushQueue(); // ShreddedPaper - Add option to flush queue in parallel
         // Paper start - Buffer joins to world
         if (Connection.currTick != net.minecraft.server.MinecraftServer.currentTick) {
             Connection.currTick = net.minecraft.server.MinecraftServer.currentTick;
