--- a/net/minecraft/server/network/config/PrepareSpawnTask.java
+++ b/net/minecraft/server/network/config/PrepareSpawnTask.java
@@ -4,6 +_,8 @@
 import java.util.Optional;
 import java.util.concurrent.CompletableFuture;
 import java.util.function.Consumer;
+
+import io.multipaper.shreddedpaper.ShreddedPaper;
 import net.minecraft.core.BlockPos;
 import net.minecraft.network.Connection;
 import net.minecraft.network.protocol.Packet;
@@ -132,9 +_,9 @@
         };
     }
 
-    public ServerPlayer spawnPlayer(final Connection connection, final CommonListenerCookie cookie) {
+    public void spawnPlayer(final Connection connection, final CommonListenerCookie cookie) { // ShreddedPaper - return void
         if (this.state instanceof PrepareSpawnTask.Ready ready) {
-            return ready.spawn(connection, cookie);
+            ready.spawn(connection, cookie); // ShreddedPaper - return void
         } else {
             throw new IllegalStateException("Player spawn was not ready");
         }
@@ -264,7 +_,9 @@
             this.spawnLevel.getChunkSource().addTicketWithRadius(TicketType.PLAYER_SPAWN, ChunkPos.containing(BlockPos.containing(this.spawnPosition)), 3);
         }
 
-        public ServerPlayer spawn(final Connection connection, final CommonListenerCookie cookie) {
+        public void spawn(final Connection connection, final CommonListenerCookie cookie) { // ShreddedPaper - return void
+            ShreddedPaper.ensureSync(this.spawnLevel, BlockPos.containing(this.spawnPosition), () -> { // ShreddedPaper - spawn on spawn positions' thread
+            if (!connection.isConnected()) return; // ShreddedPaper - ensure still connected before continuing
             ChunkPos spawnChunk = ChunkPos.containing(BlockPos.containing(this.spawnPosition));
             this.spawnLevel.waitForEntities(spawnChunk, 3);
             // Paper start - configuration api - possibly use legacy saved server player instance
@@ -315,8 +_,9 @@
                     player.loadAndSpawnEnderPearls(tag);
                     player.loadAndSpawnParentVehicle(tag);
                 });
-                return player;
+                // return player; // ShreddedPaper - return void
             }
+            }); // ShreddedPaper - spawn on spawn positions' thread
         }
     }
 
