--- a/net/minecraft/world/entity/projectile/FireworkRocketEntity.java
+++ b/net/minecraft/world/entity/projectile/FireworkRocketEntity.java
@@ -1,5 +_,6 @@
 package net.minecraft.world.entity.projectile;
 
+import ca.spottedleaf.moonrise.common.util.TickThread;
 import it.unimi.dsi.fastutil.doubles.DoubleDoubleImmutablePair;
 import java.util.List;
 import java.util.OptionalInt;
@@ -156,6 +_,12 @@
                     handHoldingItemAngle = Vec3.ZERO;
                 }
 
+                // ShreddedPaper start - remove firework rocket if entity teleported away
+                if (!TickThread.isTickThreadFor((ServerLevel) this.level(), new Vec3(this.attachedToEntity.getX() + handHoldingItemAngle.x, this.attachedToEntity.getY() + handHoldingItemAngle.y, this.attachedToEntity.getZ() + handHoldingItemAngle.z))) {
+                    this.discard(org.bukkit.event.entity.EntityRemoveEvent.Cause.DISCARD); // ShreddedPaper - add Bukkit remove cause
+                    return;
+                }
+                // ShreddedPaper end - remove firework rocket if entity teleported away
                 this.setPos(
                     this.attachedToEntity.getX() + handHoldingItemAngle.x,
                     this.attachedToEntity.getY() + handHoldingItemAngle.y,
