--- a/ca/spottedleaf/moonrise/patches/chunk_system/level/entity/EntityLookup.java
+++ b/ca/spottedleaf/moonrise/patches/chunk_system/level/entity/EntityLookup.java
@@ -422,9 +_,12 @@
             return false;
         }
 
+        final Throwable addingStackTrace = new Exception(System.currentTimeMillis() + ": Adding " + entity + " to world " + WorldUtil.getWorldName(this.world)); // ShreddedPaper - add addedStackTrace
+
         Entity currentlyMapped = this.entityById.putIfAbsent((long)entity.getId(), entity);
         if (currentlyMapped != null) {
-            LOGGER.warn("Entity id already exists: " + entity.getId() + ", mapped to " + currentlyMapped + ", can't add " + entity);
+            LOGGER.warn("Entity id already exists: " + entity.getId() + ", mapped to " + currentlyMapped + ", can't add " + entity, addingStackTrace); // ShreddedPaper - add addedStackTrace
+            LOGGER.warn("Previously added here", currentlyMapped.addedStackTrace); // ShreddedPaper - add addedStackTrace
             return false;
         }
 
@@ -432,9 +_,12 @@
         if (currentlyMapped != null) {
             // need to remove mapping for id
             this.entityById.remove((long)entity.getId(), entity);
-            LOGGER.warn("Entity uuid already exists: " + entity.getUUID() + ", mapped to " + currentlyMapped + ", can't add " + entity);
+            LOGGER.warn("Entity uuid already exists: " + entity.getUUID() + ", mapped to " + currentlyMapped + ", can't add " + entity, addingStackTrace); // ShreddedPaper - add addedStackTrace
+            LOGGER.warn("Previously added here", currentlyMapped.addedStackTrace); // ShreddedPaper - add addedStackTrace
             return false;
         }
+
+        entity.addedStackTrace = addingStackTrace; // ShreddedPaper - add addedStackTrace
 
         ((ChunkSystemEntity)entity).moonrise$setSectionX(sectionX);
         ((ChunkSystemEntity)entity).moonrise$setSectionY(sectionY);
