--- a/src/main/java/io/papermc/paper/plugin/manager/PaperPermissionManager.java
+++ b/src/main/java/io/papermc/paper/plugin/manager/PaperPermissionManager.java
@@ -106,7 +_,7 @@
 
 
     @Override
-    public void subscribeToPermission(@NotNull String permission, @NotNull Permissible permissible) {
+    public synchronized void subscribeToPermission(@NotNull String permission, @NotNull Permissible permissible) { // ShreddedPaper - synchronized
         String name = permission.toLowerCase(java.util.Locale.ENGLISH);
         Map<Permissible, Boolean> map = this.permSubs().computeIfAbsent(name, k -> new WeakHashMap<>());
 
@@ -114,7 +_,7 @@
     }
 
     @Override
-    public void unsubscribeFromPermission(@NotNull String permission, @NotNull Permissible permissible) {
+    public synchronized void unsubscribeFromPermission(@NotNull String permission, @NotNull Permissible permissible) { // ShreddedPaper - synchronized
         String name = permission.toLowerCase(java.util.Locale.ENGLISH);
         Map<Permissible, Boolean> map = this.permSubs().get(name);
 
@@ -141,14 +_,14 @@
     }
 
     @Override
-    public void subscribeToDefaultPerms(boolean op, @NotNull Permissible permissible) {
+    public synchronized void subscribeToDefaultPerms(boolean op, @NotNull Permissible permissible) { // ShreddedPaper - synchronized
         Map<Permissible, Boolean> map = this.defSubs().computeIfAbsent(op, k -> new WeakHashMap<>());
 
         map.put(permissible, true);
     }
 
     @Override
-    public void unsubscribeFromDefaultPerms(boolean op, @NotNull Permissible permissible) {
+    public synchronized void unsubscribeFromDefaultPerms(boolean op, @NotNull Permissible permissible) { // ShreddedPaper - synchronized
         Map<Permissible, Boolean> map = this.defSubs().get(op);
 
         if (map != null) {
