--- a/src/main/java/io/papermc/paper/plugin/manager/PaperPermissionManager.java
+++ b/src/main/java/io/papermc/paper/plugin/manager/PaperPermissionManager.java
@@ -108,23 +_,33 @@
     @Override
     public void subscribeToPermission(@NotNull String permission, @NotNull Permissible permissible) {
         String name = permission.toLowerCase(java.util.Locale.ENGLISH);
-        Map<Permissible, Boolean> map = this.permSubs().computeIfAbsent(name, k -> new WeakHashMap<>());
-
-        map.put(permissible, true);
+        // ShreddedPaper start - thread-safe permission manager
+        this.permSubs().compute(name, (k, map) -> {
+            if (map == null) map = new WeakHashMap<>();
+
+            map.put(permissible, true);
+
+            return map;
+        });
+        // ShreddedPaper end - thread-safe permission manager
     }
 
     @Override
     public void unsubscribeFromPermission(@NotNull String permission, @NotNull Permissible permissible) {
         String name = permission.toLowerCase(java.util.Locale.ENGLISH);
-        Map<Permissible, Boolean> map = this.permSubs().get(name);
-
-        if (map != null) {
-            map.remove(permissible);
-
-            if (map.isEmpty()) {
-                this.permSubs().remove(name);
+        // ShreddedPaper start - thread-safe permission manager
+        this.permSubs().compute(name, (k, map) -> {
+            if (map != null) {
+                map.remove(permissible);
+
+                if (map.isEmpty()) {
+                    return null; // this.permSubs().remove(name);
+                }
             }
-        }
+
+            return map;
+        });
+        // ShreddedPaper end - thread-safe permission manager
     }
 
     @Override
@@ -142,22 +_,32 @@
 
     @Override
     public void subscribeToDefaultPerms(boolean op, @NotNull Permissible permissible) {
-        Map<Permissible, Boolean> map = this.defSubs().computeIfAbsent(op, k -> new WeakHashMap<>());
-
-        map.put(permissible, true);
+        // ShreddedPaper start - thread-safe permission manager
+        this.defSubs().compute(op, (k, map) -> {
+            if (map == null) map = new WeakHashMap<>();
+
+            map.put(permissible, true);
+
+            return map;
+        });
+        // ShreddedPaper end - thread-safe permission manager
     }
 
     @Override
     public void unsubscribeFromDefaultPerms(boolean op, @NotNull Permissible permissible) {
-        Map<Permissible, Boolean> map = this.defSubs().get(op);
-
-        if (map != null) {
-            map.remove(permissible);
-
-            if (map.isEmpty()) {
-                this.defSubs().remove(op);
+        // ShreddedPaper start - thread-safe permission manager
+        this.defSubs().compute(op, (k, map) -> {
+            if (map != null) {
+                map.remove(permissible);
+
+                if (map.isEmpty()) {
+                    return null; // this.defSubs().remove(op);
+                }
             }
-        }
+
+            return map;
+        });
+        // ShreddedPaper end - thread-safe permission manager
     }
 
     @Override
