--- a/net/minecraft/world/entity/projectile/throwableitemprojectile/ThrownEnderpearl.java
+++ b/net/minecraft/world/entity/projectile/throwableitemprojectile/ThrownEnderpearl.java
@@ -1,6 +_,8 @@
 package net.minecraft.world.entity.projectile.throwableitemprojectile;
 
 import java.util.UUID;
+
+import io.multipaper.shreddedpaper.ShreddedPaper;
 import net.minecraft.core.BlockPos;
 import net.minecraft.core.SectionPos;
 import net.minecraft.core.particles.ParticleTypes;
@@ -20,6 +_,7 @@
 import net.minecraft.world.item.Item;
 import net.minecraft.world.item.ItemStack;
 import net.minecraft.world.item.Items;
+import net.minecraft.world.level.ChunkPos;
 import net.minecraft.world.level.Level;
 import net.minecraft.world.level.block.Blocks;
 import net.minecraft.world.level.block.state.BlockState;
@@ -102,6 +_,7 @@
             Entity owner = this.getOwner();
             if (owner != null && isAllowedToTeleportOwner(owner, level)) {
                 Vec3 teleportPos = this.oldPosition();
+                ShreddedPaper.ensureSync(owner, level, ChunkPos.containing(BlockPos.containing(teleportPos)), () -> { // ShreddedPaper - teleport async
                 if (owner instanceof ServerPlayer player) {
                     if (player.connection.isAcceptingMessages()) {
                         // CraftBukkit start
@@ -111,7 +_,7 @@
                         ServerPlayer newOwner = player.teleport(new TeleportTransition(
                             level, teleportPos, Vec3.ZERO, 0.0F, 0.0F, Relative.union(Relative.ROTATION, Relative.DELTA), TeleportTransition.DO_NOTHING).withCause(org.bukkit.event.player.PlayerTeleportEvent.TeleportCause.ENDER_PEARL));
                         if (newOwner == null) {
-                            this.discard(org.bukkit.event.entity.EntityRemoveEvent.Cause.HIT);
+                            // this.discard(org.bukkit.event.entity.EntityRemoveEvent.Cause.HIT); // ShreddedPaper - already removed as part of async teleport
                             return;
                         }
                         // CraftBukkit end
@@ -157,6 +_,7 @@
 
                     this.playSound(level, teleportPos);
                 }
+                }); // ShreddedPaper - teleport async
 
                 this.discard(org.bukkit.event.entity.EntityRemoveEvent.Cause.HIT); // CraftBukkit - add Bukkit remove cause
             } else {
