--- a/net/minecraft/world/item/MapItem.java
+++ b/net/minecraft/world/item/MapItem.java
@@ -66,6 +_,7 @@
     }
 
     public void update(Level level, Entity viewer, MapItemSavedData data) {
+        synchronized (data) { // ShreddedPaper - thread-safe
         if (level.dimension() == data.dimension && viewer instanceof Player) {
             int i = 1 << data.scale;
             int i1 = data.centerX;
@@ -178,6 +_,7 @@
                 }
             }
         }
+        } // ShreddedPaper - thread-safe
     }
 
     private BlockState getCorrectStateForFluidBlock(Level level, BlockState state, BlockPos pos) {
@@ -192,6 +_,7 @@
     public static void renderBiomePreviewMap(ServerLevel level, ItemStack stack) {
         MapItemSavedData savedData = getSavedData(stack, level);
         if (savedData != null) {
+            synchronized (savedData) { // ShreddedPaper - thread-safe
             savedData.isExplorerMap = true; // Purpur - Explorer Map API
             if (level.dimension() == savedData.dimension) {
                 int i = 1 << savedData.scale;
@@ -262,6 +_,7 @@
                     }
                 }
             }
+            } // ShreddedPaper - thread-safe
         }
     }
 
