--- a/net/minecraft/world/level/portal/TeleportTransition.java
+++ b/net/minecraft/world/level/portal/TeleportTransition.java
@@ -1,6 +_,9 @@
 package net.minecraft.world.level.portal;
 
 import java.util.Set;
+import java.util.concurrent.CompletableFuture;
+
+import io.multipaper.shreddedpaper.ShreddedPaper;
 import net.minecraft.core.BlockPos;
 import net.minecraft.network.protocol.game.ClientboundLevelEventPacket;
 import net.minecraft.server.level.ServerLevel;
@@ -160,12 +_,13 @@
         entity.placePortalTicket(BlockPos.containing(entity.position()));
     }
 
-    public static TeleportTransition createDefault(final ServerPlayer player, final TeleportTransition.PostTeleportTransition postTeleportTransition) {
+    public static CompletableFuture<TeleportTransition> createDefaultAsync(final ServerPlayer player, final TeleportTransition.PostTeleportTransition postTeleportTransition) { // ShreddedPaper - respawn async
         ServerLevel newLevel = player.level().getServer().findRespawnDimension();
         LevelData.RespawnData respawnData = newLevel.getRespawnData();
+        return findAdjustedSharedSpawnPosAsync(newLevel, player).thenApply(pos -> { // ShreddedPaper - respawn async
         return new TeleportTransition(
             newLevel,
-            findAdjustedSharedSpawnPos(newLevel, player),
+            pos, // ShreddedPaper - respawn async
             Vec3.ZERO,
             respawnData.yaw(),
             respawnData.pitch(),
@@ -174,14 +_,16 @@
             Set.of(),
             postTeleportTransition
         );
+        }); // ShreddedPaper - respawn async
     }
 
-    public static TeleportTransition missingRespawnBlock(final ServerPlayer player, final TeleportTransition.PostTeleportTransition postTeleportTransition) {
+    public static CompletableFuture<TeleportTransition> missingRespawnBlockAsync(final ServerPlayer player, final TeleportTransition.PostTeleportTransition postTeleportTransition) { // ShreddedPaper - respawn async
         ServerLevel newLevel = player.level().getServer().findRespawnDimension();
         LevelData.RespawnData respawnData = newLevel.getRespawnData();
+        return findAdjustedSharedSpawnPosAsync(newLevel, player).thenApply(pos -> { // ShreddedPaper - respawn async
         return new TeleportTransition(
             newLevel,
-            findAdjustedSharedSpawnPos(newLevel, player),
+            pos, // ShreddedPaper - respawn async
             Vec3.ZERO,
             respawnData.yaw(),
             respawnData.pitch(),
@@ -190,10 +_,11 @@
             Set.of(),
             postTeleportTransition
         );
+        }); // ShreddedPaper - respawn async
     }
 
-    private static Vec3 findAdjustedSharedSpawnPos(final ServerLevel newLevel, final Entity entity) {
-        return Vec3.atBottomCenterOf(entity.adjustSpawnLocation(newLevel, newLevel.getRespawnData().pos()));
+    private static CompletableFuture<Vec3> findAdjustedSharedSpawnPosAsync(final ServerLevel newLevel, final Entity entity) { // ShreddedPaper - respawn async
+        return CompletableFuture.supplyAsync(() -> Vec3.atBottomCenterOf(entity.adjustSpawnLocation(newLevel, newLevel.getRespawnData().pos())), r -> ShreddedPaper.ensureSync(newLevel, newLevel.getRespawnData().pos(), r)); // ShreddedPaper - respawn async
     }
 
     public TeleportTransition withRotation(final float yRot, final float xRot) {
