--- a/net/minecraft/world/level/block/entity/CommandBlockEntity.java
+++ b/net/minecraft/world/level/block/entity/CommandBlockEntity.java
@@ -39,6 +_,13 @@
         }
         // CraftBukkit end
 
+        // ShreddedPaper start - Add getBlockPos() for CloseableCommandBlockSource
+        @Override
+        public BlockPos getBlockPos() {
+            return CommandBlockEntity.this.getBlockPos();
+        }
+        // ShreddedPaper end - Add getBlockPos() for CloseableCommandBlockSource
+
         @Override
         public void setCommand(String command) {
             super.setCommand(command);
