--- a/net/minecraft/world/level/block/RedstoneTorchBlock.java
+++ b/net/minecraft/world/level/block/RedstoneTorchBlock.java
@@ -2,6 +_,7 @@
 
 import com.google.common.collect.Lists;
 import com.mojang.serialization.MapCodec;
+import io.multipaper.shreddedpaper.region.RegionPos;
 import java.util.List;
 import java.util.Map;
 import java.util.WeakHashMap;
@@ -74,7 +_,7 @@
     protected void tick(final BlockState state, final ServerLevel level, final BlockPos pos, final RandomSource random) {
         boolean neighborSignal = this.hasNeighborSignal(level, pos, state);
         // Paper start - Faster redstone torch rapid clock removal
-        java.util.ArrayDeque<RedstoneTorchBlock.Toggle> toggles = level.redstoneUpdateInfos;
+        java.util.ArrayDeque<RedstoneTorchBlock.Toggle> toggles = level.chunkSource.tickingRegions.get(RegionPos.forBlockPos(pos)).redstoneUpdateInfos; // ShreddedPaper - move redstoneUpdateInfos to the region
         if (toggles != null) {
             RedstoneTorchBlock.Toggle curr;
             while ((curr = toggles.peek()) != null && level.getGameTime() - curr.when > 60L) {
@@ -142,10 +_,12 @@
 
     private static boolean isToggledTooFrequently(final Level level, final BlockPos pos, final boolean add) {
         // Paper start - Faster redstone torch rapid clock removal
-        java.util.ArrayDeque<RedstoneTorchBlock.Toggle> toggles = level.redstoneUpdateInfos;
+        // ShreddedPaper start - move redstoneUpdateInfos to the region
+        java.util.ArrayDeque<RedstoneTorchBlock.Toggle> toggles = ((ServerLevel) level).chunkSource.tickingRegions.get(RegionPos.forBlockPos(pos)).redstoneUpdateInfos;
         if (toggles == null) {
-            toggles = level.redstoneUpdateInfos = new java.util.ArrayDeque<>();
+            toggles = ((ServerLevel) level).chunkSource.tickingRegions.get(RegionPos.forBlockPos(pos)).redstoneUpdateInfos = new java.util.ArrayDeque<>();
         }
+        // ShreddedPaper end - move redstoneUpdateInfos to the region
         // Paper end - Faster redstone torch rapid clock removal
         if (add) {
             toggles.add(new RedstoneTorchBlock.Toggle(pos.immutable(), level.getGameTime()));
