--- a/net/minecraft/world/level/block/WitherSkullBlock.java
+++ b/net/minecraft/world/level/block/WitherSkullBlock.java
@@ -49,7 +_,7 @@
     }
 
     public static void checkSpawn(Level level, BlockPos pos, SkullBlockEntity blockEntity) {
-        if (level.captureBlockStates) return; // CraftBukkit
+        if (level.captureBlockStates.get()) return; // CraftBukkit // ShreddedPaper - thread local
         if (!level.isClientSide()) {
             BlockState blockState = blockEntity.getBlockState();
             boolean flag = blockState.is(Blocks.WITHER_SKELETON_SKULL) || blockState.is(Blocks.WITHER_SKELETON_WALL_SKULL);
