From 0000000000000000000000000000000000000000 Mon Sep 17 00:00:00 2001
From: PureGero <puregero@gmail.com>
Date: Wed, 16 Jul 2025 18:42:23 +0900
Subject: [PATCH] Optimization: Use lazyExecute if we aren't flushing


diff --git a/src/main/java/net/minecraft/network/Connection.java b/src/main/java/net/minecraft/network/Connection.java
index d2494f38d1208e2e21daf7c45ccbfe07f3e02f7c..bb28c13015a2aa7622d2f762dc797eb0344400db 100644
--- a/src/main/java/net/minecraft/network/Connection.java
+++ b/src/main/java/net/minecraft/network/Connection.java
@@ -4,6 +4,7 @@ import com.google.common.base.Suppliers;
 import com.google.common.collect.Queues;
 import com.google.common.util.concurrent.ThreadFactoryBuilder;
 import com.mojang.logging.LogUtils;
+import io.multipaper.shreddedpaper.config.ShreddedPaperConfiguration;
 import io.netty.bootstrap.Bootstrap;
 import io.netty.channel.Channel;
 import io.netty.channel.ChannelException;
@@ -43,6 +44,7 @@ import java.util.function.Supplier;
 import javax.annotation.Nullable;
 import javax.crypto.Cipher;
 
+import io.netty.util.concurrent.AbstractEventExecutor;
 import io.papermc.paper.util.TickThread;
 import net.minecraft.SharedConstants;
 import net.minecraft.Util;
@@ -491,6 +493,14 @@ public class Connection extends SimpleChannelInboundHandler<Packet<?>> {
         if (this.channel.eventLoop().inEventLoop()) {
             this.doSendPacket(packet, callbacks, flush);
         } else {
+            // ShreddedPaper start - Use lazyExecute if we aren't flushing
+            if (ShreddedPaperConfiguration.get().optimizations.useLazyExecuteWhenNotFlushing && !flush) {
+                ((AbstractEventExecutor) this.channel.eventLoop()).lazyExecute(() -> {
+                    this.doSendPacket(packet, callbacks, flush);
+                });
+                return;
+            }
+            // ShreddedPaper end - Use lazyExecute if we aren't flushing
             this.channel.eventLoop().execute(() -> {
                 this.doSendPacket(packet, callbacks, flush);
             });
