From 0000000000000000000000000000000000000000 Mon Sep 17 00:00:00 2001
From: PureGero <puregero@gmail.com>
Date: Wed, 16 Jul 2025 19:19:00 +0900
Subject: [PATCH] Optimization: Don't check isConnected


diff --git a/src/main/java/net/minecraft/network/Connection.java b/src/main/java/net/minecraft/network/Connection.java
index bb28c13015a2aa7622d2f762dc797eb0344400db..51f67289fcf62614a5fd610474ab423544f84ebe 100644
--- a/src/main/java/net/minecraft/network/Connection.java
+++ b/src/main/java/net/minecraft/network/Connection.java
@@ -444,7 +444,7 @@ public class Connection extends SimpleChannelInboundHandler<Packet<?>> {
 
     public void send(Packet<?> packet, @Nullable PacketSendListener callbacks, boolean flush) {
         // Paper start - Optimize network: Handle oversized packets better
-        final boolean connected = this.isConnected();
+        final boolean connected = true || this.isConnected(); // ShreddedPaper - Checking if the player is connected is expensive
         if (!connected && !this.preparing) {
             return;
         }
