From 0000000000000000000000000000000000000000 Mon Sep 17 00:00:00 2001
From: PureGero <puregero@gmail.com>
Date: Sat, 25 Oct 2025 19:24:51 +0900
Subject: [PATCH] Optimization: Load chunk before joining player


diff --git a/src/main/java/net/minecraft/server/players/PlayerList.java b/src/main/java/net/minecraft/server/players/PlayerList.java
index dc99d088d6d915de789a1215df810a4a86212e42..973110b208a4026fd296663f58cd60960d9c6a1c 100644
--- a/src/main/java/net/minecraft/server/players/PlayerList.java
+++ b/src/main/java/net/minecraft/server/players/PlayerList.java
@@ -292,7 +292,7 @@ public abstract class PlayerList {
 
         Location loc = ev.getSpawnLocation();
         // ShreddedPaper start - join the region's thread
-        ShreddedPaper.ensureSync(loc, () -> placeNewPlayer3(connection, player, clientData, worldserver1, loc, optional, s, s1));
+        worldserver1.getWorld().getChunkAtAsync(loc).thenRun(() -> placeNewPlayer3(connection, player, clientData, worldserver1, loc, optional, s, s1)); // ShreddedPaper - ensure spawn chunk is loaded
     }
 
     public void placeNewPlayer3(Connection connection, ServerPlayer player, CommonListenerCookie clientData, ServerLevel worldserver1, Location loc, Optional<CompoundTag> optional, String s, String s1) {
