--- a/net/minecraft/world/level/saveddata/maps/MapIndex.java
+++ b/net/minecraft/world/level/saveddata/maps/MapIndex.java
@@ -22,7 +_,7 @@
         this.lastMapId = lastMapId;
     }
 
-    public MapId getNextMapId() {
+    public synchronized MapId getNextMapId() { // ShreddedPaper - thread-safe
         MapId mapId = new MapId(++this.lastMapId);
         this.setDirty();
         return mapId;
