From 0000000000000000000000000000000000000000 Mon Sep 17 00:00:00 2001
From: PureGero <puregero@gmail.com>
Date: Sat, 25 May 2024 18:44:13 +0900
Subject: [PATCH] Add isFoliaSupported


diff --git a/src/main/java/org/bukkit/plugin/PluginDescriptionFile.java b/src/main/java/org/bukkit/plugin/PluginDescriptionFile.java
index a857e46fa6f0c212db93193e1fdd8b0ea9c33c38..26efead73ac9a1d0d33a4137738cf3ab958af271 100644
--- a/src/main/java/org/bukkit/plugin/PluginDescriptionFile.java
+++ b/src/main/java/org/bukkit/plugin/PluginDescriptionFile.java
@@ -260,6 +260,7 @@ public final class PluginDescriptionFile implements io.papermc.paper.plugin.conf
     private Set<PluginAwareness> awareness = ImmutableSet.of();
     private String apiVersion = null;
     private List<String> libraries = ImmutableList.of();
+    private String foliaSupported = null; // ShreddedPaper
     // Paper start - plugin loader api
     private String paperPluginLoader;
     @org.jetbrains.annotations.ApiStatus.Internal @org.jetbrains.annotations.Nullable
@@ -1057,6 +1058,11 @@ public final class PluginDescriptionFile implements io.papermc.paper.plugin.conf
         return libraries;
     }
 
+    // ShreddedPaper start
+    public boolean isFoliaSupported() {
+        return "true".equalsIgnoreCase(foliaSupported);
+    }
+
     /**
      * @return unused
      * @deprecated unused
@@ -1258,6 +1264,12 @@ public final class PluginDescriptionFile implements io.papermc.paper.plugin.conf
         }
         // Paper end - plugin loader api
 
+        // ShreddedPaper start
+        if (map.get("folia-supported") != null) {
+            foliaSupported = map.get("folia-supported").toString();
+        }
+        // ShreddedPaper end
+
         try {
             lazyPermissions = (Map<?, ?>) map.get("permissions");
         } catch (ClassCastException ex) {
@@ -1336,6 +1348,12 @@ public final class PluginDescriptionFile implements io.papermc.paper.plugin.conf
             map.put("libraries", libraries);
         }
 
+        // ShreddedPaper start
+        if (foliaSupported != null) {
+            map.put("folia-supported", foliaSupported);
+        }
+        // ShreddedPaper end
+
         if (classLoaderOf != null) {
             map.put("class-loader-of", classLoaderOf);
         }
