From 0000000000000000000000000000000000000000 Mon Sep 17 00:00:00 2001
From: PureGero <puregero@gmail.com>
Date: Sun, 5 May 2024 13:59:19 +0900
Subject: [PATCH] Multithread entity ticking


diff --git a/src/main/java/io/multipaper/shreddedpaper/region/LevelChunkRegion.java b/src/main/java/io/multipaper/shreddedpaper/region/LevelChunkRegion.java
index 51ce3a1324793f3560cd718d5aa37b33ebbaafa0..616beda72d0e6fb94e54da45484559b651b7ebdd 100644
--- a/src/main/java/io/multipaper/shreddedpaper/region/LevelChunkRegion.java
+++ b/src/main/java/io/multipaper/shreddedpaper/region/LevelChunkRegion.java
@@ -1,6 +1,8 @@
 package io.multipaper.shreddedpaper.region;
 
+import io.papermc.paper.util.maplist.IteratorSafeOrderedReferenceSet;
 import net.minecraft.server.level.ServerLevel;
+import net.minecraft.world.entity.Entity;
 import net.minecraft.world.level.chunk.LevelChunk;
 
 import java.util.ArrayList;
@@ -12,6 +14,7 @@ public class LevelChunkRegion {
     private final ServerLevel level;
     private final RegionPos regionPos;
     private final List<LevelChunk> levelChunks = new ArrayList<>(RegionPos.REGION_SIZE * RegionPos.REGION_SIZE);
+    private final IteratorSafeOrderedReferenceSet<Entity> tickingEntities = new IteratorSafeOrderedReferenceSet<>(); // Use IteratorSafeOrderedReferenceSet to maintain entity tick order
 
     public LevelChunkRegion(ServerLevel level, RegionPos regionPos) {
         this.level = level;
@@ -28,6 +31,29 @@ public class LevelChunkRegion {
         }
     }
 
+    public synchronized void addTickingEntity(Entity entity) {
+        if (!tickingEntities.add(entity)) {
+            throw new IllegalStateException("Tried to add an entity that was already in the ticking list: " + entity);
+        }
+    }
+
+    public synchronized void removeTickingEntity(Entity entity) {
+        if (!tickingEntities.remove(entity)) {
+            throw new IllegalStateException("Tried to remove an entity that wasn't in the ticking list: " + entity);
+        }
+    }
+
+    public void forEachTickingEntity(Consumer<Entity> action) {
+        IteratorSafeOrderedReferenceSet.Iterator<Entity> iterator = this.tickingEntities.iterator();
+        try {
+            while (iterator.hasNext()) {
+                action.accept(iterator.next());
+            }
+        } finally {
+            iterator.finishedIterating();
+        }
+    }
+
     public ServerLevel getLevel() {
         return level;
     }
@@ -51,7 +77,9 @@ public class LevelChunkRegion {
     }
 
     public boolean isEmpty() {
-        return levelChunks.isEmpty();
+        return levelChunks.isEmpty()
+                && tickingEntities.size() == 0
+                ;
     }
 
 }
diff --git a/src/main/java/io/multipaper/shreddedpaper/region/LevelChunkRegionMap.java b/src/main/java/io/multipaper/shreddedpaper/region/LevelChunkRegionMap.java
index 89724b9d82813cc5d42eb1a0a3d4bf50e406a3ca..a7c9025433f0b0e260a1ff69022b7938dda6b541 100644
--- a/src/main/java/io/multipaper/shreddedpaper/region/LevelChunkRegionMap.java
+++ b/src/main/java/io/multipaper/shreddedpaper/region/LevelChunkRegionMap.java
@@ -2,6 +2,8 @@ package io.multipaper.shreddedpaper.region;
 
 import it.unimi.dsi.fastutil.longs.Long2ObjectOpenHashMap;
 import net.minecraft.server.level.ServerLevel;
+import net.minecraft.world.entity.Entity;
+import net.minecraft.world.level.ChunkPos;
 import net.minecraft.world.level.chunk.LevelChunk;
 import io.multipaper.shreddedpaper.util.SimpleStampedLock;
 
@@ -57,4 +59,39 @@ public class LevelChunkRegionMap {
         regionsCopy.forEach(consumer);
     }
 
+    public void addTickingEntity(Entity entity) {
+        if (entity.previousTickingChunkPosRegion != null) {
+            throw new IllegalStateException("Entity has already been added to a ticking list " + entity);
+        }
+
+        entity.previousTickingChunkPosRegion = entity.chunkPosition();
+        getOrCreate(RegionPos.forChunk(entity.chunkPosition())).addTickingEntity(entity);
+    }
+
+    public void removeTickingEntity(Entity entity) {
+        if (entity.previousTickingChunkPosRegion == null) {
+            throw new IllegalStateException("Entity has not been added to a ticking list " + entity);
+        }
+
+        getOrCreate(RegionPos.forChunk(entity.previousTickingChunkPosRegion)).removeTickingEntity(entity);
+        entity.previousTickingChunkPosRegion = null;
+    }
+
+    public void moveTickingEntity(Entity entity) {
+        if (entity.previousTickingChunkPosRegion == null) {
+            // Not ticking, ignore
+            return;
+        }
+
+        ChunkPos newChunkPos = entity.chunkPosition();
+        RegionPos fromRegion = RegionPos.forChunk(entity.previousTickingChunkPosRegion);
+        RegionPos toRegion = RegionPos.forChunk(newChunkPos);
+
+        if (!fromRegion.equals(toRegion)) {
+            entity.previousTickingChunkPosRegion = newChunkPos;
+            getOrCreate(fromRegion).removeTickingEntity(entity);
+            getOrCreate(toRegion).addTickingEntity(entity);
+        }
+    }
+
 }
\ No newline at end of file
diff --git a/src/main/java/io/multipaper/shreddedpaper/threading/ShreddedPaperChunkTicker.java b/src/main/java/io/multipaper/shreddedpaper/threading/ShreddedPaperChunkTicker.java
index 41e3f42b860ec3a9c030d0e627908ea5c84a2694..05e64b401e5ed59c79ad92e17fe47de3793f7ae2 100644
--- a/src/main/java/io/multipaper/shreddedpaper/threading/ShreddedPaperChunkTicker.java
+++ b/src/main/java/io/multipaper/shreddedpaper/threading/ShreddedPaperChunkTicker.java
@@ -59,6 +59,8 @@ public class ShreddedPaperChunkTicker {
 
             region.forEach(chunk -> this._tickChunk(level, chunk, spawnercreature_d));
 
+            region.forEachTickingEntity(ShreddedPaperEntityTicker::tickEntity);
+
             if (region.isEmpty()) {
                 level.chunkSource.tickingRegions.remove(region.getRegionPos());
             }
diff --git a/src/main/java/io/multipaper/shreddedpaper/threading/ShreddedPaperEntityTicker.java b/src/main/java/io/multipaper/shreddedpaper/threading/ShreddedPaperEntityTicker.java
new file mode 100644
index 0000000000000000000000000000000000000000..051084e44759febef79a5b5606967e677699b951
--- /dev/null
+++ b/src/main/java/io/multipaper/shreddedpaper/threading/ShreddedPaperEntityTicker.java
@@ -0,0 +1,38 @@
+package io.multipaper.shreddedpaper.threading;
+
+import net.minecraft.server.level.ServerLevel;
+import net.minecraft.world.entity.Entity;
+
+public class ShreddedPaperEntityTicker {
+
+    public static void tickEntity(Entity entity) {
+        ServerLevel level = (ServerLevel) entity.level();
+
+        entity.activatedPriorityReset = false; // Pufferfish - DAB
+        if (!entity.isRemoved()) {
+            if (false && level.shouldDiscardEntity(entity)) { // CraftBukkit - We prevent spawning in general, so this butchering is not needed
+                entity.discard();
+            } else if (!level.tickRateManager().isEntityFrozen(entity)) {
+                //gameprofilerfiller.push("checkDespawn"); // Purpur
+                entity.checkDespawn();
+                //gameprofilerfiller.pop(); // Purpur
+                if (true || level.chunkSource.chunkMap.getDistanceManager().inEntityTickingRange(entity.chunkPosition().toLong())) { // Paper - now always true if in the ticking list
+                    Entity entity1 = entity.getVehicle();
+
+                    if (entity1 != null) {
+                        if (!entity1.isRemoved() && entity1.hasPassenger(entity)) {
+                            return;
+                        }
+
+                        entity.stopRiding();
+                    }
+
+                    //gameprofilerfiller.push("tick"); // Purpur
+                    level.guardEntityTick(level::tickNonPassenger, entity);
+                    //gameprofilerfiller.pop(); // Purpur
+                }
+            }
+        }
+    }
+
+}
diff --git a/src/main/java/net/minecraft/server/level/ServerChunkCache.java b/src/main/java/net/minecraft/server/level/ServerChunkCache.java
index 7a75de1ad600505ce44b299e8eabfece9b48c387..b1c8c652babdc3549e52cba8d949edc0a9d87d88 100644
--- a/src/main/java/net/minecraft/server/level/ServerChunkCache.java
+++ b/src/main/java/net/minecraft/server/level/ServerChunkCache.java
@@ -720,12 +720,14 @@ public class ServerChunkCache extends ChunkSource {
             if (_pufferfish_spawnCountsReady.getAndSet(false)) {
                 net.minecraft.server.MinecraftServer.getServer().mobSpawnExecutor.submit(() -> {
                     int mapped = distanceManager.getNaturalSpawnChunkCount();
-                    io.papermc.paper.util.maplist.IteratorSafeOrderedReferenceSet.Iterator<Entity> objectiterator =
-                            level.entityTickList.entities.iterator(io.papermc.paper.util.maplist.IteratorSafeOrderedReferenceSet.ITERATOR_FLAG_SEE_ADDITIONS);
-                    gg.pufferfish.pufferfish.util.IterableWrapper<Entity> wrappedIterator =
-                            new gg.pufferfish.pufferfish.util.IterableWrapper<>(objectiterator);
-                    lastSpawnState = NaturalSpawner.createState(mapped, wrappedIterator, this::getFullChunk, null, true);
-                    objectiterator.finishedIterating();
+                    // ShreddedPaper start - use this.level.getAllEntities()
+                    // io.papermc.paper.util.maplist.IteratorSafeOrderedReferenceSet.Iterator<Entity> objectiterator =
+                    //         level.entityTickList.entities.iterator(io.papermc.paper.util.maplist.IteratorSafeOrderedReferenceSet.ITERATOR_FLAG_SEE_ADDITIONS);
+                    // gg.pufferfish.pufferfish.util.IterableWrapper<Entity> wrappedIterator =
+                    //         new gg.pufferfish.pufferfish.util.IterableWrapper<>(objectiterator);
+                    lastSpawnState = NaturalSpawner.createState(mapped, this.level.getAllEntities(), this::getFullChunk, null, true);
+                    // objectiterator.finishedIterating();
+                    // ShreddedPaper end - use this.level.getAllEntities()
                     _pufferfish_spawnCountsReady.set(true);
                 });
             }
diff --git a/src/main/java/net/minecraft/server/level/ServerLevel.java b/src/main/java/net/minecraft/server/level/ServerLevel.java
index c9fb25304a19a12d24da44b752d8be3b078a59d5..123d6a30e73614be92be819fc146cb87c3a49007 100644
--- a/src/main/java/net/minecraft/server/level/ServerLevel.java
+++ b/src/main/java/net/minecraft/server/level/ServerLevel.java
@@ -193,7 +193,7 @@ public class ServerLevel extends Level implements WorldGenLevel {
     private final MinecraftServer server;
     public final PrimaryLevelData serverLevelData; // CraftBukkit - type
     private int lastSpawnChunkRadius;
-    final EntityTickList entityTickList;
+    // final EntityTickList entityTickList; // ShreddedPaper - moved into each region
     //public final PersistentEntitySectionManager<Entity> entityManager; // Paper - rewrite chunk system
     private final GameEventDispatcher gameEventDispatcher;
     public boolean noSave;
@@ -703,7 +703,7 @@ public class ServerLevel extends Level implements WorldGenLevel {
         this.uuid = WorldUUID.getUUID(convertable_conversionsession.levelDirectory.path().toFile());
         // CraftBukkit end
         this.players = Lists.newArrayList();
-        this.entityTickList = new EntityTickList();
+        // this.entityTickList = new EntityTickList(); // ShreddedPaper - moved into each region
         this.blockTicks = new LevelTicks<>(this::isPositionTickingWithEntitiesLoaded, this.getProfilerSupplier());
         this.fluidTicks = new LevelTicks<>(this::isPositionTickingWithEntitiesLoaded, this.getProfilerSupplier());
         this.pathTypesByPosCache = new PathTypeCache();
@@ -917,33 +917,32 @@ public class ServerLevel extends Level implements WorldGenLevel {
 
             org.spigotmc.ActivationRange.activateEntities(this); // Spigot
             //this.timings.entityTick.startTiming(); // Spigot // Purpur
-            this.entityTickList.forEach((entity) -> {
-                entity.activatedPriorityReset = false; // Pufferfish - DAB
-                if (!entity.isRemoved()) {
-                    if (false && this.shouldDiscardEntity(entity)) { // CraftBukkit - We prevent spawning in general, so this butchering is not needed
-                        entity.discard();
-                    } else if (!tickratemanager.isEntityFrozen(entity)) {
-                        //gameprofilerfiller.push("checkDespawn"); // Purpur
-                        entity.checkDespawn();
-                        //gameprofilerfiller.pop(); // Purpur
-                        if (true || this.chunkSource.chunkMap.getDistanceManager().inEntityTickingRange(entity.chunkPosition().toLong())) { // Paper - now always true if in the ticking list
-                            Entity entity1 = entity.getVehicle();
-
-                            if (entity1 != null) {
-                                if (!entity1.isRemoved() && entity1.hasPassenger(entity)) {
-                                    return;
-                                }
-
-                                entity.stopRiding();
-                            }
-
-                            //gameprofilerfiller.push("tick"); // Purpur
-                            this.guardEntityTick(this::tickNonPassenger, entity);
-                            //gameprofilerfiller.pop(); // Purpur
-                        }
-                    }
-                }
-            });
+//            this.entityTickList.forEach((entity) -> {
+//                if (!entity.isRemoved()) {
+//                    if (false && this.shouldDiscardEntity(entity)) { // CraftBukkit - We prevent spawning in general, so this butchering is not needed
+//                        entity.discard();
+//                    } else if (!tickratemanager.isEntityFrozen(entity)) {
+//                        //gameprofilerfiller.push("checkDespawn"); // Purpur
+//                        entity.checkDespawn();
+//                        //gameprofilerfiller.pop(); // Purpur
+//                        if (true || this.chunkSource.chunkMap.getDistanceManager().inEntityTickingRange(entity.chunkPosition().toLong())) { // Paper - now always true if in the ticking list
+//                            Entity entity1 = entity.getVehicle();
+//
+//                            if (entity1 != null) {
+//                                if (!entity1.isRemoved() && entity1.hasPassenger(entity)) {
+//                                    return;
+//                                }
+//
+//                                entity.stopRiding();
+//                            }
+//
+//                            //gameprofilerfiller.push("tick"); // Purpur
+//                            this.guardEntityTick(this::tickNonPassenger, entity);
+//                            //gameprofilerfiller.pop(); // Purpur
+//                        }
+//                    }
+//                }
+//            });
             //this.timings.entityTick.stopTiming(); // Spigot // Purpur
             //this.timings.tickEntities.stopTiming(); // Spigot // Purpur
             //gameprofilerfiller.pop(); // Purpur
@@ -1012,7 +1011,7 @@ public class ServerLevel extends Level implements WorldGenLevel {
 
     }
 
-    private boolean shouldDiscardEntity(Entity entity) {
+    public boolean shouldDiscardEntity(Entity entity) { // ShreddedPaper - make public
         return !this.server.isSpawningAnimals() && (entity instanceof Animal || entity instanceof WaterAnimal) ? true : !this.server.areNpcsEnabled() && entity instanceof Npc;
     }
 
@@ -1502,7 +1501,7 @@ public class ServerLevel extends Level implements WorldGenLevel {
 
     private void tickPassenger(Entity vehicle, Entity passenger) {
         if (!passenger.isRemoved() && passenger.getVehicle() == vehicle) {
-            if (passenger instanceof Player || this.entityTickList.contains(passenger)) {
+            if (passenger instanceof Player || passenger.isTicking()) { // ShreddedPaper - use a different method to check if passenger is ticking
                 // Paper - EAR 2
                 final boolean isActive = org.spigotmc.ActivationRange.checkIfActive(passenger);
                 //co.aikar.timings.Timing timer = isActive ? passenger.getType().passengerTickTimer.startTiming() : passenger.getType().passengerInactiveTickTimer.startTiming(); // Paper // Purpur
@@ -2645,7 +2644,7 @@ public class ServerLevel extends Level implements WorldGenLevel {
 
     @Override
     public LevelEntityGetter<Entity> getEntities() {
-        org.spigotmc.AsyncCatcher.catchOp("Chunk getEntities call"); // Spigot
+        // org.spigotmc.AsyncCatcher.catchOp("Chunk getEntities call"); // Spigot // ShreddedPaper - thread safe now
         return this.entityLookup; // Paper - rewrite chunk system
     }
 
@@ -2766,11 +2765,11 @@ public class ServerLevel extends Level implements WorldGenLevel {
 
         public void onTickingStart(Entity entity) {
             if (entity instanceof net.minecraft.world.entity.Marker && !paperConfig().entities.markers.tick) return; // Paper - Configurable marker ticking
-            ServerLevel.this.entityTickList.add(entity);
+            ServerLevel.this.getChunkSource().tickingRegions.addTickingEntity(entity); // ShreddedPaper - move entityTickList to the regions
         }
 
         public void onTickingEnd(Entity entity) {
-            ServerLevel.this.entityTickList.remove(entity);
+            ServerLevel.this.getChunkSource().tickingRegions.removeTickingEntity(entity); // ShreddedPaper - move entityTickList to the regions
             // Paper start - Reset pearls when they stop being ticked
             if (paperConfig().fixes.disableUnloadedChunkEnderpearlExploit && entity instanceof net.minecraft.world.entity.projectile.ThrownEnderpearl pearl) {
                 pearl.cachedOwner = null;
@@ -2903,6 +2902,7 @@ public class ServerLevel extends Level implements WorldGenLevel {
 
         public void onSectionChange(Entity entity) {
             entity.updateDynamicGameEventListener(DynamicGameEventListener::move);
+            ServerLevel.this.chunkSource.tickingRegions.moveTickingEntity(entity); // ShreddedPaper - move ticking entity into the new region
         }
     }
 
diff --git a/src/main/java/net/minecraft/world/entity/Entity.java b/src/main/java/net/minecraft/world/entity/Entity.java
index 579ee25fc358bb6c5d4129e12036b66c35ac2d9b..9db7f7604af0a4f82867cfb63e4c3bb785d829ec 100644
--- a/src/main/java/net/minecraft/world/entity/Entity.java
+++ b/src/main/java/net/minecraft/world/entity/Entity.java
@@ -339,6 +339,7 @@ public abstract class Entity implements SyncedDataHolder, Nameable, EntityAccess
     public double xOld;
     public double yOld;
     public double zOld;
+    public @Nullable ChunkPos previousTickingChunkPosRegion; // ShreddedPaper
     public float maxUpStep; // Purpur
     public boolean noPhysics;
     public final RandomSource random;
diff --git a/src/main/java/net/minecraft/world/level/Level.java b/src/main/java/net/minecraft/world/level/Level.java
index adf6379f1ebd7ead6368d7b58f60bf793ab43be6..fb1ec5a87639402ac008a862e083a8261eac614f 100644
--- a/src/main/java/net/minecraft/world/level/Level.java
+++ b/src/main/java/net/minecraft/world/level/Level.java
@@ -1346,7 +1346,7 @@ public abstract class Level implements LevelAccessor, AutoCloseable {
     public <T extends Entity> void guardEntityTick(Consumer<T> tickConsumer, T entity) {
         try {
             tickConsumer.accept(entity);
-            MinecraftServer.getServer().executeMidTickTasks(); // Paper - execute chunk tasks mid tick
+            // MinecraftServer.getServer().executeMidTickTasks(); // Paper - execute chunk tasks mid tick // ShreddedPaper - disable
         } catch (Throwable throwable) {
             if (throwable instanceof ThreadDeath) throw throwable; // Paper
             // Paper start - Prevent block entity and entity crashes
