From 0000000000000000000000000000000000000000 Mon Sep 17 00:00:00 2001
From: PureGero <puregero@gmail.com>
Date: Sun, 5 May 2024 13:59:19 +0900
Subject: [PATCH] Multithread entity ticking


diff --git a/src/main/java/io/multipaper/shreddedpaper/region/LevelChunkRegion.java b/src/main/java/io/multipaper/shreddedpaper/region/LevelChunkRegion.java
index 51ce3a1324793f3560cd718d5aa37b33ebbaafa0..69cb88e9ca80b0c6a051f9eefadfe6a6a952d31f 100644
--- a/src/main/java/io/multipaper/shreddedpaper/region/LevelChunkRegion.java
+++ b/src/main/java/io/multipaper/shreddedpaper/region/LevelChunkRegion.java
@@ -1,6 +1,8 @@
 package io.multipaper.shreddedpaper.region;
 
+import io.papermc.paper.util.maplist.IteratorSafeOrderedReferenceSet;
 import net.minecraft.server.level.ServerLevel;
+import net.minecraft.world.entity.Entity;
 import net.minecraft.world.level.chunk.LevelChunk;
 
 import java.util.ArrayList;
@@ -12,6 +14,7 @@ public class LevelChunkRegion {
     private final ServerLevel level;
     private final RegionPos regionPos;
     private final List<LevelChunk> levelChunks = new ArrayList<>(RegionPos.REGION_SIZE * RegionPos.REGION_SIZE);
+    private final IteratorSafeOrderedReferenceSet<Entity> tickingEntities = new IteratorSafeOrderedReferenceSet<>(); // Use IteratorSafeOrderedReferenceSet to maintain entity tick order
 
     public LevelChunkRegion(ServerLevel level, RegionPos regionPos) {
         this.level = level;
@@ -28,6 +31,25 @@ public class LevelChunkRegion {
         }
     }
 
+    public synchronized void addTickingEntity(Entity entity) {
+        tickingEntities.add(entity);
+    }
+
+    public synchronized void removeTickingEntity(Entity entity) {
+        tickingEntities.remove(entity);
+    }
+
+    public synchronized void forEachTickingEntity(Consumer<Entity> action) {
+        IteratorSafeOrderedReferenceSet.Iterator<Entity> iterator = this.tickingEntities.iterator();
+        try {
+            while (iterator.hasNext()) {
+                action.accept(iterator.next());
+            }
+        } finally {
+            iterator.finishedIterating();
+        }
+    }
+
     public ServerLevel getLevel() {
         return level;
     }
diff --git a/src/main/java/io/multipaper/shreddedpaper/region/LevelChunkRegionMap.java b/src/main/java/io/multipaper/shreddedpaper/region/LevelChunkRegionMap.java
index 89724b9d82813cc5d42eb1a0a3d4bf50e406a3ca..7a1e038d13c42265ec2c8c5beab9fea1631f7686 100644
--- a/src/main/java/io/multipaper/shreddedpaper/region/LevelChunkRegionMap.java
+++ b/src/main/java/io/multipaper/shreddedpaper/region/LevelChunkRegionMap.java
@@ -2,6 +2,7 @@ package io.multipaper.shreddedpaper.region;
 
 import it.unimi.dsi.fastutil.longs.Long2ObjectOpenHashMap;
 import net.minecraft.server.level.ServerLevel;
+import net.minecraft.world.entity.Entity;
 import net.minecraft.world.level.chunk.LevelChunk;
 import io.multipaper.shreddedpaper.util.SimpleStampedLock;
 
@@ -57,4 +58,24 @@ public class LevelChunkRegionMap {
         regionsCopy.forEach(consumer);
     }
 
+    public void addTickingEntity(Entity entity) {
+        entity.previousTickingChunkPosRegion = entity.chunkPosition();
+        getOrCreate(RegionPos.forChunk(entity.chunkPosition())).addTickingEntity(entity);
+    }
+
+    public void removeTickingEntity(Entity entity) {
+        getOrCreate(RegionPos.forChunk(entity.chunkPosition())).removeTickingEntity(entity);
+    }
+
+    public void moveTickingEntity(Entity entity) {
+        RegionPos fromRegion = RegionPos.forChunk(entity.previousTickingChunkPosRegion);
+        RegionPos toRegion = RegionPos.forChunk(entity.chunkPosition());
+
+        if (!fromRegion.equals(toRegion)) {
+            entity.previousTickingChunkPosRegion = entity.chunkPosition();
+            getOrCreate(fromRegion).removeTickingEntity(entity);
+            getOrCreate(toRegion).addTickingEntity(entity);
+        }
+    }
+
 }
\ No newline at end of file
diff --git a/src/main/java/io/multipaper/shreddedpaper/threading/ShreddedPaperChunkTicker.java b/src/main/java/io/multipaper/shreddedpaper/threading/ShreddedPaperChunkTicker.java
index dfbdfca7147e4463507be08a36de63aa96238ba4..56bd2dea24afbda142602fd0c352feedc6f431dd 100644
--- a/src/main/java/io/multipaper/shreddedpaper/threading/ShreddedPaperChunkTicker.java
+++ b/src/main/java/io/multipaper/shreddedpaper/threading/ShreddedPaperChunkTicker.java
@@ -59,6 +59,8 @@ public class ShreddedPaperChunkTicker {
 
             region.forEach(chunk -> _tickChunk(level, chunk, spawnercreature_d));
 
+            region.forEachTickingEntity(ShreddedPaperEntityTicker::tickEntity);
+
             if (region.isEmpty()) {
                 level.chunkSource.tickingRegions.remove(region.getRegionPos());
             }
diff --git a/src/main/java/io/multipaper/shreddedpaper/threading/ShreddedPaperEntityTicker.java b/src/main/java/io/multipaper/shreddedpaper/threading/ShreddedPaperEntityTicker.java
new file mode 100644
index 0000000000000000000000000000000000000000..81d6ae082047fdda6f2271a830e700fa90b9d0ac
--- /dev/null
+++ b/src/main/java/io/multipaper/shreddedpaper/threading/ShreddedPaperEntityTicker.java
@@ -0,0 +1,37 @@
+package io.multipaper.shreddedpaper.threading;
+
+import net.minecraft.server.level.ServerLevel;
+import net.minecraft.world.entity.Entity;
+
+public class ShreddedPaperEntityTicker {
+
+    public static void tickEntity(Entity entity) {
+        ServerLevel level = (ServerLevel) entity.level();
+
+        if (!entity.isRemoved()) {
+            if (false && level.shouldDiscardEntity(entity)) { // CraftBukkit - We prevent spawning in general, so this butchering is not needed
+                entity.discard();
+            } else if (!level.tickRateManager().isEntityFrozen(entity)) {
+                //gameprofilerfiller.push("checkDespawn"); // Purpur
+                entity.checkDespawn();
+                //gameprofilerfiller.pop(); // Purpur
+                if (true || level.chunkSource.chunkMap.getDistanceManager().inEntityTickingRange(entity.chunkPosition().toLong())) { // Paper - now always true if in the ticking list
+                    Entity entity1 = entity.getVehicle();
+
+                    if (entity1 != null) {
+                        if (!entity1.isRemoved() && entity1.hasPassenger(entity)) {
+                            return;
+                        }
+
+                        entity.stopRiding();
+                    }
+
+                    //gameprofilerfiller.push("tick"); // Purpur
+                    level.guardEntityTick(level::tickNonPassenger, entity);
+                    //gameprofilerfiller.pop(); // Purpur
+                }
+            }
+        }
+    }
+
+}
diff --git a/src/main/java/net/minecraft/server/level/ServerChunkCache.java b/src/main/java/net/minecraft/server/level/ServerChunkCache.java
index 428994dda9fb445ab1bc0e261a7b8ad62c9adcd3..b17d31aa52ade535ebf572141e6dacc92dab5358 100644
--- a/src/main/java/net/minecraft/server/level/ServerChunkCache.java
+++ b/src/main/java/net/minecraft/server/level/ServerChunkCache.java
@@ -710,12 +710,14 @@ public class ServerChunkCache extends ChunkSource {
             if (_pufferfish_spawnCountsReady.getAndSet(false)) {
                 net.minecraft.server.MinecraftServer.getServer().mobSpawnExecutor.submit(() -> {
                     int mapped = distanceManager.getNaturalSpawnChunkCount();
-                    io.papermc.paper.util.maplist.IteratorSafeOrderedReferenceSet.Iterator<Entity> objectiterator =
-                            level.entityTickList.entities.iterator(io.papermc.paper.util.maplist.IteratorSafeOrderedReferenceSet.ITERATOR_FLAG_SEE_ADDITIONS);
-                    gg.pufferfish.pufferfish.util.IterableWrapper<Entity> wrappedIterator =
-                            new gg.pufferfish.pufferfish.util.IterableWrapper<>(objectiterator);
-                    lastSpawnState = NaturalSpawner.createState(mapped, wrappedIterator, this::getFullChunk, null, true);
-                    objectiterator.finishedIterating();
+                    // ShreddedPaper start - use this.level.getAllEntities()
+                    // io.papermc.paper.util.maplist.IteratorSafeOrderedReferenceSet.Iterator<Entity> objectiterator =
+                    //         level.entityTickList.entities.iterator(io.papermc.paper.util.maplist.IteratorSafeOrderedReferenceSet.ITERATOR_FLAG_SEE_ADDITIONS);
+                    // gg.pufferfish.pufferfish.util.IterableWrapper<Entity> wrappedIterator =
+                    //         new gg.pufferfish.pufferfish.util.IterableWrapper<>(objectiterator);
+                    lastSpawnState = NaturalSpawner.createState(mapped, this.level.getAllEntities(), this::getFullChunk, null, true);
+                    // objectiterator.finishedIterating();
+                    // ShreddedPaper end - use this.level.getAllEntities()
                     _pufferfish_spawnCountsReady.set(true);
                 });
             }
diff --git a/src/main/java/net/minecraft/server/level/ServerLevel.java b/src/main/java/net/minecraft/server/level/ServerLevel.java
index 2ff98096027d1398c72591772af5ee7cd837982a..9c6f9cc702b33b58bf2007bcfbc3cf38933244b3 100644
--- a/src/main/java/net/minecraft/server/level/ServerLevel.java
+++ b/src/main/java/net/minecraft/server/level/ServerLevel.java
@@ -192,7 +192,7 @@ public class ServerLevel extends Level implements WorldGenLevel {
     private final MinecraftServer server;
     public final PrimaryLevelData serverLevelData; // CraftBukkit - type
     private int lastSpawnChunkRadius;
-    final EntityTickList entityTickList;
+    // final EntityTickList entityTickList; // ShreddedPaper - moved into each region
     //public final PersistentEntitySectionManager<Entity> entityManager; // Paper - rewrite chunk system
     private final GameEventDispatcher gameEventDispatcher;
     public boolean noSave;
@@ -702,7 +702,7 @@ public class ServerLevel extends Level implements WorldGenLevel {
         this.uuid = WorldUUID.getUUID(convertable_conversionsession.levelDirectory.path().toFile());
         // CraftBukkit end
         this.players = Lists.newArrayList();
-        this.entityTickList = new EntityTickList();
+        // this.entityTickList = new EntityTickList(); // ShreddedPaper - moved into each region
         this.blockTicks = new LevelTicks<>(this::isPositionTickingWithEntitiesLoaded, this.getProfilerSupplier());
         this.fluidTicks = new LevelTicks<>(this::isPositionTickingWithEntitiesLoaded, this.getProfilerSupplier());
         this.pathTypesByPosCache = new PathTypeCache();
@@ -916,32 +916,32 @@ public class ServerLevel extends Level implements WorldGenLevel {
 
             org.spigotmc.ActivationRange.activateEntities(this); // Spigot
             //this.timings.entityTick.startTiming(); // Spigot // Purpur
-            this.entityTickList.forEach((entity) -> {
-                if (!entity.isRemoved()) {
-                    if (false && this.shouldDiscardEntity(entity)) { // CraftBukkit - We prevent spawning in general, so this butchering is not needed
-                        entity.discard();
-                    } else if (!tickratemanager.isEntityFrozen(entity)) {
-                        //gameprofilerfiller.push("checkDespawn"); // Purpur
-                        entity.checkDespawn();
-                        //gameprofilerfiller.pop(); // Purpur
-                        if (true || this.chunkSource.chunkMap.getDistanceManager().inEntityTickingRange(entity.chunkPosition().toLong())) { // Paper - now always true if in the ticking list
-                            Entity entity1 = entity.getVehicle();
-
-                            if (entity1 != null) {
-                                if (!entity1.isRemoved() && entity1.hasPassenger(entity)) {
-                                    return;
-                                }
-
-                                entity.stopRiding();
-                            }
-
-                            //gameprofilerfiller.push("tick"); // Purpur
-                            this.guardEntityTick(this::tickNonPassenger, entity);
-                            //gameprofilerfiller.pop(); // Purpur
-                        }
-                    }
-                }
-            });
+//            this.entityTickList.forEach((entity) -> {
+//                if (!entity.isRemoved()) {
+//                    if (false && this.shouldDiscardEntity(entity)) { // CraftBukkit - We prevent spawning in general, so this butchering is not needed
+//                        entity.discard();
+//                    } else if (!tickratemanager.isEntityFrozen(entity)) {
+//                        //gameprofilerfiller.push("checkDespawn"); // Purpur
+//                        entity.checkDespawn();
+//                        //gameprofilerfiller.pop(); // Purpur
+//                        if (true || this.chunkSource.chunkMap.getDistanceManager().inEntityTickingRange(entity.chunkPosition().toLong())) { // Paper - now always true if in the ticking list
+//                            Entity entity1 = entity.getVehicle();
+//
+//                            if (entity1 != null) {
+//                                if (!entity1.isRemoved() && entity1.hasPassenger(entity)) {
+//                                    return;
+//                                }
+//
+//                                entity.stopRiding();
+//                            }
+//
+//                            //gameprofilerfiller.push("tick"); // Purpur
+//                            this.guardEntityTick(this::tickNonPassenger, entity);
+//                            //gameprofilerfiller.pop(); // Purpur
+//                        }
+//                    }
+//                }
+//            });
             //this.timings.entityTick.stopTiming(); // Spigot // Purpur
             //this.timings.tickEntities.stopTiming(); // Spigot // Purpur
             //gameprofilerfiller.pop(); // Purpur
@@ -1009,7 +1009,7 @@ public class ServerLevel extends Level implements WorldGenLevel {
 
     }
 
-    private boolean shouldDiscardEntity(Entity entity) {
+    public boolean shouldDiscardEntity(Entity entity) { // ShreddedPaper - make public
         return !this.server.isSpawningAnimals() && (entity instanceof Animal || entity instanceof WaterAnimal) ? true : !this.server.areNpcsEnabled() && entity instanceof Npc;
     }
 
@@ -1499,7 +1499,7 @@ public class ServerLevel extends Level implements WorldGenLevel {
 
     private void tickPassenger(Entity vehicle, Entity passenger) {
         if (!passenger.isRemoved() && passenger.getVehicle() == vehicle) {
-            if (passenger instanceof Player || this.entityTickList.contains(passenger)) {
+            if (passenger instanceof Player || passenger.isTicking()) { // ShreddedPaper - use a different method to check if passenger is ticking
                 // Paper - EAR 2
                 final boolean isActive = org.spigotmc.ActivationRange.checkIfActive(passenger);
                 //co.aikar.timings.Timing timer = isActive ? passenger.getType().passengerTickTimer.startTiming() : passenger.getType().passengerInactiveTickTimer.startTiming(); // Paper // Purpur
@@ -2642,7 +2642,7 @@ public class ServerLevel extends Level implements WorldGenLevel {
 
     @Override
     public LevelEntityGetter<Entity> getEntities() {
-        org.spigotmc.AsyncCatcher.catchOp("Chunk getEntities call"); // Spigot
+        // org.spigotmc.AsyncCatcher.catchOp("Chunk getEntities call"); // Spigot // ShreddedPaper - thread safe now
         return this.entityLookup; // Paper - rewrite chunk system
     }
 
@@ -2763,11 +2763,11 @@ public class ServerLevel extends Level implements WorldGenLevel {
 
         public void onTickingStart(Entity entity) {
             if (entity instanceof net.minecraft.world.entity.Marker && !paperConfig().entities.markers.tick) return; // Paper - Configurable marker ticking
-            ServerLevel.this.entityTickList.add(entity);
+            ServerLevel.this.getChunkSource().tickingRegions.addTickingEntity(entity); // ShreddedPaper - move entityTickList to the regions
         }
 
         public void onTickingEnd(Entity entity) {
-            ServerLevel.this.entityTickList.remove(entity);
+            ServerLevel.this.getChunkSource().tickingRegions.removeTickingEntity(entity); // ShreddedPaper - move entityTickList to the regions
             // Paper start - Reset pearls when they stop being ticked
             if (paperConfig().fixes.disableUnloadedChunkEnderpearlExploit && entity instanceof net.minecraft.world.entity.projectile.ThrownEnderpearl pearl) {
                 pearl.cachedOwner = null;
@@ -2900,6 +2900,7 @@ public class ServerLevel extends Level implements WorldGenLevel {
 
         public void onSectionChange(Entity entity) {
             entity.updateDynamicGameEventListener(DynamicGameEventListener::move);
+            ServerLevel.this.chunkSource.tickingRegions.moveTickingEntity(entity); // ShreddedPaper - move ticking entity into the new region
         }
     }
 
diff --git a/src/main/java/net/minecraft/world/entity/Entity.java b/src/main/java/net/minecraft/world/entity/Entity.java
index 7c8a3a6af8d5fdbb2341fe4012686bfc4f79a312..459ed557262d2fed7cc49a7e68878cfe7ae73786 100644
--- a/src/main/java/net/minecraft/world/entity/Entity.java
+++ b/src/main/java/net/minecraft/world/entity/Entity.java
@@ -339,6 +339,7 @@ public abstract class Entity implements SyncedDataHolder, Nameable, EntityAccess
     public double xOld;
     public double yOld;
     public double zOld;
+    public ChunkPos previousTickingChunkPosRegion; // ShreddedPaper
     public float maxUpStep; // Purpur
     public boolean noPhysics;
     public final RandomSource random;
diff --git a/src/main/java/net/minecraft/world/level/Level.java b/src/main/java/net/minecraft/world/level/Level.java
index f2e988b6fe9cc62c3fc167afac4a41a2e3fc8d75..28f4e0f7ee069f891b8ff1124902d4154ab15138 100644
--- a/src/main/java/net/minecraft/world/level/Level.java
+++ b/src/main/java/net/minecraft/world/level/Level.java
@@ -1346,7 +1346,7 @@ public abstract class Level implements LevelAccessor, AutoCloseable {
     public <T extends Entity> void guardEntityTick(Consumer<T> tickConsumer, T entity) {
         try {
             tickConsumer.accept(entity);
-            MinecraftServer.getServer().executeMidTickTasks(); // Paper - execute chunk tasks mid tick
+            // MinecraftServer.getServer().executeMidTickTasks(); // Paper - execute chunk tasks mid tick // ShreddedPaper - disable
         } catch (Throwable throwable) {
             if (throwable instanceof ThreadDeath) throw throwable; // Paper
             // Paper start - Prevent block entity and entity crashes
