From 0000000000000000000000000000000000000000 Mon Sep 17 00:00:00 2001
From: PureGero <puregero@gmail.com>
Date: Fri, 17 May 2024 00:26:43 +0900
Subject: [PATCH] Level ticks


diff --git a/src/main/java/io/multipaper/shreddedpaper/region/LevelTicksRegionProxy.java b/src/main/java/io/multipaper/shreddedpaper/region/LevelTicksRegionProxy.java
new file mode 100644
index 0000000000000000000000000000000000000000..051e74363aea1aff0a9d3da4300fca9002369fc3
--- /dev/null
+++ b/src/main/java/io/multipaper/shreddedpaper/region/LevelTicksRegionProxy.java
@@ -0,0 +1,114 @@
+package io.multipaper.shreddedpaper.region;
+
+import com.mojang.logging.LogUtils;
+import it.unimi.dsi.fastutil.longs.Long2ObjectMap;
+import it.unimi.dsi.fastutil.longs.Long2ObjectOpenHashMap;
+import net.minecraft.core.BlockPos;
+import net.minecraft.core.Vec3i;
+import net.minecraft.util.profiling.ProfilerFiller;
+import net.minecraft.world.level.ChunkPos;
+import net.minecraft.world.level.levelgen.structure.BoundingBox;
+import net.minecraft.world.ticks.LevelChunkTicks;
+import net.minecraft.world.ticks.LevelTicks;
+import net.minecraft.world.ticks.ScheduledTick;
+import org.slf4j.Logger;
+import io.multipaper.shreddedpaper.util.SimpleStampedLock;
+
+import java.util.Optional;
+import java.util.function.BiConsumer;
+import java.util.function.LongPredicate;
+import java.util.function.Supplier;
+
+public class LevelTicksRegionProxy<T> extends LevelTicks<T> {
+
+    private static final Logger LOGGER = LogUtils.getClassLogger();
+
+    private final LongPredicate tickingFutureReadyPredicate;
+    private final Supplier<ProfilerFiller> profilerGetter;
+    private final Long2ObjectMap<LevelTicks<T>> regions = new Long2ObjectOpenHashMap<>();
+    private final SimpleStampedLock regionsLock = new SimpleStampedLock();
+
+    public LevelTicksRegionProxy(LongPredicate tickingFutureReadyPredicate, Supplier<ProfilerFiller> profilerGetter) {
+        super(tickingFutureReadyPredicate, profilerGetter);
+        this.tickingFutureReadyPredicate = tickingFutureReadyPredicate;
+        this.profilerGetter = profilerGetter;
+    }
+
+    private LevelTicks<T> createRegionLevelTicks() {
+        return new LevelTicks<>(tickingFutureReadyPredicate, profilerGetter);
+    }
+
+    public Optional<LevelTicks<T>> get(BlockPos pos) {
+        return get(new ChunkPos(pos));
+    }
+
+    public Optional<LevelTicks<T>> get(ChunkPos pos) {
+        return get(RegionPos.forChunk(pos));
+    }
+
+    public Optional<LevelTicks<T>> get(RegionPos pos) {
+        return Optional.ofNullable(regionsLock.optimisticRead(() -> regions.get(pos.longKey)));
+    }
+
+    public void addContainer(ChunkPos pos, LevelChunkTicks<T> scheduler) {
+        get(pos).orElseGet(() -> {
+            return regionsLock.write(() -> regions.computeIfAbsent(RegionPos.forChunk(pos).longKey, k -> createRegionLevelTicks()));
+        }).addContainer(pos, scheduler);
+    }
+
+    public void removeContainer(ChunkPos pos) {
+        get(pos).ifPresent(v -> {
+            v.removeContainer(pos);
+
+            if (v.isEmpty()) {
+                regionsLock.write(() -> regions.remove(RegionPos.forChunk(pos).longKey));
+            }
+        });
+    }
+
+    @Override
+    public void schedule(ScheduledTick<T> orderedTick) {
+        get(orderedTick.pos()).orElseThrow(() -> new IllegalArgumentException("Chunk not loaded: " + orderedTick.pos())).schedule(orderedTick);
+    }
+
+    public void tick(RegionPos regionPos, long time, int maxTicks, BiConsumer<BlockPos, T> ticker) {
+        get(regionPos).ifPresent(v -> {
+            v.tick(time, maxTicks, ticker);
+        });
+    }
+
+    @Override
+    public void tick(long time, int maxTicks, BiConsumer<BlockPos, T> ticker) {
+        // Do nothing
+    }
+
+    @Override
+    public boolean hasScheduledTick(BlockPos pos, T type) {
+        return get(pos).map(v -> v.hasScheduledTick(pos, type)).orElse(false);
+    }
+
+    @Override
+    public boolean willTickThisTick(BlockPos pos, T type) {
+        return get(pos).map(v -> v.willTickThisTick(pos, type)).orElse(false);
+    }
+
+    @Override
+    public void clearArea(BoundingBox box) {
+        // Surely no one will miss this
+    }
+
+    @Override
+    public void copyArea(BoundingBox box, Vec3i offset) {
+        // Surely no one will miss this
+    }
+
+    @Override
+    public void copyAreaFrom(LevelTicks<T> scheduler, BoundingBox box, Vec3i offset) {
+        // Surely no one will miss this
+    }
+
+    @Override
+    public int count() {
+        return -1;
+    }
+}
diff --git a/src/main/java/io/multipaper/shreddedpaper/threading/ShreddedPaperChunkTicker.java b/src/main/java/io/multipaper/shreddedpaper/threading/ShreddedPaperChunkTicker.java
index 59c5a855c4ea4a990076559dd247e05cfb24d523..a51db4e5505903376656aad80971f8faabc0cc53 100644
--- a/src/main/java/io/multipaper/shreddedpaper/threading/ShreddedPaperChunkTicker.java
+++ b/src/main/java/io/multipaper/shreddedpaper/threading/ShreddedPaperChunkTicker.java
@@ -66,6 +66,10 @@ public class ShreddedPaperChunkTicker {
             });
 
             region.tickTasks();
+
+            level.blockTicks.tick(region.getRegionPos(), level.getGameTime(), level.paperConfig().environment.maxBlockTicks, level::tickBlock);
+            level.fluidTicks.tick(region.getRegionPos(), level.getGameTime(), level.paperConfig().environment.maxBlockTicks, level::tickFluid);
+
             region.forEachTickingChunk(chunk -> _tickChunk(level, chunk, spawnercreature_d));
 
             region.forEachTickingEntity(ShreddedPaperEntityTicker::tickEntity);
diff --git a/src/main/java/net/minecraft/server/level/ServerLevel.java b/src/main/java/net/minecraft/server/level/ServerLevel.java
index 8176fe4cb0d2f4cbb9bc9b8cbedec2aa200a39af..7402eeaa02be21a8b441812e2db6b398f9ba5974 100644
--- a/src/main/java/net/minecraft/server/level/ServerLevel.java
+++ b/src/main/java/net/minecraft/server/level/ServerLevel.java
@@ -1,5 +1,6 @@
 package net.minecraft.server.level;
 
+import ca.spottedleaf.moonrise.common.util.TickThread;
 import com.google.common.annotations.VisibleForTesting;
 import co.aikar.timings.TimingHistory; // Paper
 import com.google.common.collect.Lists;
@@ -176,6 +177,7 @@ import org.bukkit.event.entity.CreatureSpawnEvent;
 import org.bukkit.event.server.MapInitializeEvent;
 import org.bukkit.event.weather.LightningStrikeEvent;
 import org.bukkit.event.world.TimeSkipEvent;
+import io.multipaper.shreddedpaper.region.LevelTicksRegionProxy;
 import io.multipaper.shreddedpaper.region.RegionPos;
 import io.multipaper.shreddedpaper.threading.ShreddedPaperRegionScheduler;
 // CraftBukkit end
@@ -202,8 +204,8 @@ public class ServerLevel extends Level implements WorldGenLevel, ca.spottedleaf.
     private final SleepStatus sleepStatus;
     private int emptyTime;
     private final PortalForcer portalForcer;
-    private final LevelTicks<Block> blockTicks;
-    private final LevelTicks<Fluid> fluidTicks;
+    public final LevelTicksRegionProxy<Block> blockTicks; // ShreddedPaper
+    public final LevelTicksRegionProxy<Fluid> fluidTicks; // ShreddedPaper
     private final PathTypeCache pathTypesByPosCache;
     final Set<Mob> navigatingMobs;
     volatile boolean isUpdatingNavigations;
@@ -513,8 +515,8 @@ public class ServerLevel extends Level implements WorldGenLevel, ca.spottedleaf.
         // CraftBukkit end
         this.players = new CopyOnWriteArrayList<>(); // ShreddedPaper - thread-safe (players should not be changing worlds often, thus copy-on-write is sufficient)
         // this.entityTickList = new EntityTickList(); // ShreddedPaper - moved into each region
-        this.blockTicks = new LevelTicks<>(this::isPositionTickingWithEntitiesLoaded, this.getProfilerSupplier());
-        this.fluidTicks = new LevelTicks<>(this::isPositionTickingWithEntitiesLoaded, this.getProfilerSupplier());
+        this.blockTicks = new LevelTicksRegionProxy<>(this::isPositionTickingWithEntitiesLoaded, this.getProfilerSupplier()); // ShreddedPaper
+        this.fluidTicks = new LevelTicksRegionProxy<>(this::isPositionTickingWithEntitiesLoaded, this.getProfilerSupplier()); // ShreddedPaper
         this.pathTypesByPosCache = new PathTypeCache();
         this.navigatingMobs = new ObjectOpenHashSet();
         this.blockEvents = new ObjectLinkedOpenHashSet();
@@ -1255,7 +1257,8 @@ public class ServerLevel extends Level implements WorldGenLevel, ca.spottedleaf.
         this.emptyTime = 0;
     }
 
-    private void tickFluid(BlockPos pos, Fluid fluid) {
+    public void tickFluid(BlockPos pos, Fluid fluid) { // ShreddedPaper - make public
+        TickThread.ensureTickThread(this, pos, "Cannot tick fluid outside of tick thread"); // ShreddedPaper
         FluidState fluid1 = this.getFluidState(pos);
 
         if (fluid1.is(fluid)) {
@@ -1263,13 +1266,14 @@ public class ServerLevel extends Level implements WorldGenLevel, ca.spottedleaf.
         }
         // Paper start - rewrite chunk system
         if ((++this.tickedBlocksOrFluids & 7L) != 0L) {
-            ((ca.spottedleaf.moonrise.patches.chunk_system.server.ChunkSystemMinecraftServer)this.server).moonrise$executeMidTickTasks();
+        //     ((ca.spottedleaf.moonrise.patches.chunk_system.server.ChunkSystemMinecraftServer)this.server).moonrise$executeMidTickTasks(); // ShreddedPaper - unnecessary now
         }
         // Paper end - rewrite chunk system
 
     }
 
-    private void tickBlock(BlockPos pos, Block block) {
+    public void tickBlock(BlockPos pos, Block block) { // ShreddedPaper - make public
+        TickThread.ensureTickThread(this, pos, "Cannot tick block outside of tick thread"); // ShreddedPaper
         BlockState iblockdata = this.getBlockState(pos);
 
         if (iblockdata.is(block)) {
@@ -1277,7 +1281,7 @@ public class ServerLevel extends Level implements WorldGenLevel, ca.spottedleaf.
         }
         // Paper start - rewrite chunk system
         if ((++this.tickedBlocksOrFluids & 7L) != 0L) {
-            ((ca.spottedleaf.moonrise.patches.chunk_system.server.ChunkSystemMinecraftServer)this.server).moonrise$executeMidTickTasks();
+            // ((ca.spottedleaf.moonrise.patches.chunk_system.server.ChunkSystemMinecraftServer)this.server).moonrise$executeMidTickTasks(); // ShreddedPaper - unnecessary now
         }
         // Paper end - rewrite chunk system
 
diff --git a/src/main/java/net/minecraft/world/level/LevelAccessor.java b/src/main/java/net/minecraft/world/level/LevelAccessor.java
index 54d13eebc9b01e9d77f51011b7de95b80bc21669..585a2c51be2b3d78fcca2ed0a44babca1863d97d 100644
--- a/src/main/java/net/minecraft/world/level/LevelAccessor.java
+++ b/src/main/java/net/minecraft/world/level/LevelAccessor.java
@@ -1,6 +1,8 @@
 package net.minecraft.world.level;
 
 import javax.annotation.Nullable;
+
+import ca.spottedleaf.moonrise.common.util.TickThread;
 import net.minecraft.core.BlockPos;
 import net.minecraft.core.Direction;
 import net.minecraft.core.Holder;
@@ -8,6 +10,7 @@ import net.minecraft.core.particles.ParticleOptions;
 import net.minecraft.core.registries.Registries;
 import net.minecraft.resources.ResourceKey;
 import net.minecraft.server.MinecraftServer;
+import net.minecraft.server.level.ServerLevel;
 import net.minecraft.sounds.SoundEvent;
 import net.minecraft.sounds.SoundSource;
 import net.minecraft.util.RandomSource;
@@ -47,20 +50,24 @@ public interface LevelAccessor extends CommonLevelAccessor, LevelTimeAccess {
     }
 
     default void scheduleTick(BlockPos pos, Block block, int delay, TickPriority priority) {
+        if (this instanceof ServerLevel serverLevel) TickThread.ensureTickThread(serverLevel, pos, "Cannot schedule outside of tick thread"); // ShreddedPaper
         this.getBlockTicks().schedule(this.createTick(pos, block, delay, priority));
     }
 
     default void scheduleTick(BlockPos pos, Block block, int delay) {
+        if (this instanceof ServerLevel serverLevel) TickThread.ensureTickThread(serverLevel, pos, "Cannot schedule outside of tick thread"); // ShreddedPaper
         this.getBlockTicks().schedule(this.createTick(pos, block, delay));
     }
 
     LevelTickAccess<Fluid> getFluidTicks();
 
     default void scheduleTick(BlockPos pos, Fluid fluid, int delay, TickPriority priority) {
+        if (this instanceof ServerLevel serverLevel) TickThread.ensureTickThread(serverLevel, pos, "Cannot schedule outside of tick thread"); // ShreddedPaper
         this.getFluidTicks().schedule(this.createTick(pos, fluid, delay, priority));
     }
 
     default void scheduleTick(BlockPos pos, Fluid fluid, int delay) {
+        if (this instanceof ServerLevel serverLevel) TickThread.ensureTickThread(serverLevel, pos, "Cannot schedule outside of tick thread"); // ShreddedPaper
         this.getFluidTicks().schedule(this.createTick(pos, fluid, delay));
     }
 
diff --git a/src/main/java/net/minecraft/world/ticks/LevelTicks.java b/src/main/java/net/minecraft/world/ticks/LevelTicks.java
index a6d62abd3102770652f914b9d697c6d3c2533cfc..5d7aeaa8dc31906713e65808f675830df23cd0e4 100644
--- a/src/main/java/net/minecraft/world/ticks/LevelTicks.java
+++ b/src/main/java/net/minecraft/world/ticks/LevelTicks.java
@@ -282,4 +282,10 @@ public class LevelTicks<T> implements LevelTickAccess<T> {
     interface PosAndContainerConsumer<T> {
         void accept(long chunkPos, LevelChunkTicks<T> chunkTickScheduler);
     }
+
+    // ShreddedPaper start
+    public boolean isEmpty() {
+        return this.allContainers.isEmpty();
+    }
+    // ShreddedPaper end
 }
