From 0000000000000000000000000000000000000000 Mon Sep 17 00:00:00 2001
From: PureGero <puregero@gmail.com>
Date: Sun, 5 May 2024 17:02:23 +0900
Subject: [PATCH] Thread-safe navigatingMobs


diff --git a/src/main/java/io/multipaper/shreddedpaper/region/LevelChunkRegion.java b/src/main/java/io/multipaper/shreddedpaper/region/LevelChunkRegion.java
index 1ae8b6b9b0a63c1c93e9add5c5dfe4739476444d..8086978b2a48acf23bf3de1a4082efb4ad6b8f28 100644
--- a/src/main/java/io/multipaper/shreddedpaper/region/LevelChunkRegion.java
+++ b/src/main/java/io/multipaper/shreddedpaper/region/LevelChunkRegion.java
@@ -9,6 +9,7 @@ import it.unimi.dsi.fastutil.objects.ReferenceArrayList;
 import net.minecraft.server.level.ServerLevel;
 import net.minecraft.server.level.ServerPlayer;
 import net.minecraft.world.entity.Entity;
+import net.minecraft.world.entity.Mob;
 import net.minecraft.world.level.block.entity.TickingBlockEntity;
 import net.minecraft.world.level.chunk.LevelChunk;
 
@@ -31,6 +32,7 @@ public class LevelChunkRegion {
     public final LongLinkedOpenHashSet unloadQueue = new LongLinkedOpenHashSet();
     public final List<TickingBlockEntity> tickingBlockEntities = new ReferenceArrayList<>();
     public final List<TickingBlockEntity> pendingBlockEntityTickers = new ReferenceArrayList<>();
+    private final ObjectOpenHashSet<Mob> navigatingMobs = new ObjectOpenHashSet<>();
 
     public LevelChunkRegion(ServerLevel level, RegionPos regionPos) {
         this.level = level;
@@ -114,6 +116,18 @@ public class LevelChunkRegion {
         return this.players.isEmpty() ? List.of() : new ObjectArrayList<>(this.players);
     }
 
+    public synchronized void addNavigationMob(Mob mob) {
+        navigatingMobs.add(mob);
+    }
+
+    public synchronized void removeNavigationMob(Mob mob) {
+        navigatingMobs.remove(mob);
+    }
+
+    public synchronized void collectNavigatingMobs(List<Mob> collection) {
+        collection.addAll(this.navigatingMobs);
+    }
+
     public RegionPos getRegionPos() {
         return regionPos;
     }
@@ -157,6 +171,7 @@ public class LevelChunkRegion {
                 && tickingBlockEntities.isEmpty()
                 && pendingBlockEntityTickers.isEmpty()
                 && trackedEntities.isEmpty()
+                && navigatingMobs.isEmpty()
                 ;
     }
 }
diff --git a/src/main/java/io/multipaper/shreddedpaper/region/LevelChunkRegionMap.java b/src/main/java/io/multipaper/shreddedpaper/region/LevelChunkRegionMap.java
index 52c7ce154c4cef680143f635d99d162bdac569c4..80acf2a25921c7b674d0758c0c29a7352746726b 100644
--- a/src/main/java/io/multipaper/shreddedpaper/region/LevelChunkRegionMap.java
+++ b/src/main/java/io/multipaper/shreddedpaper/region/LevelChunkRegionMap.java
@@ -1,11 +1,14 @@
 package io.multipaper.shreddedpaper.region;
 
 import it.unimi.dsi.fastutil.longs.Long2ObjectOpenHashMap;
+import it.unimi.dsi.fastutil.objects.ObjectArrayList;
 import net.minecraft.server.level.ServerLevel;
 import net.minecraft.server.level.ServerPlayer;
 import net.minecraft.world.entity.Entity;
+import net.minecraft.world.entity.Mob;
 import net.minecraft.world.level.ChunkPos;
 import net.minecraft.world.level.chunk.LevelChunk;
+import io.multipaper.shreddedpaper.threading.ShreddedPaperRegionLocker;
 import io.multipaper.shreddedpaper.util.SimpleStampedLock;
 
 import java.util.ArrayList;
@@ -103,6 +106,10 @@ public class LevelChunkRegionMap {
 
         entity.previousTrackedChunkPosRegion = entity.chunkPosition();
         getOrCreate(RegionPos.forChunk(entity.chunkPosition())).addTrackedEntity(entity);
+
+        if (entity instanceof Mob mob) {
+            getOrCreate(RegionPos.forChunk(entity.chunkPosition())).addNavigationMob(mob);
+        }
     }
 
     public void removeTrackedEntity(Entity entity) {
@@ -110,6 +117,10 @@ public class LevelChunkRegionMap {
             throw new IllegalStateException("Entity is not being tracked " + entity);
         }
 
+        if (entity instanceof Mob mob) {
+            getOrCreate(RegionPos.forChunk(entity.chunkPosition())).removeNavigationMob(mob);
+        }
+
         getOrCreate(RegionPos.forChunk(entity.previousTrackedChunkPosRegion)).removeTrackedEntity(entity);
         entity.previousTrackedChunkPosRegion = null;
     }
@@ -128,6 +139,11 @@ public class LevelChunkRegionMap {
             entity.previousTrackedChunkPosRegion = newChunkPos;
             getOrCreate(fromRegion).removeTrackedEntity(entity);
             getOrCreate(toRegion).addTrackedEntity(entity);
+
+            if (entity instanceof Mob mob) {
+                getOrCreate(fromRegion).removeNavigationMob(mob);
+                getOrCreate(toRegion).addNavigationMob(mob);
+            }
         }
     }
 
@@ -182,4 +198,29 @@ public class LevelChunkRegionMap {
             getOrCreate(toRegion).addPlayer(player);
         }
     }
+
+    public List<Mob> collectRelevantNavigatingMobs(RegionPos regionPos) {
+        if (!level.chunkScheduler.getRegionLocker().hasLock(regionPos)) {
+            // We care about the navigating mobs in at least this region, ensure it's locked
+            throw new IllegalStateException("Collecting navigating mobs outside of region's thread");
+        }
+
+        ObjectArrayList<Mob> navigatingMobs = new ObjectArrayList<>();
+
+        for (int x = -ShreddedPaperRegionLocker.REGION_LOCK_RADIUS; x <= ShreddedPaperRegionLocker.REGION_LOCK_RADIUS; x++) {
+            for (int z = -ShreddedPaperRegionLocker.REGION_LOCK_RADIUS; z <= ShreddedPaperRegionLocker.REGION_LOCK_RADIUS; z++) {
+                RegionPos i = new RegionPos(regionPos.x + x, regionPos.z + z);
+
+                // Only collect mobs from regions that are locked - if it's not locked, it should be too far away to matter
+                if (!level.chunkScheduler.getRegionLocker().hasLock(i)) continue;
+
+                LevelChunkRegion region = get(i);
+                if (region == null) continue;
+
+                region.collectNavigatingMobs(navigatingMobs);
+            }
+        }
+
+        return navigatingMobs;
+    }
 }
diff --git a/src/main/java/net/minecraft/server/level/ServerChunkCache.java b/src/main/java/net/minecraft/server/level/ServerChunkCache.java
index bcba012eff482031f7160cd113753128ede85760..fc32dbe44f4bbd8da0795c1e573b9666dc99534c 100644
--- a/src/main/java/net/minecraft/server/level/ServerChunkCache.java
+++ b/src/main/java/net/minecraft/server/level/ServerChunkCache.java
@@ -416,6 +416,7 @@ public class ServerChunkCache extends ChunkSource {
     }
 
     public boolean pollTask() {
+        if (ShreddedPaperTickThread.isShreddedPaperTickThread()) return false; // ShreddedPaper - don't run main thread tasks on worker threads
         return this.mainThreadProcessor.pollTask();
     }
 
diff --git a/src/main/java/net/minecraft/server/level/ServerLevel.java b/src/main/java/net/minecraft/server/level/ServerLevel.java
index 5ba89a75bfcfd9f3d28e5bf25cbfeb7eded4cc3a..099251e47f8a8bd303fec64a157f78eeda416f21 100644
--- a/src/main/java/net/minecraft/server/level/ServerLevel.java
+++ b/src/main/java/net/minecraft/server/level/ServerLevel.java
@@ -6,6 +6,7 @@ import com.google.common.collect.Lists;
 import com.mojang.datafixers.DataFixer;
 import com.mojang.datafixers.util.Pair;
 import com.mojang.logging.LogUtils;
+import io.papermc.paper.util.TickThread;
 import it.unimi.dsi.fastutil.ints.Int2ObjectMap;
 import it.unimi.dsi.fastutil.ints.Int2ObjectOpenHashMap;
 import it.unimi.dsi.fastutil.longs.LongSet;
@@ -206,7 +207,7 @@ public class ServerLevel extends Level implements WorldGenLevel {
     public final LevelTicksRegionProxy<Fluid> fluidTicks; // ShreddedPaper
     private final PathTypeCache pathTypesByPosCache;
     final Set<Mob> navigatingMobs;
-    volatile boolean isUpdatingNavigations;
+    final ThreadLocal<Boolean> isUpdatingNavigations = ThreadLocal.withInitial(() -> false); // ShreddedPaper - make thread local
     protected final Raids raids;
     private final ObjectLinkedOpenHashSet<BlockEventData> blockEvents;
     private final List<BlockEventData> blockEventsToReschedule;
@@ -1920,7 +1921,7 @@ public class ServerLevel extends Level implements WorldGenLevel {
 
     @Override
     public void sendBlockUpdated(BlockPos pos, BlockState oldState, BlockState newState, int flags) {
-        if (this.isUpdatingNavigations) {
+        if (this.isUpdatingNavigations.get()) { // ShreddedPaper - use thread local
             String s = "recursive call to sendBlockUpdated";
 
             Util.logAndPauseIfInIde("recursive call to sendBlockUpdated", new IllegalStateException("recursive call to sendBlockUpdated"));
@@ -1934,7 +1935,7 @@ public class ServerLevel extends Level implements WorldGenLevel {
 
         if (Shapes.joinIsNotEmpty(voxelshape, voxelshape1, BooleanOp.NOT_SAME)) {
             List<PathNavigation> list = new ObjectArrayList();
-            Iterator iterator = this.navigatingMobs.iterator();
+            Iterator iterator = chunkSource.tickingRegions.collectRelevantNavigatingMobs(RegionPos.forChunk(new ChunkPos(pos))).iterator(); // ShreddedPaper
 
             while (iterator.hasNext()) {
                 // CraftBukkit start - fix SPIGOT-6362
@@ -1957,7 +1958,7 @@ public class ServerLevel extends Level implements WorldGenLevel {
             }
 
             try {
-                this.isUpdatingNavigations = true;
+                this.isUpdatingNavigations.set(true); // ShreddedPaper - use thread local
                 iterator = list.iterator();
 
                 while (iterator.hasNext()) {
@@ -1966,7 +1967,7 @@ public class ServerLevel extends Level implements WorldGenLevel {
                     navigationabstract1.recomputePath();
                 }
             } finally {
-                this.isUpdatingNavigations = false;
+                this.isUpdatingNavigations.set(false); // ShreddedPaper - use thread local
             }
 
         }
@@ -2784,7 +2785,7 @@ public class ServerLevel extends Level implements WorldGenLevel {
         }
 
         public void onTrackingStart(Entity entity) {
-            org.spigotmc.AsyncCatcher.catchOp("entity register"); // Spigot
+            TickThread.ensureTickThread(entity, "entity register"); // org.spigotmc.AsyncCatcher.catchOp("entity register"); // Spigot // ShreddedPaper
             // ServerLevel.this.getChunkSource().addEntity(entity); // Paper - ignore and warn about illegal addEntity calls instead of crashing server; moved down below valid=true
             if (entity instanceof ServerPlayer entityplayer) {
                 ServerLevel.this.players.add(entityplayer);
@@ -2793,13 +2794,13 @@ public class ServerLevel extends Level implements WorldGenLevel {
             }
 
             if (entity instanceof Mob entityinsentient) {
-                if (false && ServerLevel.this.isUpdatingNavigations) { // Paper - Remove unnecessary onTrackingStart during navigation warning
+                if (false && ServerLevel.this.isUpdatingNavigations.get()) { // Paper - Remove unnecessary onTrackingStart during navigation warning // ShreddedPaper - use thread local
                     String s = "onTrackingStart called during navigation iteration";
 
                     Util.logAndPauseIfInIde("onTrackingStart called during navigation iteration", new IllegalStateException("onTrackingStart called during navigation iteration"));
                 }
 
-                ServerLevel.this.navigatingMobs.add(entityinsentient);
+                // ServerLevel.this.navigatingMobs.add(entityinsentient); // ShreddedPaper - handled ourselves
             }
 
             if (entity instanceof EnderDragon entityenderdragon) {
@@ -2830,7 +2831,7 @@ public class ServerLevel extends Level implements WorldGenLevel {
         }
 
         public void onTrackingEnd(Entity entity) {
-            org.spigotmc.AsyncCatcher.catchOp("entity unregister"); // Spigot
+            TickThread.ensureTickThread(entity, "entity unregister"); // org.spigotmc.AsyncCatcher.catchOp("entity unregister"); // Spigot // ShreddedPaper
             // Spigot start
             if ( entity instanceof Player )
             {
@@ -2875,13 +2876,13 @@ public class ServerLevel extends Level implements WorldGenLevel {
             }
 
             if (entity instanceof Mob entityinsentient) {
-                if (false && ServerLevel.this.isUpdatingNavigations) { // Paper - Remove unnecessary onTrackingStart during navigation warning
+                if (false && ServerLevel.this.isUpdatingNavigations.get()) { // Paper - Remove unnecessary onTrackingStart during navigation warning // ShreddedPaper - use thread local
                     String s = "onTrackingStart called during navigation iteration";
 
                     Util.logAndPauseIfInIde("onTrackingStart called during navigation iteration", new IllegalStateException("onTrackingStart called during navigation iteration"));
                 }
 
-                ServerLevel.this.navigatingMobs.remove(entityinsentient);
+                // ServerLevel.this.navigatingMobs.remove(entityinsentient); // ShreddedPaper - handled ourselves
             }
 
             if (entity instanceof EnderDragon entityenderdragon) {
