From 0000000000000000000000000000000000000000 Mon Sep 17 00:00:00 2001
From: ProdPreva1l <prod.preva1l@gmail.com>
Date: Wed, 5 Jun 2024 20:27:53 +1000
Subject: [PATCH] Add various API for Folia plugin compatibility


diff --git a/src/main/java/io/papermc/paper/threadedregions/RegionizedServer.java b/src/main/java/io/papermc/paper/threadedregions/RegionizedServer.java
new file mode 100644
index 0000000000000000000000000000000000000000..69db3090c24ab058549cdd8e78ca1beb7be5e7a8
--- /dev/null
+++ b/src/main/java/io/papermc/paper/threadedregions/RegionizedServer.java
@@ -0,0 +1,7 @@
+package io.papermc.paper.threadedregions;
+
+@SuppressWarnings("unused")
+public class RegionizedServer {
+    /* We have nothing here because this is just to fix compat
+     for Folia plugins that have hard coded checks for Folia */
+}
diff --git a/src/main/java/net/minecraft/server/MinecraftServer.java b/src/main/java/net/minecraft/server/MinecraftServer.java
index a51cf2bed02780f741f0465e23a76413a85289e1..ae722ce431941c46e1ae05d51735306586fd702b 100644
--- a/src/main/java/net/minecraft/server/MinecraftServer.java
+++ b/src/main/java/net/minecraft/server/MinecraftServer.java
@@ -17,6 +17,7 @@ import com.mojang.datafixers.DataFixer;
 import com.mojang.logging.LogUtils;
 import io.multipaper.shreddedpaper.ShreddedPaper;
 import io.multipaper.shreddedpaper.threading.ShreddedPaperPlayerTicker;
+import io.papermc.paper.threadedregions.RegionizedServerInitEvent;
 import it.unimi.dsi.fastutil.longs.LongIterator;
 import it.unimi.dsi.fastutil.objects.ObjectArrayList;
 import it.unimi.dsi.fastutil.objects.ObjectArraySet;
@@ -1249,6 +1250,8 @@ public abstract class MinecraftServer extends ReentrantBlockableEventLoop<TickTa
             this.statusIcon = (ServerStatus.Favicon) this.loadStatusIcon().orElse(null); // CraftBukkit - decompile error
             this.status = this.buildServerStatus();
 
+            new RegionizedServerInitEvent().callEvent(); // ShreddedPaper - Folia plugin support
+
             // Spigot start
             // Paper start - Improved Watchdog Support
             LOGGER.info("Running delayed init tasks");
diff --git a/src/main/java/org/bukkit/craftbukkit/CraftServer.java b/src/main/java/org/bukkit/craftbukkit/CraftServer.java
index fab5e3420e05139722590dc643b278ebf4a100a0..2cb6d417f6d51ac32ac63b41d556675c1f76a216 100644
--- a/src/main/java/org/bukkit/craftbukkit/CraftServer.java
+++ b/src/main/java/org/bukkit/craftbukkit/CraftServer.java
@@ -1,5 +1,6 @@
 package org.bukkit.craftbukkit;
 
+import ca.spottedleaf.moonrise.common.util.TickThread;
 import com.google.common.base.Charsets;
 import com.google.common.base.Function;
 import com.google.common.base.Joiner;
@@ -13,6 +14,7 @@ import com.mojang.brigadier.StringReader;
 import com.mojang.brigadier.exceptions.CommandSyntaxException;
 import com.mojang.serialization.Dynamic;
 import com.mojang.serialization.Lifecycle;
+import io.multipaper.shreddedpaper.threading.ShreddedPaperTickThread;
 import it.unimi.dsi.fastutil.objects.Object2IntOpenHashMap;
 import java.awt.image.BufferedImage;
 import java.io.ByteArrayOutputStream;
@@ -383,6 +385,13 @@ public final class CraftServer implements Server {
     }
     // Paper end - Folia reagion threading API
 
+    // ShreddedPaper start - Folia region api
+    @Override
+    public boolean isGlobalTickThread() {
+        return TickThread.isTickThread() && !ShreddedPaperTickThread.isShreddedPaperTickThread();
+    }
+    // ShreddedPaper end - Folia region api
+
     static {
         ConfigurationSerialization.registerClass(CraftOfflinePlayer.class);
         ConfigurationSerialization.registerClass(CraftPlayerProfile.class);
