From 0000000000000000000000000000000000000000 Mon Sep 17 00:00:00 2001
From: PureGero <puregero@gmail.com>
Date: Wed, 26 Jun 2024 10:23:46 +0900
Subject: [PATCH] /kill command


diff --git a/src/main/java/io/multipaper/shreddedpaper/ShreddedPaper.java b/src/main/java/io/multipaper/shreddedpaper/ShreddedPaper.java
index d3ff21ac78f4e5c64bbb1e72aecf7397164cabab..7aebc3ecb70fc42b795460a0d21b66776cf19004 100644
--- a/src/main/java/io/multipaper/shreddedpaper/ShreddedPaper.java
+++ b/src/main/java/io/multipaper/shreddedpaper/ShreddedPaper.java
@@ -8,6 +8,8 @@ import org.bukkit.Location;
 import org.bukkit.craftbukkit.CraftWorld;
 import io.multipaper.shreddedpaper.region.RegionPos;
 
+import java.util.function.Consumer;
+
 public class ShreddedPaper {
 
     public static void runSync(Location location, Runnable runnable) {
@@ -38,6 +40,14 @@ public class ShreddedPaper {
         }
     }
 
+    public static void ensureSync(Entity entity, Consumer<Entity> consumer) {
+        if (!isSync((ServerLevel) entity.level(), entity.chunkPosition())) {
+            entity.getBukkitEntity().taskScheduler.schedule(consumer, null, 1);
+        } else {
+            consumer.accept(entity);
+        }
+    }
+
     public static void ensureSync(ServerLevel serverLevel, BlockPos blockPos, Runnable runnable) {
         ensureSync(serverLevel, new ChunkPos(blockPos), runnable);
     }
diff --git a/src/main/java/net/minecraft/server/commands/KillCommand.java b/src/main/java/net/minecraft/server/commands/KillCommand.java
index c2974a6bd6851b54d1df2689195d896baf4906ee..610e43fa1b388e34ec50d03f5c87ab3372f64822 100644
--- a/src/main/java/net/minecraft/server/commands/KillCommand.java
+++ b/src/main/java/net/minecraft/server/commands/KillCommand.java
@@ -3,6 +3,9 @@ package net.minecraft.server.commands;
 import com.google.common.collect.ImmutableList;
 import com.mojang.brigadier.CommandDispatcher;
 import java.util.Collection;
+import java.util.function.Consumer;
+
+import io.multipaper.shreddedpaper.ShreddedPaper;
 import net.minecraft.commands.CommandSourceStack;
 import net.minecraft.commands.Commands;
 import net.minecraft.commands.arguments.EntityArgument;
@@ -24,7 +27,7 @@ public class KillCommand {
 
     private static int kill(CommandSourceStack source, Collection<? extends Entity> targets) {
         for (Entity entity : targets) {
-            entity.kill();
+            ShreddedPaper.ensureSync(entity, Entity::kill); // ShreddedPaper - run on right thread
         }
 
         if (targets.size() == 1) {
