From 0000000000000000000000000000000000000000 Mon Sep 17 00:00:00 2001
From: PureGero <puregero@gmail.com>
Date: Wed, 5 Jun 2024 20:01:02 +0900
Subject: [PATCH] Block events


diff --git a/src/main/java/io/multipaper/shreddedpaper/region/LevelChunkRegion.java b/src/main/java/io/multipaper/shreddedpaper/region/LevelChunkRegion.java
index f92725d4da17326988809c382046fe4d20d6ba0f..b94453f7f06ec1c25dc01a72a719a74d8c1c4bc4 100644
--- a/src/main/java/io/multipaper/shreddedpaper/region/LevelChunkRegion.java
+++ b/src/main/java/io/multipaper/shreddedpaper/region/LevelChunkRegion.java
@@ -4,12 +4,14 @@ import ca.spottedleaf.concurrentutil.executor.standard.PrioritisedThreadedTaskQu
 import io.papermc.paper.util.maplist.IteratorSafeOrderedReferenceSet;
 import it.unimi.dsi.fastutil.longs.LongLinkedOpenHashSet;
 import it.unimi.dsi.fastutil.objects.ObjectArrayList;
+import it.unimi.dsi.fastutil.objects.ObjectLinkedOpenHashSet;
 import it.unimi.dsi.fastutil.objects.ObjectOpenHashSet;
 import it.unimi.dsi.fastutil.objects.ReferenceArrayList;
 import net.minecraft.server.level.ServerLevel;
 import net.minecraft.server.level.ServerPlayer;
 import net.minecraft.world.entity.Entity;
 import net.minecraft.world.entity.Mob;
+import net.minecraft.world.level.BlockEventData;
 import net.minecraft.world.level.block.entity.TickingBlockEntity;
 import net.minecraft.world.level.chunk.LevelChunk;
 
@@ -31,6 +33,7 @@ public class LevelChunkRegion {
     public final LongLinkedOpenHashSet unloadQueue = new LongLinkedOpenHashSet();
     public final List<TickingBlockEntity> tickingBlockEntities = new ReferenceArrayList<>();
     public final List<TickingBlockEntity> pendingBlockEntityTickers = new ReferenceArrayList<>();
+    private final ObjectLinkedOpenHashSet<BlockEventData> blockEvents = new ObjectLinkedOpenHashSet<>();
 
     public LevelChunkRegion(ServerLevel level, RegionPos regionPos) {
         this.level = level;
@@ -143,9 +146,21 @@ public class LevelChunkRegion {
         toRun.forEach(DelayedTask::run);
     }
 
+    public synchronized void addBlockEvent(BlockEventData blockEvent) {
+        blockEvents.add(blockEvent);
+    }
+
+    public synchronized List<BlockEventData> takeBlockEvents() {
+        List<BlockEventData> retValue = blockEvents.isEmpty() ? List.of() : new ArrayList<>(blockEvents);
+        blockEvents.clear();
+        return retValue;
+    }
+
     public boolean isEmpty() {
         return levelChunks.isEmpty() && scheduledTasks.isEmpty() && internalTasks.getTotalTasksExecuted() >= internalTasks.getTotalTasksScheduled() && navigatingMobs.isEmpty()
                 && players.isEmpty()
-                && unloadQueue.isEmpty();
+                && unloadQueue.isEmpty()
+                && blockEvents.isEmpty()
+                ;
     }
 }
diff --git a/src/main/java/io/multipaper/shreddedpaper/region/LevelChunkRegionMap.java b/src/main/java/io/multipaper/shreddedpaper/region/LevelChunkRegionMap.java
index ae8581242d52b4422f95b8fd3581e9ae2b442ab1..bd9c18c72d8e99267c9e7541481ce10564e9f5e5 100644
--- a/src/main/java/io/multipaper/shreddedpaper/region/LevelChunkRegionMap.java
+++ b/src/main/java/io/multipaper/shreddedpaper/region/LevelChunkRegionMap.java
@@ -6,6 +6,7 @@ import net.minecraft.server.level.ServerLevel;
 import net.minecraft.server.level.ServerPlayer;
 import net.minecraft.world.entity.Entity;
 import net.minecraft.world.entity.Mob;
+import net.minecraft.world.level.BlockEventData;
 import net.minecraft.world.level.ChunkPos;
 import net.minecraft.world.level.chunk.LevelChunk;
 import io.multipaper.shreddedpaper.threading.ShreddedPaperRegionLocker;
@@ -183,4 +184,8 @@ public class LevelChunkRegionMap {
             getOrCreate(toRegion).addPlayer(player);
         }
     }
+
+    public void addBlockEvent(BlockEventData blockEvent) {
+        getOrCreate(RegionPos.forBlockPos(blockEvent.pos())).addBlockEvent(blockEvent);
+    }
 }
\ No newline at end of file
diff --git a/src/main/java/io/multipaper/shreddedpaper/threading/ShreddedPaperChunkTicker.java b/src/main/java/io/multipaper/shreddedpaper/threading/ShreddedPaperChunkTicker.java
index 627eaf4d0f9204a25d3b8524461a6723bb204065..82447a7e4ab08ab0a333e8549c646170a9076fe9 100644
--- a/src/main/java/io/multipaper/shreddedpaper/threading/ShreddedPaperChunkTicker.java
+++ b/src/main/java/io/multipaper/shreddedpaper/threading/ShreddedPaperChunkTicker.java
@@ -82,6 +82,8 @@ public class ShreddedPaperChunkTicker {
 
             region.forEach(chunk -> _tickChunk(level, chunk, spawnercreature_d));
 
+            level.runBlockEvents(region.takeBlockEvents());
+
             region.forEachTickingEntity(ShreddedPaperEntityTicker::tickEntity);
 
             level.tickBlockEntities(region.tickingBlockEntities, region.pendingBlockEntityTickers);
diff --git a/src/main/java/net/minecraft/server/level/ServerLevel.java b/src/main/java/net/minecraft/server/level/ServerLevel.java
index d98cae9c61dc1832cb73b6acfdd5f49bc70ed286..6c0dd6ab9d10ec4e39da1944d96687c2d4b405e4 100644
--- a/src/main/java/net/minecraft/server/level/ServerLevel.java
+++ b/src/main/java/net/minecraft/server/level/ServerLevel.java
@@ -23,6 +23,7 @@ import java.io.Writer;
 import java.nio.file.Files;
 import java.nio.file.Path;
 import java.util.ArrayList;
+import java.util.Collection;
 import java.util.Comparator;
 import java.util.Iterator;
 import java.util.List;
@@ -903,7 +904,7 @@ public class ServerLevel extends Level implements WorldGenLevel {
         //gameprofilerfiller.popPush("blockEvents"); // Purpur
         if (flag) {
             // this.timings.doSounds.startTiming(); // Spigot // Purpur
-            this.runBlockEvents();
+            // this.runBlockEvents(); // ShreddedPaper - handled locally in the region
             // this.timings.doSounds.stopTiming(); // Spigot // Purpur
         }
 
@@ -2044,25 +2045,25 @@ public class ServerLevel extends Level implements WorldGenLevel {
 
     @Override
     public void blockEvent(BlockPos pos, Block block, int type, int data) {
-        this.blockEvents.add(new BlockEventData(pos, block, type, data));
+        this.chunkSource.tickingRegions.addBlockEvent(new BlockEventData(pos, block, type, data)); // ShreddedPaper
     }
 
-    private void runBlockEvents() {
-        this.blockEventsToReschedule.clear();
+    public void runBlockEvents(List<BlockEventData> blockEvents) { // ShreddedPaper
+        // this.blockEventsToReschedule.clear(); // ShreddedPaper
 
-        while (!this.blockEvents.isEmpty()) {
-            BlockEventData blockactiondata = (BlockEventData) this.blockEvents.removeFirst();
+        for (int i = 0; i < blockEvents.size(); i++) { // ShreddedPaper
+            BlockEventData blockactiondata = (BlockEventData) blockEvents.get(i); // ShreddedPaper
 
             if (this.shouldTickBlocksAt(blockactiondata.pos())) {
                 if (this.doBlockEvent(blockactiondata)) {
                     this.server.getPlayerList().broadcast((Player) null, (double) blockactiondata.pos().getX(), (double) blockactiondata.pos().getY(), (double) blockactiondata.pos().getZ(), 64.0D, this.dimension(), new ClientboundBlockEventPacket(blockactiondata.pos(), blockactiondata.block(), blockactiondata.paramA(), blockactiondata.paramB()));
                 }
             } else {
-                this.blockEventsToReschedule.add(blockactiondata);
+                this.chunkSource.tickingRegions.addBlockEvent(blockactiondata); // ShreddedPaper
             }
         }
 
-        this.blockEvents.addAll(this.blockEventsToReschedule);
+        // this.blockEvents.addAll(this.blockEventsToReschedule); // ShreddedPaper
     }
 
     private boolean doBlockEvent(BlockEventData event) {
