From 0000000000000000000000000000000000000000 Mon Sep 17 00:00:00 2001
From: PureGero <puregero@gmail.com>
Date: Thu, 6 Jun 2024 17:09:45 +0900
Subject: [PATCH] 
 allow-unsupported-plugins-to-modify-chunks-via-global-scheduler


diff --git a/src/main/java/io/multipaper/shreddedpaper/threading/ShreddedPaperChunkTicker.java b/src/main/java/io/multipaper/shreddedpaper/threading/ShreddedPaperChunkTicker.java
index f32aecd69d929c005ca8226054c218c7dd37e6fc..3ebb70ecb854ba0172148e48a07578dc9c90d853 100644
--- a/src/main/java/io/multipaper/shreddedpaper/threading/ShreddedPaperChunkTicker.java
+++ b/src/main/java/io/multipaper/shreddedpaper/threading/ShreddedPaperChunkTicker.java
@@ -21,11 +21,15 @@ import java.util.concurrent.CompletableFuture;
 
 public class ShreddedPaperChunkTicker {
 
+    public static boolean tickingChunks = false;
+
     public static void tickChunks(ServerLevel level, NaturalSpawner.SpawnState spawnercreature_d) {
         List<CompletableFuture<Void>> futures = new ArrayList<>();
 
         MinecraftServer.getServer().executeMidTickTasks();
 
+        tickingChunks = true;
+
         level.chunkSource.tickingRegions.forEach(
                 region -> futures.add(tickRegion(level, region, spawnercreature_d))
         );
@@ -36,6 +40,8 @@ public class ShreddedPaperChunkTicker {
             }
         }
 
+        tickingChunks = false;
+
         MinecraftServer.getServer().executeMidTickTasks();
     }
 
diff --git a/src/main/java/io/multipaper/shreddedpaper/threading/ShreddedPaperRegionLocker.java b/src/main/java/io/multipaper/shreddedpaper/threading/ShreddedPaperRegionLocker.java
index 5e5cf3cb72ab51a20fa5c91fb9b7f0f942ae9cc8..b30147827ffa329c09703f48f01733c34f5e63e3 100644
--- a/src/main/java/io/multipaper/shreddedpaper/threading/ShreddedPaperRegionLocker.java
+++ b/src/main/java/io/multipaper/shreddedpaper/threading/ShreddedPaperRegionLocker.java
@@ -1,5 +1,6 @@
 package io.multipaper.shreddedpaper.threading;
 
+import io.papermc.paper.util.TickThread;
 import it.unimi.dsi.fastutil.objects.ObjectOpenHashSet;
 import io.multipaper.shreddedpaper.region.RegionPos;
 
@@ -40,7 +41,7 @@ public class ShreddedPaperRegionLocker {
      * Checks if the current thread holds a read lock for the given region
      */
     public boolean hasLock(RegionPos regionPos) {
-        return localLocks.get().contains(regionPos);
+        return localLocks.get().contains(regionPos) || TickThread.canBypassTickThreadCheck();
     }
 
     /**
@@ -49,7 +50,7 @@ public class ShreddedPaperRegionLocker {
      * syncing conflicts with other servers.
      */
     public boolean hasWriteLock(RegionPos regionPos) {
-        return writeLocks.get().contains(regionPos);
+        return writeLocks.get().contains(regionPos) || TickThread.canBypassTickThreadCheck();
     }
 
     /**
diff --git a/src/main/java/io/papermc/paper/util/TickThread.java b/src/main/java/io/papermc/paper/util/TickThread.java
index 313f30948c9ebd09094c1623bfe2c3594f30077c..2cc58dacd64689415b9fe1e35ff8d23ad103f5fd 100644
--- a/src/main/java/io/papermc/paper/util/TickThread.java
+++ b/src/main/java/io/papermc/paper/util/TickThread.java
@@ -1,6 +1,9 @@
 package io.papermc.paper.util;
 
+import io.multipaper.shreddedpaper.config.ShreddedPaperConfiguration;
+import io.multipaper.shreddedpaper.threading.ShreddedPaperChunkTicker;
 import io.multipaper.shreddedpaper.threading.ShreddedPaperTickThread;
+import io.multipaper.shreddedpaper.threading.SynchronousPluginExecution;
 import net.minecraft.core.BlockPos;
 import net.minecraft.server.MinecraftServer;
 import net.minecraft.server.level.ServerLevel;
@@ -69,6 +72,16 @@ public class TickThread extends Thread {
     }
     // ShreddedPaper end
 
+    // ShreddedPaper start
+    public static boolean canBypassTickThreadCheck() {
+        return
+                ShreddedPaperConfiguration.get().multithreading.allowUnsupportedPluginsToModifyChunksViaGlobalScheduler &&
+                SynchronousPluginExecution.getCurrentPlugin() != null &&
+                Thread.currentThread() == MinecraftServer.getServer().getRunningThread() &&
+                !ShreddedPaperChunkTicker.tickingChunks;
+    }
+    // ShreddedPaper end
+
     /**
      * @deprecated
      */
