From 0000000000000000000000000000000000000000 Mon Sep 17 00:00:00 2001
From: PureGero <puregero@gmail.com>
Date: Sat, 20 Jul 2024 01:23:29 +0900
Subject: [PATCH] Debug dependency list print


diff --git a/src/main/java/io/multipaper/shreddedpaper/threading/SynchronousPluginExecution.java b/src/main/java/io/multipaper/shreddedpaper/threading/SynchronousPluginExecution.java
index 9ad9e19bcd29cb94c4395d48f1c78a6dd9758bc6..c625c04eb6286f842c6705446c5c5e3868192c16 100644
--- a/src/main/java/io/multipaper/shreddedpaper/threading/SynchronousPluginExecution.java
+++ b/src/main/java/io/multipaper/shreddedpaper/threading/SynchronousPluginExecution.java
@@ -60,6 +60,7 @@ public class SynchronousPluginExecution {
             pluginsToLock = cachedDependencyLists.computeIfAbsent(plugin.getName(), (name) -> {
                 TreeSet<String> dependencyList = new TreeSet<>(Comparator.naturalOrder());
                 fillPluginsToLock(plugin, dependencyList);
+                LOGGER.info("Dependency list calculated for " + plugin.getName() + ": " + dependencyList);
                 return new ArrayList<>(dependencyList);
             });
         }
