From 0000000000000000000000000000000000000000 Mon Sep 17 00:00:00 2001
From: PureGero <puregero@gmail.com>
Date: Sat, 20 Jul 2024 01:23:29 +0900
Subject: [PATCH] Debug dependency list print


diff --git a/src/main/java/io/multipaper/shreddedpaper/threading/SynchronousPluginExecution.java b/src/main/java/io/multipaper/shreddedpaper/threading/SynchronousPluginExecution.java
index 593c40c5726d6a958c59178cf8e48e1209bb2330..1dad8cd9a882259ba6faefa4f3c5625519dd416d 100644
--- a/src/main/java/io/multipaper/shreddedpaper/threading/SynchronousPluginExecution.java
+++ b/src/main/java/io/multipaper/shreddedpaper/threading/SynchronousPluginExecution.java
@@ -57,6 +57,7 @@ public class SynchronousPluginExecution {
             pluginsToLock = cachedDependencyLists.computeIfAbsent(plugin.getName(), (name) -> {
                 TreeSet<String> dependencyList = new TreeSet<>(Comparator.naturalOrder());
                 fillPluginsToLock(plugin, dependencyList);
+                LOGGER.info("Dependency list calculated for " + plugin.getName() + ": " + dependencyList);
                 return new ArrayList<>(dependencyList);
             });
         }
