From 0000000000000000000000000000000000000000 Mon Sep 17 00:00:00 2001
From: PureGero <puregero@gmail.com>
Date: Sat, 20 Jul 2024 01:23:29 +0900
Subject: [PATCH] Debug dependency list print


diff --git a/src/main/java/io/multipaper/shreddedpaper/threading/SynchronousPluginExecution.java b/src/main/java/io/multipaper/shreddedpaper/threading/SynchronousPluginExecution.java
index 3625ee53f17766aae695c0d9e35d755e331e6c56..4ac12d5e57177a70d323dc8f241588cf8eff561e 100644
--- a/src/main/java/io/multipaper/shreddedpaper/threading/SynchronousPluginExecution.java
+++ b/src/main/java/io/multipaper/shreddedpaper/threading/SynchronousPluginExecution.java
@@ -56,6 +56,7 @@ public class SynchronousPluginExecution {
             pluginsToLock = cachedDependencyLists.computeIfAbsent(plugin.getName(), (name) -> {
                 TreeSet<String> dependencyList = new TreeSet<>(Comparator.naturalOrder());
                 fillPluginsToLock(plugin, dependencyList);
+                LOGGER.info("Dependency list calculated for " + plugin.getName() + ": " + dependencyList);
                 return new ArrayList<>(dependencyList);
             });
         }
