From 0000000000000000000000000000000000000000 Mon Sep 17 00:00:00 2001
From: PureGero <puregero@gmail.com>
Date: Sat, 3 Aug 2024 19:37:13 +0900
Subject: [PATCH] handlingTick


diff --git a/src/main/java/io/multipaper/shreddedpaper/threading/ShreddedPaperChunkTicker.java b/src/main/java/io/multipaper/shreddedpaper/threading/ShreddedPaperChunkTicker.java
index 327eadaab7d0df52e6bb392fe29ef4ee381163ed..7933c7cb4ca7b5292529f93ac23f48af4754cda0 100644
--- a/src/main/java/io/multipaper/shreddedpaper/threading/ShreddedPaperChunkTicker.java
+++ b/src/main/java/io/multipaper/shreddedpaper/threading/ShreddedPaperChunkTicker.java
@@ -77,6 +77,8 @@ public class ShreddedPaperChunkTicker {
 
             region.tickTasks();
 
+            level.handlingTickThreadLocal.set(true);
+
             level.blockTicks.tick(region.getRegionPos(), level.getGameTime(), level.paperConfig().environment.maxBlockTicks, level::tickBlock);
             level.fluidTicks.tick(region.getRegionPos(), level.getGameTime(), level.paperConfig().environment.maxBlockTicks, level::tickFluid);
 
@@ -84,6 +86,8 @@ public class ShreddedPaperChunkTicker {
 
             level.runBlockEvents(region);
 
+            level.handlingTickThreadLocal.set(false);
+
             region.forEachTickingEntity(ShreddedPaperEntityTicker::tickEntity);
 
             region.forEachTrackedEntity(ShreddedPaperEntityTicker::processTrackQueue);
diff --git a/src/main/java/net/minecraft/server/level/ServerLevel.java b/src/main/java/net/minecraft/server/level/ServerLevel.java
index b6ffd2ac046777466657eaedbf13f2cbaa8456b3..5f99874eb7d8e33686c2fb662e1220d05883fd0a 100644
--- a/src/main/java/net/minecraft/server/level/ServerLevel.java
+++ b/src/main/java/net/minecraft/server/level/ServerLevel.java
@@ -210,7 +210,7 @@ public class ServerLevel extends Level implements WorldGenLevel {
     protected final Raids raids;
     // private final ObjectLinkedOpenHashSet<BlockEventData> blockEvents; // ShreddedPaper - moved into each region
     private final ThreadLocal<List<BlockEventData>> blockEventsToRescheduleThreadLocal; // ShreddedPaper
-    private boolean handlingTick;
+    public ThreadLocal<Boolean> handlingTickThreadLocal = ThreadLocal.withInitial(() -> false); // ShreddedPaper
     private final List<CustomSpawner> customSpawners;
     @Nullable
     private EndDragonFight dragonFight;
@@ -837,7 +837,7 @@ public class ServerLevel extends Level implements WorldGenLevel {
     public CompletableFuture<Void> tick(BooleanSupplier shouldKeepTicking) { // ShreddedPaper - run async
         //ProfilerFiller gameprofilerfiller = this.getProfiler(); // Purpur
 
-        this.handlingTick = true;
+        this.handlingTickThreadLocal.set(true); // ShreddedPaper
         TickRateManager tickratemanager = this.tickRateManager();
         boolean flag = tickratemanager.runsNormally();
 
@@ -906,7 +906,7 @@ public class ServerLevel extends Level implements WorldGenLevel {
             // this.timings.doSounds.stopTiming(); // Spigot // Purpur
         }
 
-        this.handlingTick = false;
+        this.handlingTickThreadLocal.set(false); // ShreddedPaper
         //gameprofilerfiller.pop(); // Purpur
         boolean flag1 = !paperConfig().unsupportedSettings.disableWorldTickingWhenEmpty || !this.players.isEmpty() || !this.getForcedChunks().isEmpty(); // CraftBukkit - this prevents entity cleanup, other issues on servers with no players // Paper - restore this
 
@@ -1225,7 +1225,7 @@ public class ServerLevel extends Level implements WorldGenLevel {
     }
 
     public boolean isHandlingTick() {
-        return this.handlingTick;
+        return this.handlingTickThreadLocal.get();
     }
 
     public boolean canSleepThroughNights() {
