From 0000000000000000000000000000000000000000 Mon Sep 17 00:00:00 2001
From: PureGero <puregero@gmail.com>
Date: Wed, 14 Aug 2024 01:13:18 +0900
Subject: [PATCH] Send ping packet later


diff --git a/src/main/java/net/minecraft/server/MinecraftServer.java b/src/main/java/net/minecraft/server/MinecraftServer.java
index 76a105451cce3a264be2a25ed15c6632e229d86b..1717f6e9b2d46b2867aec0bb51f79e873db9672e 100644
--- a/src/main/java/net/minecraft/server/MinecraftServer.java
+++ b/src/main/java/net/minecraft/server/MinecraftServer.java
@@ -1922,6 +1922,7 @@ public abstract class MinecraftServer extends ReentrantBlockableEventLoop<TickTa
             ServerPlayer entityplayer = (ServerPlayer) iterator.next();
 
             entityplayer.connection.chunkSender.sendNextChunks(entityplayer);
+            entityplayer.connection.keepConnectionAlive(); // ShreddedPaper - send the ping packet right before flushing
             entityplayer.connection.resumeFlushing();
         }
 
diff --git a/src/main/java/net/minecraft/server/network/ServerCommonPacketListenerImpl.java b/src/main/java/net/minecraft/server/network/ServerCommonPacketListenerImpl.java
index d142bb4572a9cacfe61f3d09fa97fc26aa446588..f01b28732851f26465c9e246fe0dc35eea457d76 100644
--- a/src/main/java/net/minecraft/server/network/ServerCommonPacketListenerImpl.java
+++ b/src/main/java/net/minecraft/server/network/ServerCommonPacketListenerImpl.java
@@ -269,7 +269,7 @@ public abstract class ServerCommonPacketListenerImpl implements ServerCommonPack
         this.disconnect(ServerCommonPacketListenerImpl.DISCONNECT_UNEXPECTED_QUERY, PlayerKickEvent.Cause.INVALID_COOKIE); // Paper - kick event cause
     }
 
-    protected void keepConnectionAlive() {
+    public void keepConnectionAlive() { // ShreddedPaper - make public
         //this.server.getProfiler().push("keepAlive"); // Purpur
         // Paper start - give clients a longer time to respond to pings as per pre 1.12.2 timings
         // This should effectively place the keepalive handling back to "as it was" before 1.12.2
diff --git a/src/main/java/net/minecraft/server/network/ServerGamePacketListenerImpl.java b/src/main/java/net/minecraft/server/network/ServerGamePacketListenerImpl.java
index 640ac03286b7653888497c9cee8864101488ac42..2b25e48d03cc90ca9c0cb1caaf07ee2edd9d0e50 100644
--- a/src/main/java/net/minecraft/server/network/ServerGamePacketListenerImpl.java
+++ b/src/main/java/net/minecraft/server/network/ServerGamePacketListenerImpl.java
@@ -397,7 +397,7 @@ public class ServerGamePacketListenerImpl extends ServerCommonPacketListenerImpl
             this.aboveGroundVehicleTickCount = 0;
         }
 
-        this.keepConnectionAlive();
+        // this.keepConnectionAlive(); // ShreddedPaper - called elsewhere
         // CraftBukkit start
         for (int spam; (spam = this.chatSpamTickCount.get()) > 0 && !this.chatSpamTickCount.compareAndSet(spam, spam - 1); ) ;
         if (tabSpamLimiter.get() > 0) tabSpamLimiter.getAndDecrement(); // Paper - configurable tab spam limits
