From 0000000000000000000000000000000000000000 Mon Sep 17 00:00:00 2001
From: PureGero <puregero@gmail.com>
Date: Tue, 6 Aug 2024 22:08:49 +0900
Subject: [PATCH] PotentialCalculator


diff --git a/src/main/java/net/minecraft/world/level/PotentialCalculator.java b/src/main/java/net/minecraft/world/level/PotentialCalculator.java
index d18eedff64d1dbf5257798c5d41da7ef76d19e3e..c4e80757fda67ec763545fcfde6be6299ba4d63f 100644
--- a/src/main/java/net/minecraft/world/level/PotentialCalculator.java
+++ b/src/main/java/net/minecraft/world/level/PotentialCalculator.java
@@ -9,7 +9,7 @@ public class PotentialCalculator {
 
     public void addCharge(BlockPos pos, double mass) {
         if (mass != 0.0) {
-            this.charges.add(new PotentialCalculator.PointCharge(pos, mass));
+            this.charges.add(new PotentialCalculator.PointCharge(pos, mass)); // ShreddedPaper - doesn't need a lock, if a value isn't added it's not a big deal
         }
     }
 
@@ -19,7 +19,8 @@ public class PotentialCalculator {
         } else {
             double d = 0.0;
 
-            for (PotentialCalculator.PointCharge pointCharge : this.charges) {
+            for (int i = 0; i < this.charges.size(); i++) { // ShreddedPaper - prevent CME - list is only ever added to
+                PotentialCalculator.PointCharge pointCharge = this.charges.get(i); // ShreddedPaper - prevent CME - list is only ever added to
                 d += pointCharge.getPotentialChange(pos);
             }
 
