From 0000000000000000000000000000000000000000 Mon Sep 17 00:00:00 2001
From: PureGero <puregero@gmail.com>
Date: Thu, 1 May 2025 23:21:56 +0900
Subject: [PATCH] Optimization: vanish api


diff --git a/src/main/java/org/bukkit/craftbukkit/entity/CraftPlayer.java b/src/main/java/org/bukkit/craftbukkit/entity/CraftPlayer.java
index 494920321af0098e390d6b2669403acc97dab984..cf651f132a125cded54bd36014da8b61dd571f5a 100644
--- a/src/main/java/org/bukkit/craftbukkit/entity/CraftPlayer.java
+++ b/src/main/java/org/bukkit/craftbukkit/entity/CraftPlayer.java
@@ -5,6 +5,7 @@ import com.google.common.collect.ImmutableSet;
 import com.google.common.io.BaseEncoding;
 import com.mojang.authlib.GameProfile;
 import com.mojang.datafixers.util.Pair;
+import io.multipaper.shreddedpaper.config.ShreddedPaperConfiguration;
 import io.netty.buffer.Unpooled;
 import io.papermc.paper.util.TickThread;
 import it.unimi.dsi.fastutil.shorts.ShortArraySet;
@@ -2235,15 +2236,18 @@ public class CraftPlayer extends CraftHumanEntity implements Player {
 
     @Override
     public boolean canSee(Player player) {
+        if (ShreddedPaperConfiguration.get().optimizations.disableVanishApi) return true; // ShreddedPaper - optimize canSee
         return this.canSee((org.bukkit.entity.Entity) player);
     }
 
     @Override
     public boolean canSee(org.bukkit.entity.Entity entity) {
+        if (ShreddedPaperConfiguration.get().optimizations.disableVanishApi) return true; // ShreddedPaper - optimize canSee
         return this.equals(entity) || entity.isVisibleByDefault() ^ this.invertedVisibilityEntities.containsKey(entity.getUniqueId()); // SPIGOT-7312: Can always see self
     }
 
     public boolean canSeePlayer(UUID uuid) {
+        if (ShreddedPaperConfiguration.get().optimizations.disableVanishApi) return true; // ShreddedPaper - optimize canSee
         org.bukkit.entity.Entity entity = this.getServer().getPlayer(uuid);
 
         return (entity != null) ? this.canSee(entity) : false; // If we can't find it, we can't see it
