--- a/src/main/java/org/bukkit/craftbukkit/Main.java
+++ b/src/main/java/org/bukkit/craftbukkit/Main.java
@@ -177,6 +_,14 @@
                     .describedAs("Yml file");
                 // Purpur end - Purpur config files
 
+                // ShreddedPaper start
+                acceptsAll(asList("shreddedpaper", "shreddedpaper-settings"), "File for shreddedpaper settings")
+                        .withRequiredArg()
+                        .ofType(File.class)
+                        .defaultsTo(new File("shreddedpaper.yml"))
+                        .describedAs("Yml file");
+                // ShreddedPaper end
+
                 this.accepts("server-name", "Name of the server")
                         .withRequiredArg()
                         .ofType(String.class)
